---
title: "rRNADepletion_s4ter"
author: "Lachlan Baer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    fig_width: 8
    fig_height: 6
    fig_align: "center"
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  error = FALSE, 
  fig.align = "center"
)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

# Setup 

```{r packages}
library(tidyverse)
library(magrittr)
library(ngsReports)
library(here)
library(scales)
library(ggpubr)
library(kableExtra)
library(AnnotationHub)
library(ensembldb)
library(edgeR)
library(corrplot)
library(DT)
library(pander)
library(ggrepel)
library(pheatmap)
library(ggdendro)
library(Biostrings)
library(Gviz)
library(Rsamtools)
library(viridis)
library(ggfortify)
```

```{r options}
if (interactive()) setwd(here::here())
theme_set(theme_bw())
cores <- detectCores() - 1
```

# Coverage patterns

## Plot setup

```{r GvizOpts}
options(ucscChromosomeNames = FALSE)
```

```{r exons}
exons_Dr <- exonsBy(
  ensDb_Dr,
  by = "tx",
  columns = c("exon_id", "gene_id", "tx_id", "gene_name"),
  filter = GeneIdFilter(c(deDown, deUp))) %>%
  unlist()
colnames(mcols(exons_Dr)) <- str_remove_all(colnames(mcols(exons_Dr)), "_id") %>%
  str_replace_all("tx", "transcript")
mcols(exons_Dr) <- mcols(exons_Dr)[c("gene_name", "gene", "exon", "transcript")]
```

```{r orders}
order <- dgeList$samples %>%
  dplyr::arrange(filenames) %>%
  as_tibble() %>% 
  rownames_to_column("order") %>% 
  dplyr::arrange(-rRNA)
```

```{r}
geneCov <- function(geneId, kVal, col){
  
  ## Define user input
  goi <- geneId
  kUp <- get(paste0("k", kVal, "deUp"))
  kDown <- get(paste0("k", kVal, "deDown"))
  
  ## Gene region & genome axis tracks
  sym <- mcols(dgeList$genes)[goi,] %>%
    .$gene_name
  counts <- dgeList$counts[goi,]
  gm <- subset(exons_Dr, gene == goi) 
  # %>% subset(transcript %in% rownames(dgeFiltTx))
  gmRed <- gm %>% 
    GenomicRanges::reduce()
  st <- GenomicRanges::start(gmRed)
  en <- GenomicRanges::end(gmRed)
  chr <- seqnames(gmRed)@values
  str <- gmRed %>% 
    strand() %>%
    .@values
  gTrack <- GenomeAxisTrack()
  rrTrack <- GeneRegionTrack(
    gmRed,
    name = "Gene model",
    rotation.title = 0,
    col.title = "black",
    fontface.title = 2,
    fontsize = 14,
    fill = "darkorange"
  )
  rTrack <- GeneRegionTrack(
    gm,
    name = "Tx",
    rotation.title = 0,
    col.title = "black",
    fontface.title = 2,
    fontsize = 14
  )
  
  ## GC content track
  win <- slidingWindows(gmRed, width = 1) %>% 
    unlist()
  start <- win %>%
    start()
  end <- win %>%
    end()
  seq <- getSeq(dna, win)
  gc <- seq %>%
    letterFrequency(letters = "GC", OR = "") %>%
    .[,1]
  tbl <- tibble(
    start = start,
    end = end,
    gc = gc
  ) %>% 
    mutate(
      seqnames = chr,
      strand = str,
      cumsum = cumsum(gc),
      cummean = cummean(gc),
      lag = dplyr::lag(cumsum, n = 50, default = NA),
      lead = dplyr::lead(cumsum, n = 50, default = NA),
      gcsum = lead - lag,
      gcprop = gcsum / 101
    ) 
  gcTrack <- tbl %>% 
    dplyr::select(seqnames, start, end, strand, gcprop) %>%
    makeGRangesFromDataFrame(
      keep.extra.columns = TRUE
    ) %>%
    DataTrack(
      type = c("b"),
      ylim = c(0, 1), 
      cex = 0.2,
      cex.axis = 0.3,
      name = "GC %",
      col.title="black",
      fontface.title=2,
      fontsize=14,
      grid = TRUE,
      lwd.grid = 0.5,
      lty.grid = "dashed",
      v = 0
    )
  
  ## Coverage tracks
  files <- list.files(
    "/hpcfs/users/a1647910/20200310_rRNADepletion/1_Psen2S4Ter/2_alignedData/bam", 
    pattern = ".bam$", 
    full.names = TRUE
  )
  param <- ScanBamParam(
    what = c("pos", "qwidth"),
    which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
    flag=scanBamFlag(isUnmappedQuery=FALSE)
  )
  covFunc <- function(x){
    prop <- order$rRNA[order$filenames == x] %>%
      percent(accuracy = 0.1)
    reads <- counts[order$sample[order$filenames == x]]
    # pileup(x, scanBamParam = param) %>%
    #   dplyr::filter(strand == str) %>%
    #   mutate(start = pos, end = pos, count = count/reads) %>% 
    pileup(x, scanBamParam = param) %>%
      dplyr::filter(strand == str) %>%
      group_by(pos) %>% 
      summarise(count = sum(count)) %>%
      ungroup() %>%
      mutate(
        seqnames = chr, 
        start = pos, 
        end = pos, 
        strand = str,
        norm = count/reads
      ) %>% 
      dplyr::select(seqnames, start, end, strand, norm) %>%
      makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>%
      DataTrack(
        type = c("heatmap"),
        gradient = viridis(100, option = col),
        ncolor = 20,
        cex = 0.2, 
        name = prop,
        rotation.title = 0,
        col.title = "black",
        fontface.title = 2,
        fontsize = 14,
        showAxis = FALSE
      )
  }
  covList <- mclapply(files, covFunc, mc.cores = cores) 
  covOrd <- covList[as.integer(order$order)]
  
  ## k-mer track
  gmSeq <- getSeq(dna, gmRed)
  ## Up k-mers
  merLocUp <- lapply(kUp, function(x){
    vmatchPattern(x, gmSeq) %>%
      mapply(function(y,z){
        start(y) + z
      }, y = ., z = st) %>%
      unlist()
  }) %>%
    unlist()
  kTrack <- DataTrack(
    name = "K-mer",
    rotation.title = 0,
    col.title = "black",
    fontface.title = 2,
    fontsize = 14,
    showAxis = FALSE
  )
  if (length(merLocUp) > 0) {
    kTrack %<>%
      HighlightTrack(
        trackList = ., 
        start = merLocUp,
        width = kVal,
        chromosome = chr,
        inBackground = TRUE,
        col = "seagreen",
        fill = "seagreen",
        showAxis = FALSE
      )
  }
  # ## Down k-mers
  # merLocDown <- lapply(kDown, function(x){
  #   vmatchPattern(x, gmSeq) %>%
  #     mapply(function(y,z){
  #       start(y) + z
  #     }, y = ., z = st) %>%
  #     unlist()
  # }) %>%
  #   unlist()
  # downTrack <- DataTrack(
  #   name = "Down",
  #   rotation.title = 0,
  #   col.title = "black",
  #   fontface.title = 2,
  #   fontsize = 14,
  #   showAxis = FALSE
  # )
  # if (length(merLocDown) > 0) {
  #   downTrack %<>%
  #     HighlightTrack(
  #       trackList = ., 
  #       start = merLocDown,
  #       width = kVal,
  #       chromosome = chr,
  #       inBackground = TRUE,
  #       col = "white",
  #       fill = "red3",
  #       showAxis = FALSE
  #     )
  # }
  
  ## Plot the tracks
  plotTracks(
    c(gTrack, rrTrack, rTrack, covOrd, gcTrack, kTrack),
    from = min(st) - 100,
    to = max(en) + 100,
    chromosome = chr,
    sizes = c(1,0.5,length(unique(mcols(gm)$transcript))*0.25,rep(0.5, length(covOrd)),1, 0.5),
    main = sym,
    cex.main = 1.5,
    fontface.main = 1
  )
  
  ###################
  ## Plot Checking ##
  ###################
  # gmSeq <- getSeq(dna, gmRed)
  # merLoc <- lapply(k8deUp, function(x){
  #   vmatchPattern(x, gmSeq) %>%
  #     mapply(function(y,z){
  #       start(y) + z
  #     }, y = ., z = st) %>%
  #     unlist()
  # }) %>%
  #   unlist()
  # checkTrack <- DataTrack(
  #   name = "Check",
  #   rotation.title = 0,
  #   col.title = "black",
  #   fontface.title = 2,
  #   fontsize = 14,
  #   showAxis = FALSE
  # ) %>%
  #   HighlightTrack(
  #     trackList = .,
  #     start = merLoc,
  #     width = 8,
  #     chromosome = chr,
  #     inBackground = TRUE,
  #     col = "white",
  #     fill = "blue"
  #   )
  # plotTracks(
  #   c(gTrack, rrTrack, rTrack, covOrd, gcTrack, upTrack, checkTrack, downTrack),
  #   from = min(st) - 100,
  #   to = max(en) + 100,
  #   chromosome = chr,
  #   sizes = c(1,0.5,length(unique(mcols(gm)$transcript))*0.25,rep(0.5, length(covOrd)),1, 0.5, 0.5, 0.5), 
  #   main = sym,
  #   cex.main = 1.5,
  #   fontface.main = 1
  # )
}
```

## LM test

```{r}
lmTest <- function(goi){
  
  ## Exon info
  counts <- dgeList$counts[goi,]
  gm <- subset(exons_Dr, gene == goi)
  gmRed <- gm %>% 
    GenomicRanges::reduce()
  st <- GenomicRanges::start(gmRed)
  en <- GenomicRanges::end(gmRed)
  chr <- seqnames(gmRed)@values
  str <- gmRed %>% 
    strand() %>%
    .@values
  gmRange <- mapply(function(x, y){
    seq(x,y,1)
  }, x = st, y = en, SIMPLIFY = FALSE) %>%
    unlist() %>%
    enframe(name = NULL, value = "pos")
  
  ## Kmer info
  gmSeq <- getSeq(dna, gmRed)
  merLoc <- lapply(k8deUp, function(x){
    vmatchPattern(x, gmSeq) %>%
      mapply(function(y,z){
        start(y) + z
      }, y = ., z = st) %>%
      unlist()
  }) %>%
    unlist() %>%
    lapply(
      function(x){
        sapply(x, function(y){
          seq(y, y+7, 1)
        })
      }
    ) %>%
    unlist() %>%
    unique()
  inKmer <- gmRange$pos %in% merLoc
  
  ## Coverage info
  files <- list.files(
    "/hpcfs/users/a1647910/20200310_rRNADepletion/1_Psen2S4Ter/2_alignedData/bam", 
    pattern = ".bam$", 
    full.names = TRUE
  )
  param <- ScanBamParam(
    what = c("pos", "qwidth"),
    which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
    flag = scanBamFlag(isUnmappedQuery=FALSE)
  )
  covFunc <- function(x){
    prop <- order$rRNA[order$filenames == x] %>%
      percent(accuracy = 0.1)
    reads <- counts[order$sample[order$filenames == x]]
    pileup(x, scanBamParam = param) %>%
      dplyr::filter(strand == str) %>%
      group_by(pos) %>% 
      summarise(count = sum(count)) %>%
      ungroup() %>%
      mutate(
        seqnames = chr, 
        start = pos, 
        end = pos, 
        strand = str,
        count = count/reads
      ) %>% 
      dplyr::select(seqnames, start, end, strand, count) %>%
      makeGRangesFromDataFrame(keep.extra.columns = TRUE)
  }
  covList <- mclapply(files, covFunc, mc.cores = cores)
  
  ## lm test
  names <- dgeList$samples %>% 
    dplyr::arrange(filenames) %>%
    .$sample
  test <- lapply(covList, function(x){
    tibble(
      pos = start(x),
      cov = mcols(x)$count
    )
  }) %>%
    purrr::reduce(full_join, by = "pos") %>%
    set_colnames(c("pos", names)) %>%
    right_join(gmRange) %>%
    cbind(inKmer) %>%
    as_tibble() %>%
    pivot_longer(
      cols = contains(c("Heter", "Hom", "WT")), 
      names_to = "sample", 
      values_to = "coverage"
    ) %>%
    left_join(dgeList$samples[,c("sample", "rRNA", "lib.size")], by = "sample")
  test[is.na(test)] <- as.double(0)
  test <- test %>%
    with(
      lm(coverage ~ inKmer * rRNA)
    ) %>%
    summary()
  
  if (sum(inKmer) > 0) {
    return(
      tibble(
        Geneid = goi,
        Estimate = test$coefficients["inKmerTRUE:rRNA","Estimate"],
        Std.error = test$coefficients["inKmerTRUE:rRNA", "Std. Error"],
        t = test$coefficients["inKmerTRUE:rRNA","t value"],
        P = test$coefficients["inKmerTRUE:rRNA","Pr(>|t|)"],
        adj.R.squared = test$adj.r.squared
      )
    )
  } else {
    return(
      tibble(
        Geneid = goi,
        Estimate = NA,
        Std.error = NA,
        t = NA,
        P = NA,
        adj.R.squared = NA
      )
    )
  }
  
}
```

```{r}
checkLm <- function(goi){
  
  ## Exon info
  counts <- dgeList$counts[goi,]
  gm <- subset(exons_Dr, gene == goi)
  gmRed <- gm %>% 
    GenomicRanges::reduce()
  st <- GenomicRanges::start(gmRed)
  en <- GenomicRanges::end(gmRed)
  chr <- seqnames(gmRed)@values
  str <- gmRed %>% 
    strand() %>%
    .@values
  gmRange <- mapply(function(x, y){
    seq(x,y,1)
  }, x = st, y = en, SIMPLIFY = FALSE) %>%
    unlist() %>%
    enframe(name = NULL, value = "pos")
  
  ## Kmer info
  gmSeq <- getSeq(dna, gmRed)
  merLoc <- lapply(k8deUp, function(x){
    vmatchPattern(x, gmSeq) %>%
      mapply(function(y,z){
        start(y) + z
      }, y = ., z = st) %>%
      unlist()
  }) %>%
    unlist() %>%
    lapply(
      function(x){
        sapply(x, function(y){
          seq(y, y+7, 1)
        })
      }
    ) %>%
    unlist() %>%
    unique()
  inKmer <- gmRange$pos %in% merLoc
  
  ## Coverage info
  files <- list.files(
    "/hpcfs/users/a1647910/20200310_rRNADepletion/1_Psen2S4Ter/2_alignedData/bam", 
    pattern = ".bam$", 
    full.names = TRUE
  )
  param <- ScanBamParam(
    what = c("pos", "qwidth"),
    which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
    flag = scanBamFlag(isUnmappedQuery=FALSE)
  )
  covFunc <- function(x){
    prop <- order$rRNA[order$filenames == x] %>%
      percent(accuracy = 0.1)
    reads <- counts[order$sample[order$filenames == x]]
    pileup(x, scanBamParam = param) %>%
      dplyr::filter(strand == str) %>%
      group_by(pos) %>% 
      summarise(count = sum(count)) %>%
      ungroup() %>%
      mutate(
        seqnames = chr, 
        start = pos, 
        end = pos, 
        strand = str,
        count = count/reads
      ) %>% 
      dplyr::select(seqnames, start, end, strand, count) %>%
      makeGRangesFromDataFrame(keep.extra.columns = TRUE)
  }
  covList <- mclapply(files, covFunc, mc.cores = cores)
  
  ## lm test
  names <- dgeList$samples %>% 
    dplyr::arrange(filenames) %>%
    .$sample
  test <- lapply(covList, function(x){
    tibble(
      pos = start(x),
      cov = mcols(x)$count
    )
  }) %>%
    purrr::reduce(full_join, by = "pos") %>%
    set_colnames(c("pos", names)) %>%
    right_join(gmRange) %>%
    cbind(inKmer) %>%
    as_tibble() %>%
    pivot_longer(
      cols = contains(c("Heter", "Hom", "WT")), 
      names_to = "sample", 
      values_to = "coverage"
    ) %>%
    left_join(dgeList$samples[,c("sample", "rRNA", "lib.size")], by = "sample")
  test[is.na(test)] <- as.double(0)
  test %>%
    with(
      lm(coverage ~ inKmer * rRNA)
    ) %>%
    autoplot()
  
}
```

```{r}
lmRes <- mclapply(deUp, lmTest, mc.cores = cores) 
lmRes %>%
  purrr::reduce(rbind) %>%
  dplyr::arrange(P) %>%
  left_join(k8overlapUU, by = "Geneid") %>%
  dplyr::select(Geneid, Symbol, Estimate, Std.error, t, -Length, -Overlap, -Norm, P) %>%
  mutate(
    adjP = p.adjust(P, "fdr"),
    sig = adjP < 0.05
  )
saveRDS(lmRes, here::here("R/output/lmRes.Rds"))
lmRes <- readRDS(here::here("R/output/lmRes.Rds"))
```

## GLM test

```{r covTest}
glmTest <- function(goi){
  
  ## Exon info
  counts <- dgeList$counts[goi,]
  gm <- subset(exons_Dr, gene == goi)
  gmRed <- gm %>% 
    GenomicRanges::reduce()
  st <- GenomicRanges::start(gmRed)
  en <- GenomicRanges::end(gmRed)
  chr <- seqnames(gmRed)@values
  str <- gmRed %>% 
    strand() %>%
    .@values
  gmRange <- mapply(function(x, y){
    seq(x,y,1)
  }, x = st, y = en, SIMPLIFY = FALSE) %>%
    unlist() %>%
    enframe(name = NULL, value = "pos")
  
  ## Kmer info
  gmSeq <- getSeq(dna, gmRed)
  merLoc <- lapply(k8deUp, function(x){
    vmatchPattern(x, gmSeq) %>%
      mapply(function(y,z){
        start(y) + z
      }, y = ., z = st) %>%
      unlist()
  }) %>%
    unlist() %>%
    lapply(
      function(x){
        sapply(x, function(y){
          seq(y, y+7, 1)
        })
      }
    ) %>%
    unlist() %>%
    unique()
  inKmer <- gmRange$pos %in% merLoc
  
  ## Coverage info
  files <- list.files(
    "/hpcfs/users/a1647910/20200310_rRNADepletion/1_Psen2S4Ter/2_alignedData/bam", 
    pattern = ".bam$", 
    full.names = TRUE
  )
  param <- ScanBamParam(
    what = c("pos", "qwidth"),
    which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
    flag = scanBamFlag(isUnmappedQuery=FALSE)
  )
  covFunc <- function(x){
    prop <- order$rRNA[order$filenames == x] %>%
      percent(accuracy = 0.1)
    reads <- counts[order$sample[order$filenames == x]]
    pileup(x, scanBamParam = param) %>%
      dplyr::filter(strand == str) %>%
      group_by(pos) %>% 
      summarise(count = sum(count)) %>%
      ungroup() %>%
      mutate(
        seqnames = chr, 
        start = pos, 
        end = pos, 
        strand = str
      ) %>% 
      dplyr::select(seqnames, start, end, strand, count) %>%
      makeGRangesFromDataFrame(keep.extra.columns = TRUE)
  }
  covList <- mclapply(files, covFunc, mc.cores = cores)
  
  ## test
  reads <- dgeList$counts["ENSDARG00000071566",] %>%
    enframe(name = "sample", value = "reads")
  names <- dgeList$samples %>% 
    dplyr::arrange(filenames) %>%
    .$sample
  test <- lapply(covList, function(x){
    tibble(
      pos = start(x),
      cov = mcols(x)$count
    )
  }) %>%
    purrr::reduce(full_join, by = "pos") %>%
    set_colnames(c("pos", names)) %>%
    right_join(gmRange) %>%
    cbind(inKmer) %>%
    as_tibble() %>%
    pivot_longer(
      cols = contains(c("Heter", "Hom", "WT")), 
      names_to = "sample", 
      values_to = "coverage"
    ) %>%
    left_join(dgeList$samples[,c("sample", "rRNA", "lib.size")], by = "sample") %>%
    left_join(reads, by = "sample")
  test[is.na(test)] <- 0
  
  test <- test %>%
    with(
      glm(coverage ~ inKmer * rRNA + offset(log(reads)), family = poisson(link = "log"))
    ) 
  
  if (sum(inKmer) > 0) {
    return(
      tibble(
        Geneid = goi,
        Estimate = test$coefficients["inKmerTRUE:rRNA","Estimate"],
        Std.error = test$coefficients["inKmerTRUE:rRNA", "Std. Error"],
        z = test$coefficients["inKmerTRUE:rRNA","z value"],
        P = test$coefficients["inKmerTRUE:rRNA","Pr(>|z|)"]
      )
    )
  } else {
    return(
      tibble(
        Geneid = goi,
        Estimate = NA,
        Std.error = NA,
        z = NA,
        P = NA
      )
    )
  }
  
}
```

```{r covTest}
checkGlm <- function(goi){
  
  ## Exon info
  counts <- dgeList$counts[goi,]
  gm <- subset(exons_Dr, gene == goi)
  gmRed <- gm %>% 
    GenomicRanges::reduce()
  st <- GenomicRanges::start(gmRed)
  en <- GenomicRanges::end(gmRed)
  chr <- seqnames(gmRed)@values
  str <- gmRed %>% 
    strand() %>%
    .@values
  gmRange <- mapply(function(x, y){
    seq(x,y,1)
  }, x = st, y = en, SIMPLIFY = FALSE) %>%
    unlist() %>%
    enframe(name = NULL, value = "pos")
  
  ## Kmer info
  gmSeq <- getSeq(dna, gmRed)
  merLoc <- lapply(k8deUp, function(x){
    vmatchPattern(x, gmSeq) %>%
      mapply(function(y,z){
        start(y) + z
      }, y = ., z = st) %>%
      unlist()
  }) %>%
    unlist() %>%
    lapply(
      function(x){
        sapply(x, function(y){
          seq(y, y+7, 1)
        })
      }
    ) %>%
    unlist() %>%
    unique()
  inKmer <- gmRange$pos %in% merLoc
  
  ## Coverage info
  files <- list.files(
    "/hpcfs/users/a1647910/20200310_rRNADepletion/1_Psen2S4Ter/2_alignedData/bam", 
    pattern = ".bam$", 
    full.names = TRUE
  )
  param <- ScanBamParam(
    what = c("pos", "qwidth"),
    which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
    flag = scanBamFlag(isUnmappedQuery=FALSE)
  )
  covFunc <- function(x){
    prop <- order$rRNA[order$filenames == x] %>%
      percent(accuracy = 0.1)
    reads <- counts[order$sample[order$filenames == x]]
    pileup(x, scanBamParam = param) %>%
      dplyr::filter(strand == str) %>%
      group_by(pos) %>% 
      summarise(count = sum(count)) %>%
      ungroup() %>%
      mutate(
        seqnames = chr, 
        start = pos, 
        end = pos, 
        strand = str
      ) %>% 
      dplyr::select(seqnames, start, end, strand, count) %>%
      makeGRangesFromDataFrame(keep.extra.columns = TRUE)
  }
  covList <- mclapply(files, covFunc, mc.cores = cores)
  
  ## lm test
  names <- dgeList$samples %>% 
    dplyr::arrange(filenames) %>%
    .$sample
  test <- lapply(covList, function(x){
    tibble(
      pos = start(x),
      cov = mcols(x)$count
    )
  }) %>%
    purrr::reduce(full_join, by = "pos") %>%
    set_colnames(c("pos", names)) %>%
    right_join(gmRange) %>%
    cbind(inKmer) %>%
    as_tibble() %>%
    pivot_longer(
      cols = contains(c("Heter", "Hom", "WT")), 
      names_to = "sample", 
      values_to = "coverage"
    ) %>%
    left_join(dgeList$samples[,c("sample", "rRNA", "lib.size")], by = "sample")
  test[is.na(test)] <- 0
  par(mfrow = c(2, 2))
  test %>%
    with(
      glm(coverage ~ inKmer * rRNA + offset(log(lib.size)), family = poisson(link = "log"))
    ) %>%
    plot()
  
}
```

```{r}
# glmRes <- mclapply(deUp, glmTest, mc.cores = cores) %>%
#   purrr::reduce(rbind) %>%
#   dplyr::arrange(P) %>%
#   left_join(k8overlapUU, by = "Geneid") %>%
#   dplyr::select(Geneid, Symbol, Estimate, Std.error, z, -Length, -Overlap, -Norm, P) %>%
#   mutate(
#     adjP = p.adjust(P, "fdr"),
#     sig = adjP < 0.05
#   )
# saveRDS(glmRes, here::here("R/output/glmRes.Rds"))
glmRes <- readRDS(here::here("R/output/glmRes.Rds"))
```

```{r}
geneCov(glmRes$Geneid[32], 8)
geneCov(glmRes$Geneid[41], 8)
geneCov(glmRes$Geneid[78], 8)
geneCov(glmRes$Geneid[80], 8)
geneCov(glmRes$Geneid[104], 8)
geneCov(glmRes$Geneid[116], 8)
geneCov(glmRes$Geneid[141], 8, "B")
geneCov(glmRes$Geneid[146], 8, "B")
geneCov(glmRes$Geneid[168], 8, "B")
geneCov(glmRes$Geneid[220], 8, "B")
geneCov(glmRes$Geneid[223], 8, "B")
```

## KW test

```{r}
kwTest <- function(goi){
  
  ## Exon info
  counts <- dgeList$counts[goi,]
  gm <- subset(exons_Dr, gene == goi)
  gmRed <- gm %>% 
    GenomicRanges::reduce()
  st <- GenomicRanges::start(gmRed)
  en <- GenomicRanges::end(gmRed)
  chr <- seqnames(gmRed)@values
  str <- gmRed %>% 
    strand() %>%
    .@values
  gmRange <- mapply(function(x, y){
    seq(x,y,1)
  }, x = st, y = en, SIMPLIFY = FALSE) %>%
    unlist() %>%
    enframe(name = NULL, value = "pos")
  
  ## Kmer info
  gmSeq <- getSeq(dna, gmRed)
  merLoc <- lapply(k8deUp, function(x){
    vmatchPattern(x, gmSeq) %>%
      mapply(function(y,z){
        start(y) + z
      }, y = ., z = st) %>%
      unlist()
  }) %>%
    unlist() %>%
    lapply(
      function(x){
        sapply(x, function(y){
          seq(y, y+7, 1)
        })
      }
    ) %>%
    unlist() %>%
    unique()
  inKmer <- gmRange$pos %in% merLoc
  
  ## Coverage info
  files <- list.files(
    "/hpcfs/users/a1647910/20200310_rRNADepletion/1_Psen2S4Ter/2_alignedData/bam", 
    pattern = ".bam$", 
    full.names = TRUE
  )
  param <- ScanBamParam(
    what = c("pos", "qwidth"),
    which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
    flag = scanBamFlag(isUnmappedQuery=FALSE)
  )
  covFunc <- function(x){
    prop <- order$rRNA[order$filenames == x] %>%
      percent(accuracy = 0.1)
    reads <- counts[order$sample[order$filenames == x]]
    pileup(x, scanBamParam = param) %>%
      dplyr::filter(strand == str) %>%
      group_by(pos) %>% 
      summarise(count = sum(count)) %>%
      ungroup() %>%
      mutate(
        seqnames = chr, 
        start = pos, 
        end = pos, 
        strand = str,
        count = count/reads
      ) %>% 
      dplyr::select(seqnames, start, end, strand, count) %>%
      makeGRangesFromDataFrame(keep.extra.columns = TRUE)
  }
  covList <- mclapply(files, covFunc, mc.cores = cores)
  
  ## lm test
  names <- dgeList$samples %>% 
    dplyr::arrange(filenames) %>%
    .$sample
  test <- lapply(covList, function(x){
    tibble(
      pos = start(x),
      cov = mcols(x)$count
    )
  }) %>%
    purrr::reduce(full_join, by = "pos") %>%
    set_colnames(c("pos", names)) %>%
    right_join(gmRange) %>%
    cbind(inKmer) %>%
    as_tibble() %>%
    pivot_longer(
      cols = contains(c("Heter", "Hom", "WT")), 
      names_to = "sample", 
      values_to = "coverage"
    ) %>%
    left_join(dgeList$samples[,c("sample", "rRNA", "lib.size")], by = "sample")
  test[is.na(test)] <- as.double(0)
  kmerT <- test %>%
    dplyr::filter(inKmer) %>%
    with(
      kruskal.test(coverage ~ rRNA)
    )
  kmerF <- test %>%
    dplyr::filter(!inKmer) %>%
    with(
      kruskal.test(coverage ~ rRNA)
    )
  
  return(
    tibble(
      Geneid = goi,
      kmerT.p = kmerT$p.value,
      kmerF.p = kmerF$p.value
    )
  )
  
  # if (sum(inKmer) > 0) {
  #   return(
  #     tibble(
  #       Geneid = goi,
  #       kmerT.p = kmerT$p.value,
  #       kmerF.p = kmerF$p.value
  #     )
  #   )
  # } else {
  #   return(
  #     tibble(
  #       Geneid = goi,
  #       kmerT.p = NA,
  #       kmerF.p = NA
  #     )
  #   )
  # }
  
}
```

```{r}
kwRes <- mclapply(deUp, kwTest, mc.cores = 6) %>%
  purrr::reduce(rbind) %>%
  dplyr::arrange(P) %>%
  left_join(k8overlapUU, by = "Geneid") %>%
  dplyr::select(Geneid, Symbol, -Length, -Overlap, -Norm, P) %>%
  mutate(
    adjP = p.adjust(P, "fdr"),
    sig = adjP < 0.05
  )
saveRDS(kwRes, here::here("R/output/kwRes.Rds"))
kwRes <- readRDS(here::here("R/output/kwRes.Rds"))
```

## ttmep

```{r}
## Define user input
kVal <- 8
goi <- glmRes$Geneid[168]
kUp <- k8deUp

## Gene region & genome axis tracks
sym <- mcols(dgeList$genes)[goi,] %>%
  .$gene_name
counts <- dgeList$counts[goi,]
gm <- subset(exons_Dr, gene == goi)
gmRed <- gm %>% 
  GenomicRanges::reduce()
st <- GenomicRanges::start(gmRed)
en <- GenomicRanges::end(gmRed)
chr <- seqnames(gmRed)@values
str <- gmRed %>% 
  strand() %>%
  .@values
gTrack <- GenomeAxisTrack()
rrTrack <- GeneRegionTrack(
  gmRed,
  name = "Gene model",
  rotation.title = 0,
  col.title = "black",
  fontface.title = 2,
  fontsize = 14,
  fill = "darkorange"
)
rTrack <- GeneRegionTrack(
  gm,
  name = "Gene model",
  rotation.title = 0,
  col.title = "black",
  fontface.title = 2,
  fontsize = 14,
  fill = "darkorange"
)

## GC content track
win <- slidingWindows(gmRed, width = 1) %>% 
  unlist()
start <- win %>%
  start()
end <- win %>%
  end()
seq <- getSeq(dna, win)
gc <- seq %>%
  letterFrequency(letters = "GC", OR = "") %>%
  .[,1]
tbl <- tibble(
  start = start,
  end = end,
  gc = gc
) %>% 
  mutate(
    seqnames = chr,
    strand = str,
    cumsum = cumsum(gc),
    cummean = cummean(gc),
    lag = dplyr::lag(cumsum, n = 50, default = NA),
    lead = dplyr::lead(cumsum, n = 50, default = NA),
    gcsum = lead - lag,
    gcprop = gcsum / 101
  ) 
gcTrack <- tbl %>% 
  dplyr::select(seqnames, start, end, strand, gcprop) %>%
  makeGRangesFromDataFrame(
    keep.extra.columns = TRUE
  ) %>%
  DataTrack(
    type = c("b"),
    ylim = c(0, 1), 
    cex = 0.2,
    cex.axis = 0.3,
    name = "GC %",
    col.title="black",
    fontface.title=2,
    fontsize=14,
    grid = TRUE,
    lwd.grid = 0.5,
    lty.grid = "dashed",
    v = 0
  )

## Coverage tracks
files <- list.files(
  "/hpcfs/users/a1647910/20200310_rRNADepletion/1_Psen2S4Ter/2_alignedData/bam", 
  pattern = ".bam$", 
  full.names = TRUE
)
param <- ScanBamParam(
  what = c("pos", "qwidth"),
  which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
  flag=scanBamFlag(isUnmappedQuery=FALSE)
)
covFunc <- function(x){
  prop <- order$rRNA[order$filenames == x] %>%
    percent(accuracy = 0.1)
  reads <- counts[order$sample[order$filenames == x]]
  pileup(x, scanBamParam = param) %>%
    dplyr::filter(strand == str) %>%
    group_by(pos) %>% 
    summarise(count = sum(count)) %>%
    ungroup() %>%
    mutate(
      seqnames = chr, 
      start = pos, 
      end = pos, 
      strand = str,
      norm = count/reads
    ) %>% 
    dplyr::select(seqnames, start, end, strand, norm) %>%
    makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>%
    DataTrack(
      type = c("heatmap"),
      gradient = viridis(100, option = "B"),
      ncolor = 20,
      cex = 0.2, 
      name = prop,
      rotation.title = 0,
      col.title = "black",
      fontface.title = 2,
      fontsize = 14,
      showAxis = FALSE
    )
}
covList <- mclapply(files, covFunc, mc.cores = cores) 
covOrd <- covList[as.integer(order$order)]

## k-mer track
gmSeq <- getSeq(dna, gmRed)
## Up k-mers
merLocUp <- lapply(kUp, function(x){
  vmatchPattern(x, gmSeq) %>%
    mapply(function(y,z){
      start(y) + z
    }, y = ., z = st) %>%
    unlist()
}) %>%
  unlist()
kTrack <- DataTrack(
  name = "K-mer",
  rotation.title = 0,
  col.title = "black",
  fontface.title = 2,
  fontsize = 14,
  showAxis = FALSE
)
if (length(merLocUp) > 0) {
  kTrack %<>%
    HighlightTrack(
      trackList = ., 
      start = merLocUp,
      width = kVal,
      chromosome = chr,
      inBackground = TRUE,
      col = "seagreen",
      fill = "seagreen",
      showAxis = FALSE
    )
}

## Plot the tracks
plotTracks(
  c(gTrack, rTrack, covOrd, kTrack),
  from = min(st) - 100,
  to = max(en) + 100,
  chromosome = chr,
  sizes = c(1,1,rep(0.5, length(covOrd)), 0.5),
  main = sym,
  cex.main = 1.5,
  fontface.main = 1,
  background.title = "#c6cbd7"
)
```

```{r}
data <- grep("_R1", fqName(rawFqc))
lineCols <- RColorBrewer::brewer.pal(min(12, n), 
                "Paired")
            lineCols <- colorRampPalette(lineCols)(n)
            lineCols <- c("#000000", lineCols)
  plotGcContent(
    x = rawFqc[data], 
    plotType = "line",
    gcType = "Transcriptome",
    species = "Drerio"
  ) +
    labs(title = "GC Distribution", subtitle = c()) +
    scale_color_manual(values = lineCols,
                labels = c("Theoretical Distribution", "Sample 1", "Sample 2", "Sample 3", "Sample 4", "Sample 5", "Sample 6",
                                  "Sample 7", "Sample 8", "Sample 9", "Sample 10", "Sample 11", "Sample 12")) +
    theme(panel.background = element_rect(fill = "#c6cbd7"),
          plot.background = element_rect(fill = "#c6cbd7", colour = "#c6cbd7"),
          legend.background = element_rect(fill = "#c6cbd7"),
          legend.key = element_rect(fill = "#c6cbd7"))
```