---
title: "rRNADepletion_s4ter"
author: "Lachlan Baer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    fig_width: 8
    fig_height: 6
    fig_align: "center"
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
    echo = TRUE,
    message = FALSE,
    warning = FALSE,
    error = FALSE, 
    fig.align = "center"
)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

# Setup 

```{r packages}
library(tidyverse)
library(magrittr)
library(here)
library(scales)
library(AnnotationHub)
library(Biostrings)
library(Rsamtools)
```

```{r options}
if (interactive()) setwd(here::here())
theme_set(theme_bw())
```

```{r ah}
ah <- AnnotationHub() %>%
    subset(species == "Danio rerio") %>%
    subset(rdataclass == "EnsDb")
ensDb <- ah[["AH83189"]]
```

# Load data

```{r}
dgeFilt <- readRDS(here("1_Psen2S4Ter/Rdata/dgeFilt.Rds"))
```

# Exploration

Now that differential expression of $k$-mers has been detected, the next step is to develop a normalisation method that can lessen the extent of gene-level differential expression bias due to presence of particular $k$-mers. The initial aim of this exploration is to define a single value per gene, $\kappa$, that represents the magnitude of bias due to the genes constituent $k$-mers.

## Using logFC

In an initial attempt to develop a value for $\kappa$, logFC values from previously derived $k$-mer differential expression testing will be evaluated. Firstly, for each gene's coding sequence, a count for the presence of all detected $k$-mers will be calculated. This count, $n$, will then be multiplied by the $k$-mers logFC value to determine $\kappa$: \[\kappa = n_1LFC_1 + n_2LFC_2 + n_3LFC_3\]

```{r ids}
ids <- rownames(dgeFilt)
```

```{r}
## Reduced exon sequences were saved to speed up document
ranges <- exonsBy(ensDb, by = "gene", filter = GeneIdFilter(ids)) %>%
    lapply(function(x){
        GenomicRanges::reduce(x)
    }) %>%
    GRangesList()
```

```{r}
seqs <- lapply(ranges, function(x){
    getSeq(
        FaFile(
            file = "/mnt/hpcfs/a1647910/workflows/smk-ens-refs/results/danio_rerio/release_101/genome.fa",
            index = "/mnt/hpcfs/a1647910/workflows/smk-ens-refs/results/danio_rerio/release_101/genome.fa.fai"
        ),
        x
    ) %>%
        unlist()
}) %>%
    DNAStringSet()
```

```{r}
lapply(names(seqs), \(x){
    writeXStringSet(
        seqs[x],
        here(paste0(
            "1_Psen2S4Ter/analysis-gene_kmer/results/seqs/", x, ".fa"
        ))
    )
})
```
