---
title: "rRNA Depletion (transcript-level)"
author: "Lachlan Baer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    fig_width: 8
    fig_height: 6
    fig_align: "center"
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  error = FALSE, 
  fig.align = "center"
)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

# Setup 

```{r packages}
library(tidyverse)
library(ngsReports)
library(here)
library(scales)
library(ggpubr)
library(kableExtra)
library(AnnotationHub)
library(ensembldb)
library(edgeR)
library(corrplot)
library(cqn)
library(DT)
```

```{r options}
if (interactive()) setwd(here::here())
theme_set(theme_bw())
```

```{r samples}
files <- list.files(
  path = "0_rawData/FastQC",
  pattern = "zip",
  full.names = TRUE
) %>%
  basename()
samples <- tibble(
  sample = str_remove(files, "_fastqc.zip"),
  dataset = NA,
  organism = NA
) %>%
  mutate(
    dataset = ifelse(
      str_detect(sample, "SRR213"), "E-GEOD-71609", dataset
    ),
    dataset = ifelse(
      str_detect(sample, "SRR218"), "E-GEOD-72322", dataset
    ),
    organism = ifelse(
      str_detect(sample, "(SRR213|SRR218)"), "zebrafish", organism
    ),
    dataset = ifelse(
      str_detect(sample, "ERR313"), "E-MTAB-7636", dataset
    ),
    dataset = ifelse(
      str_detect(sample, "ERR268"), "E-MTAB-6972", dataset
    ),
    organism = ifelse(
      str_detect(sample, "(ERR313|ERR268)"), "mouse", organism
    )
  )
datasets <- samples$dataset %>% 
  unique()
```

```{r ah_Dr, cache=TRUE}
ah_Dr <- AnnotationHub() %>%
  subset(species == "Danio rerio") %>%
  subset(rdataclass == "EnsDb")
ensDb_Dr <- ah_Dr[["AH74989"]]
trEns_Dr <- transcripts(ensDb_Dr) %>%
  mcols() %>% 
  as_tibble()
trLen_Dr <- exonsBy(ensDb_Dr, "tx") %>%
  width() %>%
  vapply(sum, integer(1))
geneGcLen_Dr <- trLen_Dr %>%
  enframe() %>%
  set_colnames(c("tx_id", "length")) %>%
  left_join(trEns_Dr) %>%
  group_by(gene_id) %>% 
  summarise(
    aveLen = mean(length),
    maxLen = max(length), 
    aveGc = sum(length * gc_content) / sum(length),
    longestGc = gc_content[which.max(length)[[1]]]
  ) %>%
  mutate(
    aveGc =  aveGc / 100,
    longestGc = longestGc / 100
  )
trGcLen_Dr <- trLen_Dr %>%
  enframe() %>%
  set_colnames(c("tx_id", "length")) %>%
  left_join(trEns_Dr) %>%
  group_by(tx_id) %>% 
  summarise(
    aveLen = mean(length),
    maxLen = max(length), 
    aveGc = sum(length * gc_content) / sum(length),
    longestGc = gc_content[which.max(length)[[1]]]
  ) %>%
  mutate(
    aveGc =  aveGc / 100,
    longestGc = longestGc / 100
  )
genesGR_Dr <- genes(ensDb_Dr)
mcols(genesGR_Dr) <- mcols(genesGR_Dr)[c("gene_id", "gene_name", 
                                         "gene_biotype", "entrezid")]
txGR_Dr <- transcripts(ensDb_Dr)
mcols(txGR_Dr) <- mcols(txGR_Dr)[c("tx_id", "tx_name", 
                                   "tx_biotype", "tx_id_version", "gene_id")]
```

```{r ah_Mm, cache=TRUE}
ah_Mm <- AnnotationHub() %>%
  subset(species == "Mus musculus") %>%
  subset(rdataclass == "EnsDb")
ensDb_Mm <- ah_Mm[["AH75036"]]
trEns_Mm <- transcripts(ensDb_Mm) %>%
  mcols() %>% 
  as_tibble()
trLen_Mm <- exonsBy(ensDb_Mm, "tx") %>%
  width() %>%
  vapply(sum, integer(1))
geneGcLen_Mm <- trLen_Mm %>%
  enframe() %>%
  set_colnames(c("tx_id", "length")) %>%
  left_join(trEns_Mm) %>%
  group_by(gene_id) %>% 
  summarise(
    aveLen = mean(length),
    maxLen = max(length), 
    aveGc = sum(length * gc_content) / sum(length),
    longestGc = gc_content[which.max(length)[[1]]]
  ) %>%
  mutate(
    aveGc =  aveGc / 100,
    longestGc = longestGc / 100
  )
trGcLen_Mm <- trLen_Mm %>%
  enframe() %>%
  set_colnames(c("tx_id", "length")) %>%
  left_join(trEns_Mm) %>%
  group_by(tx_id) %>% 
  summarise(
    aveLen = mean(length),
    maxLen = max(length), 
    aveGc = sum(length * gc_content) / sum(length),
    longestGc = gc_content[which.max(length)[[1]]]
  ) %>%
  mutate(
    aveGc =  aveGc / 100,
    longestGc = longestGc / 100
  )
genesGR_Mm <- genes(ensDb_Mm)
mcols(genesGR_Mm) <- mcols(genesGR_Mm)[c("gene_id", "gene_name", 
                                         "gene_biotype", "entrezid")]
```

```{r ah_Hs, cache=TRUE}
ah_Hs <- AnnotationHub() %>%
  subset(species == "Homo sapiens") %>%
  subset(rdataclass == "EnsDb")
ensDb_Hs <- ah_Hs[["AH75011"]]
trEns_Hs <- transcripts(ensDb_Hs) %>%
  mcols() %>% 
  as_tibble()
trLen_Hs <- exonsBy(ensDb_Hs, "tx") %>%
  width() %>%
  vapply(sum, integer(1))
geneGcLen_Hs <- trLen_Hs %>%
  enframe() %>%
  set_colnames(c("tx_id", "length")) %>%
  left_join(trEns_Hs) %>%
  group_by(gene_id) %>% 
  summarise(
    aveLen = mean(length),
    maxLen = max(length), 
    aveGc = sum(length * gc_content) / sum(length),
    longestGc = gc_content[which.max(length)[[1]]]
  ) %>%
  mutate(
    aveGc =  aveGc / 100,
    longestGc = longestGc / 100
  )
trGcLen_Hs <- trLen_Hs %>%
  enframe() %>%
  set_colnames(c("tx_id", "length")) %>%
  left_join(trEns_Hs) %>%
  group_by(tx_id) %>% 
  summarise(
    aveLen = mean(length),
    maxLen = max(length), 
    aveGc = sum(length * gc_content) / sum(length),
    longestGc = gc_content[which.max(length)[[1]]]
  ) %>%
  mutate(
    aveGc =  aveGc / 100,
    longestGc = longestGc / 100
  )
genesGR_Hs <- genes(ensDb_Hs)
mcols(genesGR_Hs) <- mcols(genesGR_Hs)[c("gene_id", "gene_name", 
                                         "gene_biotype", "entrezid")]
```

An `EnsDb` object was obtained for Ensembl release 98 using the `AnnotationHub` package. This provided the GC content and length for every gene and transcript in the release. 

- For *zebrafish*, this consisted of `r nrow(geneGcLen_Dr)` genes and `r nrow(trGcLen_Dr)` transcripts.
- For *mouse*, this consisted of `r nrow(geneGcLen_Mm)` genes and `r nrow(trGcLen_Mm)` transcripts.
- For *human*, this consisted of `r nrow(geneGcLen_Hs)` genes and `r nrow(trGcLen_Hs)` transcripts.

# Raw data

Raw data was sourced from publically available datasets on ArrayExpress. Samples from zebrafish (*Danio rerio*) were chosen for an initial inspection as it is expected that rRNA sequences are more divergent from commonly chosen model organisms such as mice. Hence, inefficient ribosomal RNA (rRNA) depletion by kits that are generally optimised for these common model organisms may produce a stronger signal in zebrafish. 

Following an initial inspection, further analysis on human and mouse datasets was performed.

The following analysis involves `r nrow(samples)` samples across `r length(datasets)` datasets: `r paste(datasets, sep = ", ")`.

## Library sizes

```{r rawFqcList}
rawFqc <- list.files(
  path = "0_rawData/FastQC",
  pattern = "zip",
  full.names = TRUE
) %>%
  FastqcDataList()
```

```{r rawLibPlots}
data1 <- grepl("SRR213", fqName(rawFqc))
rawLib1 <- plotReadTotals(rawFqc[data1]) +
  labs(subtitle = "E-GEOD-71609")
data2 <- grepl("SRR218", fqName(rawFqc))
rawLib2 <- plotReadTotals(rawFqc[data2]) +
  labs(subtitle = "E-GEOD-72322")
data3 <- grepl("ERR313", fqName(rawFqc))
rawLib3 <- plotReadTotals(rawFqc[data3]) +
  labs(subtitle = "E-MTAB-7636")
data4 <- grepl("ERR268", fqName(rawFqc))
rawLib4 <- plotReadTotals(rawFqc[data4]) +
  labs(subtitle = "E-MTAB-6972")
```

The library sizes of unprocessed datasets ranged between `r comma(min(readTotals(rawFqc)$Total_Sequences))` and `r comma(max(readTotals(rawFqc)$Total_Sequences))` reads.

```{r plotRawLibs, fig.height=12}
ggarrange(rawLib1, rawLib2, rawLib3, rawLib4, ncol = 1, nrow = 4)
```

## GC content

rRNA transcripts are known to have high GC content. Therefore, surveying the GC content of the raw reads serves as a good starting point for detecting incomplete rRNA removal. A spike in GC content at > 70% is expected if this is the case.

```{r gcPlots}
gc1 <- plotGcContent(
  x = rawFqc[data1], 
  plotType = "line",
  gcType = "Transcriptome",
  species = "Drerio"
) +
  labs(subtitle = "E-GEOD-71609")
gc2 <- plotGcContent(
  x = rawFqc[data2], 
  plotType = "line",
  gcType = "Transcriptome",
  species = "Drerio"
) +
  labs(subtitle = "E-GEOD-72322")
gc3 <- plotGcContent(
  x = rawFqc[data3], 
  plotType = "line",
  gcType = "Transcriptome",
  species = "Mmusculus"
) +
  labs(subtitle = "E-MTAB-7636")
gc4 <- plotGcContent(
  x = rawFqc[data4], 
  plotType = "line",
  gcType = "Transcriptome",
  species = "Mmusculus"
) +
  labs(subtitle = "E-MTAB-6972")
```

```{r plotGc, fig.height=12, fig.cap="*GC content of reads in each dataset. No obvious spikes above 70% GC are observed, however unusual distibution shapes in some samples are worth further exploration.*"}
ggarrange(gc1, gc2, gc3, gc4, ncol = 1, nrow = 4)
```

```{r 70GC, fig.cap="*Percentages of each library which contain >70% GC. Using the known theoretical distribution, this should be 0.19% of the total library.*"}
gcMod <- getModule(rawFqc, "Per_sequence_GC") %>%
  mutate(sample = str_remove(Filename, ".fastq.gz"))
rawGC <- gcMod %>% 
  group_by(sample) %>%
  mutate(Freq = Count / sum(Count)) %>%
  dplyr::filter(GC_Content > 70) %>% 
  summarise(Freq = sum(Freq)) %>% 
  arrange(desc(Freq)) %>%
  left_join(samples) 
rawGC %>%
  ggplot(aes(sample, Freq)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(~dataset, scales = "free_x") +
  scale_y_continuous(labels = percent) +
  labs(x = "Sample", y = "Percent of Total") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
```

```{r gcDev, fig.cap="*Standard deviations of observed GC frequency Vs Expected GC frequency using the theoretical GC.*"}
gcDev <- gcMod %>%
  group_by(sample) %>%
  mutate(Freq = Count / sum(Count)) %>%
  left_join(
    getGC(gcTheoretical, "Drerio", "Trans")
  ) %>%
  dplyr::rename(actual = Drerio) %>%
  mutate(res = Freq - actual) %>%
  summarise(ss = sum(res^2), n = n()) %>%
  ungroup() %>%
  mutate(sd = sqrt(ss / (n - 1)))
gcDev %>%
  left_join(samples) %>%
  ggplot(aes(sample, sd)) +
  geom_bar(stat= "identity", position ="dodge") +
  facet_wrap(~dataset, scales = "free_x") +
  scale_fill_manual() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
```

```{r rRnaContDetect, include=FALSE, eval=FALSE, fig.cap="*Comparison of the two measures used for assessing rRNA contamination.*"}
gcDev %>%
  left_join(rawGC) %>%
  ggplot(aes(sd, Freq)) +
  geom_point() +
  geom_smooth(method = "lm") +
  labs(
    x = "SD (GC - Theoretical)",
    y = "%Library > GC70"
  )
```

## Overrepresented sequences

The top 30 overrepresented sequences were analysed using `blastn` and  were found to be predominantly rRNA sequences.

```{r overrep}
getModule(rawFqc, "Overrep") %>% 
  group_by(Sequence, Possible_Source) %>% 
  summarise(`Found In` = n(), `Highest Percentage` = max(Percentage)) %>% 
  arrange(desc(`Highest Percentage`), desc(`Found In`)) %>% 
  ungroup() %>% 
  dplyr::slice(1:30) %>%
  mutate(`Highest Percentage` = percent_format(0.01)(`Highest Percentage`/100)) %>%
  kable(
    align = "llrr", 
    caption = paste(
      "Top", nrow(.),"Overrepresented sequences.",
      "The number of samples they were found in is shown,",
      "along with the percentage of the most 'contaminated' sample."
    )
  ) %>%
  kable_styling(
    bootstrap_options = c("striped", "hover", "condensed", "responsive")
  )
```

## Trimming

```{r trimFqc}
trimFqc <- list.files(
  path = "1_trimmedData/FastQC", 
  pattern = "zip", 
  full.names = TRUE
) %>%
  FastqcDataList()
trimStats <- readTotals(rawFqc) %>%
  dplyr::rename(Raw = Total_Sequences) %>%
  left_join(readTotals(trimFqc), by = "Filename") %>%
  dplyr::rename(Trimmed = Total_Sequences) %>%
  mutate(
    Discarded = 1 - Trimmed/Raw,
    Retained = Trimmed / Raw
  )
```

After trimming of adapters between `r percent(min(trimStats$Discarded), 0.01)` and `r percent(max(trimStats$Discarded), 0.01)` of reads were discarded.

# Aligned data

Trimmed reads were firstly aligned to rRNA sequences using the `BWA-MEM` algorithm to calculate the proportion of reads that were of rRNA origin within each sample. `BWA-MEM` is recommended for high-quality queries of reads ranging from 70bp to 1Mbp as it is faster and more accurate that alternative algorithms `BWA-backtrack` and `BWA-SW`.

## rRNA proportions

```{r, include=FALSE, eval=FALSE}
library(biomaRt)
mart <- useMart("ENSEMBL_MART_ENSEMBL", "drerio_gene_ensembl")
rRnaBM <- getBM(values="rRNA", 
                filters="biotype", 
                attributes=c("ensembl_gene_id", "external_gene_name", "gene_biotype", "description"), 
                mart = mart) %>%
  as_tibble()
```

```{r rRnaProp}
prop1 <- samples %>%
  dplyr::filter(dataset == "E-GEOD-71609") %>%
  # cbind(tibble(proportion = c(0.0344, 0.0339, 0.0409, 0.0347, 0.0362, 0.0397, 0.0288, 0.0346, 0.0429, 0.0297))) 
  cbind(tibble(proportion = c(0.0363, 0.0367, 0.0451, 0.0367, 0.0388, 0.0427, 0.0307, 0.0369, 0.0468, 0.0312)))
prop2 <- samples %>%
  dplyr::filter(dataset == "E-GEOD-72322") %>%
  # cbind(tibble(proportion = c(0.1465, 0.1048, 0.1304, 0.1498, 0.1136, 0.1274, 0.1206, 0.1318, 0.1351, 0.1385, 0.1456, 0.1353, 0.1634, 0.1480, 0.1492, 0.1361, 0.1683, 0.1709, 0.1639, 0.1654, 0.1940, 0.1782, 0.1837, 0.1666)))
  cbind(tibble(proportion = c(0.1673, 0.1060, 0.1410, 0.1758, 0.1154, 0.1292, 0.1213, 0.1361, 0.1368, 0.1449, 0.1510, 0.1370, 0.1745, 0.1580, 0.1562, 0.1441, 0.1743, 0.1749, 0.1715, 0.1713, 0.2110, 0.1887, 0.1895, 0.1739)))
prop3 <- samples %>%
  dplyr::filter(dataset == "E-MTAB-7636") %>%
  cbind(tibble(proportion = c(0.0614, 0.0517, 0.0458, 0.0633, 0.0547, 0.0564, 0.0525, 0.0520, 0.0551, 0.0513, 0.0473, 0.0448, 0.0407, 0.0238, 0.0499, 0.0307, 0.0383, 0.0178, 0.0232, 0.0384, 0.0461, 0.0545, 0.0389, 0.0353, 0.0866, 0.0763)))
prop4 <- samples %>%
  dplyr::filter(dataset == "E-MTAB-6972") %>%
  cbind(tibble(proportion = c(0.0123, 0.0122, 0.0121, 0.0121, 0.0121, 0.0131, 0.0131, 0.0131, 0.0130, 0.0131, 0.0929, 0.0924, 0.0931, 0.0929, 0.0927, 0.0224, 0.0223, 0.0222, 0.0223, 0.0222)))
rRnaProp <- rbind(prop1, prop2, prop3, prop4)
```

```{r plotRRnaProp, fig.cap="*Percentages of each library that align to rRNA sequences with `bwa mem`.*"}
rRnaProp %>%
  ggplot(aes(sample, proportion)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(~dataset, scales = "free_x") +
  scale_y_continuous(labels = percent) +
  labs(x = "Sample", y = "Percent of Total") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
```

Following alignment to rRNA sequences, reads that mapped to rRNA sequences were removed and a separate `fastq` file was created.

These sequences were aligned to the *Danio rerio* GRCz11 genome (Ensembl release 98) using `STAR 2.7.0d` and summarised with `featureCounts`.

# Transcript-level analysis

## Gene GC content and length

```{r salmon}
paths <- list.dirs("4_alignedDataSalmon/quant", recursive = FALSE)
salmon <- catchSalmon(paths = paths)
```

```{r counts_tx}
dgeList_1 <- salmon$counts %>%
  set_colnames(basename(colnames(.))) %>%
  as.data.frame() %>%
  dplyr::select(str_subset(colnames(.), "SRR213")) %>%
  rownames_to_column("tx_id") %>%
  mutate(tx_id = str_remove(tx_id, "\\.[0-9]+$")) %>%
  column_to_rownames("tx_id") %>%
  DGEList() %>%
  calcNormFactors()
dgeList_2 <- salmon$counts %>%
  set_colnames(basename(colnames(.))) %>%
  as.data.frame() %>%
  dplyr::select(str_subset(colnames(.), "SRR218")) %>%
  rownames_to_column("tx_id") %>%
  mutate(tx_id = str_remove(tx_id, "\\.[0-9]+$")) %>%
  column_to_rownames("tx_id") %>%
  DGEList() %>%
  calcNormFactors()
```

```{r addInfo_tx}
dgeList_1$tx <- txGR_Dr[rownames(dgeList_1),]
mcols(dgeList_1$tx) %<>% 
  as.data.frame() %>% 
  left_join(trGcLen_Dr) 
dgeList_1$samples %<>%
  rownames_to_column("rowname") %>%
  mutate(sample = rowname) %>%
  left_join(rRnaProp) %>%
  column_to_rownames("rowname")
dgeList_2$tx <- txGR_Dr[rownames(dgeList_2),]
mcols(dgeList_2$tx) %<>% 
  as.data.frame() %>% 
  left_join(trGcLen_Dr)
dgeList_2$samples %<>%
  rownames_to_column("rowname") %>%
  mutate(sample = rowname) %>%
  left_join(rRnaProp) %>%
  column_to_rownames("rowname")
```

```{r gcFunc_tx}
gcInfo <- function(x) {
  x$counts %>%
    as.data.frame() %>%
    rownames_to_column("tx_id") %>%
    as_tibble() %>%
    pivot_longer(
      cols = contains("SRR"), 
      names_to = "sample", 
      values_to = "counts"
    ) %>%
    dplyr::filter(
      counts > 0
    ) %>%
    left_join(
      trGcLen_Dr
    ) %>%
    dplyr::select(
      ends_with("id"), sample, counts, aveGc, maxLen
    ) %>%
    split(f = .$sample) %>%
    lapply(
      function(x){
        DataFrame(
          gc = Rle(x$aveGc, x$counts),
          logLen = Rle(log10(x$maxLen), x$counts)
        )
      }
    ) 
}
gcSummary <- function(x) {
  x %>%
    vapply(function(x){
      c(mean(x$gc), sd(x$gc), mean(x$logLen), sd(x$logLen))
    }, numeric(4)
    ) %>%
    t() %>%
    set_colnames(
      c("mn_gc", "sd_gc", "mn_logLen", "sd_logLen")
    ) %>%
    as.data.frame() %>%
    rownames_to_column("sample") %>%
    as_tibble()
}
gcCors <-function(x) {
  x %>%
    left_join(rawGC) %>%
    left_join(gcDev) %>%
    dplyr::select(Freq, mn_gc, mn_logLen, sd) %>%
    cor()
}
```

```{r rle_tx}
rle_1 <- gcInfo(dgeList_1)
rle_2 <- gcInfo(dgeList_2)
```

```{r gcSummary_tx}
sum_1 <- gcSummary(rle_1)
sum_2 <- gcSummary(rle_2)
```

```{r gcCors_tx}
cor_1 <- gcCors(sum_1)
cor_2 <- gcCors(sum_2)
```

```{r gc_1_tx, fig.height=4, fig.cap="*Comparison of residual bias potentially introduced by incomplete rRNA removal. Regression lines are shown along with standard error bands for each comparison.*"}
a_1 <- sum_1 %>%
  left_join(rRnaProp) %>%
  ggplot(aes(proportion, mn_logLen)) +
  geom_point(size = 3) +
  geom_smooth(method = "lm") +
  scale_x_continuous(labels = percent) +
  labs(
    x = "rRNA Proportion of Initial Library",
    y = "Mean log(Length)"
  ) 
b_1 <- sum_1 %>%
  left_join(rRnaProp) %>%
  ggplot(aes(proportion, mn_gc)) +
  geom_point(size = 3) +
  geom_smooth(method = "lm") +
  scale_y_continuous(labels = percent) +
  scale_x_continuous(labels = percent) +
  labs(
    x = "rRNA Proportion of Initial Library",
    y = "Mean GC Content"
  ) 
ggarrange(a_1, b_1, ncol = 2, nrow = 1) %>%
  annotate_figure("Dataset 1, E-GEOD-71609")
```

```{r gc_2_tx, fig.height=4, fig.cap="*Comparison of residual bias potentially introduced by incomplete rRNA removal. Regression lines are shown along with standard error bands for each comparison.*"}
a_2 <- sum_2 %>%
  left_join(rRnaProp) %>%
  ggplot(aes(proportion, mn_logLen)) +
  geom_point(size = 3) +
  geom_smooth(method = "lm") +
  scale_x_continuous(labels = percent) +
  labs(
    x = "rRNA Proportion of Initial Library",
    y = "Mean log(Length)"
  ) 
b_2 <- sum_2 %>%
  left_join(rRnaProp) %>%
  ggplot(aes(proportion, mn_gc)) +
  geom_point(size = 3) +
  geom_smooth(method = "lm") +
  scale_y_continuous(labels = percent) +
  scale_x_continuous(labels = percent) +
  labs(
    x = "rRNA Proportion of Initial Library",
    y = "Mean GC Content"
  )
ggarrange(a_2, b_2, ncol = 2, nrow = 1) %>%
  annotate_figure("Dataset 2, E-GEOD-72322")
```

## PCA 

```{r dgeFilt_tx}
genes2keep_1 <- dgeList_1 %>%
  cpm() %>%
  is_greater_than(1) %>%
  rowSums() %>%
  is_weakly_greater_than(5)
dgeFilt_1 <- dgeList_1[genes2keep_1,, keep.lib.sizes = FALSE] %>%
  calcNormFactors()
genes2keep_2 <- dgeList_2 %>%
  cpm() %>%
  is_greater_than(1) %>%
  rowSums() %>%
  is_weakly_greater_than(12)
dgeFilt_2 <- dgeList_2[genes2keep_2,, keep.lib.sizes = FALSE] %>%
  calcNormFactors()
```

```{r pca_tx}
pca_1 <- cpm(dgeFilt_1, log = TRUE) %>%
  t() %>%
  prcomp()
pca_2 <- cpm(dgeFilt_2, log = TRUE) %>%
  t() %>%
  prcomp()
```

```{r pcaCor_tx}
pcaCor_1 <- pca_1$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sum_1) %>%
  as_tibble() %>% 
  left_join(rRnaProp) %>%
  dplyr::select(
    PC1, PC2, PC3, 
    Mean_GC = mn_gc, 
    Mean_Length = mn_logLen, 
    rRna_Proportion = proportion
  ) %>% 
  cor()
pcaCor_2 <- pca_2$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sum_2) %>%
  as_tibble() %>% 
  left_join(rRnaProp) %>%
  dplyr::select(
    PC1, PC2, PC3, 
    Mean_GC = mn_gc, 
    Mean_Length = mn_logLen, 
    rRna_Proportion = proportion
  ) %>% 
  cor()
```

### Dataset 1 (*D. rerio*)

```{r pcaPlots_1_tx, fig.cap="*PCA plot showing rRNA proportion, mean GC content and mean log(length) after summarisation to gene-level.*"}
a_1 <- pca_1$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  as_tibble() %>%
  ggplot(aes(PC1, PC2)) +
  geom_point(size = 2) +
  labs(
    x = paste0("PC1 (", percent(summary(pca_1)$importance["Proportion of Variance","PC1"]),")"),
    y = paste0("PC2 (", percent(summary(pca_1)$importance["Proportion of Variance","PC2"]),")")
  )
b_1 <- pca_1$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(rRnaProp) %>%
  ggplot(aes(PC1, proportion)) +
  geom_point(size = 2) +
  geom_smooth(method = "lm") +
  scale_y_continuous(labels = percent) +
  labs(
    x = paste0("PC1 (", percent(summary(pca_1)$importance["Proportion of Variance","PC1"]),")"),
    y = "rRNA Proportion of Initial Library"
  )
c_1 <- pca_1$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sum_1) %>%
  left_join(rawGC) %>%
  as_tibble() %>%
  ggplot(aes(PC1, mn_logLen)) +
  geom_point(size = 2) +
  geom_smooth(method = "lm") +
  labs(
    x = paste0("PC1 (", percent(summary(pca_1)$importance["Proportion of Variance","PC1"]),")"),
    y = "Mean log(Length)"
  )
d_1 <- pca_1$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sum_1) %>%
  left_join(rawGC) %>%
  as_tibble() %>%
  ggplot(aes(PC1, mn_gc)) +
  geom_point(size = 2) +
  geom_smooth(method = "lm") +
  scale_y_continuous(labels = percent) +
  labs(
    x = paste0("PC1 (", percent(summary(pca_1)$importance["Proportion of Variance","PC1"]),")"),
    y = "Mean GC"
  )
ggarrange(a_1, b_1, c_1, d_1, ncol = 2, nrow = 2) %>%
  annotate_figure("E-GEOD-71609")
```

```{r corrplot_1_tx, fig.cap="*Correlations between the first three principal components and measured variables: mean GC content, mean log(length) and rRNA proportion.*"}
corrplot(
  pcaCor_1,
  type = "lower", 
  diag = FALSE, 
  addCoef.col = 1, addCoefasPercent = TRUE
)
```

### Dataset 2 (*D. rerio*)

```{r pcaPlots_2_tx, fig.cap="*PCA plot showing rRNA proportion, mean GC content and mean log(length) after summarisation to gene-level.*"}
a_2 <- pca_2$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  as_tibble() %>%
  ggplot(aes(PC1, PC2)) +
  geom_point(size = 2) +
  labs(
    x = paste0("PC1 (", percent(summary(pca_2)$importance["Proportion of Variance","PC1"]),")"),
    y = paste0("PC2 (", percent(summary(pca_2)$importance["Proportion of Variance","PC2"]),")")
  )
b_2 <- pca_2$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(rRnaProp) %>%
  ggplot(aes(PC1, proportion)) +
  geom_point(size = 2) +
  geom_smooth(method = "lm") +
  scale_y_continuous(labels = percent) +
  labs(
    x = paste0("PC1 (", percent(summary(pca_2)$importance["Proportion of Variance","PC1"]),")"),
    y = "rRNA Proportion of Initial Library"
  )
c_2 <- pca_2$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sum_2) %>%
  left_join(rawGC) %>%
  as_tibble() %>%
  ggplot(aes(PC1, mn_logLen)) +
  geom_point(size = 2) +
  geom_smooth(method = "lm") +
  labs(
    x = paste0("PC1 (", percent(summary(pca_2)$importance["Proportion of Variance","PC1"]),")"),
    y = "Mean log(Length)"
  )
d_2 <- pca_2$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sum_2) %>%
  left_join(rawGC) %>%
  as_tibble() %>%
  ggplot(aes(PC1, mn_gc)) +
  geom_point(size = 2) +
  geom_smooth(method = "lm") +
  scale_y_continuous(labels = percent) +
  labs(
    x = paste0("PC1 (", percent(summary(pca_2)$importance["Proportion of Variance","PC1"]),")"),
    y = "Mean GC"
  )
ggarrange(a_2, b_2, c_2, d_2, ncol = 2, nrow = 2) %>%
  annotate_figure("E-GEOD-72322")
```

```{r corrplot_2_tx, fig.cap="*Correlations between the first three principal components and measured variables: mean GC content, mean log(length) and rRNA proportion.*"}
corrplot(
  pcaCor_2,
  type = "lower", 
  diag = FALSE, 
  addCoef.col = 1, addCoefasPercent = TRUE
)
```

## Differential abundance

```{r design_tx}
design_1 <- model.matrix(~proportion, data = dgeFilt_1$samples)
design_2 <- model.matrix(~proportion, data = dgeFilt_2$samples)
```

```{r voom_tx}
voom_1 <- voom(dgeFilt_1, design = design_1, plot = T)
voom_2 <- voom(dgeFilt_2, design = design_2, plot = T)
```

```{r fit_tx}
fit_1 <- lmFit(voom_1, design = design_1)
fit_2 <- lmFit(voom_2, design = design_2)
```

```{r eBayes_tx}
eBayes_1 <- eBayes(fit_1)
eBayes_2 <- eBayes(fit_2)
```

