---
title: "rRNADepletion_StrippedSerum"
author: "Lachlan Baer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    fig_width: 8
    fig_height: 6
    fig_align: "center"
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  error = FALSE, 
  fig.align = "center"
)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

# Setup 

```{r packages}
library(tidyverse)
library(magrittr)
library(parallel)
library(ngsReports)
library(here)
library(scales)
library(ggpubr)
library(kableExtra)
library(AnnotationHub)
library(ensembldb)
library(edgeR)
library(corrplot)
library(DT)
library(ggrepel)
library(msigdbr)
library(fgsea)
```

```{r options}
if (interactive()) setwd(here::here())
theme_set(theme_bw())
cores <- detectCores() - 2
```

# Sequence information

```{r ah}
ah <- AnnotationHub() %>%
  subset(species == "Homo sapiens") %>%
  subset(rdataclass == "EnsDb")
ensDb <- ah[["AH83216"]]
trEns <- transcripts(ensDb) %>%
  mcols() %>% 
  as_tibble()
trLen <- exonsBy(ensDb, "tx") %>%
  width() %>%
  vapply(sum, integer(1))
geneGcLen <- trLen %>%
  enframe() %>%
  set_colnames(c("tx_id", "length")) %>%
  left_join(trEns) %>%
  group_by(gene_id) %>% 
  summarise(
    aveLen = mean(length),
    maxLen = max(length), 
    aveGc = sum(length * gc_content) / sum(length),
    longestGc = gc_content[which.max(length)[[1]]]
  ) %>%
  mutate(
    aveGc =  aveGc / 100,
    longestGc = longestGc / 100
  )
trGcLen <- trLen %>%
  enframe() %>%
  set_colnames(c("tx_id", "length")) %>%
  left_join(trEns) %>%
  group_by(tx_id) %>% 
  summarise(
    aveLen = mean(length),
    maxLen = max(length), 
    aveGc = sum(length * gc_content) / sum(length),
    longestGc = gc_content[which.max(length)[[1]]]
  ) %>%
  mutate(
    aveGc =  aveGc / 100,
    longestGc = longestGc / 100
  )
genesGR <- genes(ensDb)
mcols(genesGR) <- mcols(genesGR)[c("gene_id", "gene_name", 
                                   "gene_biotype", "entrezid")]
txGR <- transcripts(ensDb)
mcols(txGR) <- mcols(txGR)[c("tx_id", "tx_name", 
                             "tx_biotype", "tx_id_version", "gene_id")]
```

An `EnsDb` object was obtained for Ensembl release 101 using the `AnnotationHub` package. This provided the GC content and length for every gene and transcript in the release. For humans, this consists of `r nrow(geneGcLen)` genes and `r nrow(trGcLen)` transcripts.

# Raw data

## Sample information

```{r files}
files <- list.files(
  path = "/hpcfs/users/a1647910/20200310_rRNADepletion/4_T47D_ZR75_DHT_StrippedSerum/0_rawData/FastQC",
  pattern = "zip",
  full.names = TRUE
)
```

```{r samples}
samples <- tibble(
  sample = str_remove(basename(files), "_fastqc.zip"),
  dataset = NA,
  organism = NA
) %>%
  mutate(
    dataset = "StrippedSerum",
    organism = "human"
  )
datasets <- samples$dataset %>% 
  unique()
```

The following analysis involves `r nrow(samples)/2` paired-end samples across `r length(datasets)` dataset(s): `r paste(datasets, sep = ", ")`.

## Library sizes

```{r rawFqcList}
rawFqc <- files %>%
  FastqcDataList()
```

```{r rawLibPlots}
data <- grep("GLL", fqName(rawFqc))
labels <- rawFqc[data] %>%
  fqName() %>%
  str_remove("_001\\.fastq\\.gz") %>%
  str_remove("Ps2Ex3M1_")
rawLib <- plotReadTotals(rawFqc[data]) +
  labs(subtitle = "StrippedSerum") + 
  scale_x_discrete(labels = labels)
```

The library sizes of the unprocessed dataset(s) range between `r comma(min(readTotals(rawFqc)$Total_Sequences))` and `r comma(max(readTotals(rawFqc)$Total_Sequences))` reads.

```{r plotRawLibs}
rawLib
```

## GC content

rRNA transcripts are known to have high GC content. Therefore, inspecting the GC content of the raw reads is a logical start point for detecting incomplete rRNA removal. A spike in GC content at ~ 70% is expected if this is the case.

```{r gcDist, fig.cap="*GC content of reads in the dataset. Clear spikes at about 70% GC are observed, which is likely due to incomplete rRNA depletion.*"}
plotly::ggplotly(
  plotGcContent(
    x = rawFqc[data], 
    plotType = "line",
    gcType = "Transcriptome",
    species = "Hsapiens"
  ) +
    labs(title = "StrippedSerum Dataset (H. sapiens)") + 
    theme(legend.position="none")
) 
```

## Overrepresented sequences

```{r overrep}
getModule(rawFqc, "Overrep") %>% 
  group_by(Sequence, Possible_Source) %>% 
  summarise(`Found In` = n(), `Highest Percentage` = max(Percentage)) %>% 
  arrange(desc(`Highest Percentage`), desc(`Found In`)) %>% 
  ungroup() %>% 
  dplyr::slice(1:30) %>%
  mutate(`Highest Percentage` = percent_format(0.01)(`Highest Percentage`/100)) %>%
  kable(
    align = "llrr", 
    caption = paste(
      "Top", nrow(.),"Overrepresented sequences.",
      "The number of samples they were found in is shown,",
      "along with the percentage of the most 'contaminated' sample."
    )
  ) %>%
  kable_styling(
    bootstrap_options = c("striped", "hover", "condensed", "responsive")
  )
```

# Trimmed data

Raw libraries were trimmed using `cutadapt v1.14` to remove Illumina adapter sequences. Bases with PHRED score < 30, NextSeq-induced polyG runs and reads shorter than 35bp were also removed.

```{r trimFqc}
trimFqc <- list.files(
  path = "/hpcfs/users/a1647910/20200310_rRNADepletion/4_T47D_ZR75_DHT_StrippedSerum/1_trimmedData/FastQC",
  pattern = "zip",
  full.names = TRUE
) %>%
  FastqcDataList()
trimStats <- readTotals(rawFqc) %>%
  dplyr::rename(Raw = Total_Sequences) %>%
  left_join(readTotals(trimFqc), by = "Filename") %>%
  dplyr::rename(Trimmed = Total_Sequences) %>%
  mutate(
    Discarded = 1 - Trimmed/Raw,
    Retained = Trimmed / Raw
  )
```

After trimming of adapters between `r percent(min(trimStats$Discarded), 0.01)` and `r percent(max(trimStats$Discarded), 0.01)` of reads were discarded.

# Aligned data

Trimmed reads were:

1. Aligned to rRNA sequences using the `BWA-MEM` algorithm to estimate the proportion of reads that were of rRNA origin within each sample. `BWA-MEM` is recommended for high-quality queries of reads ranging from 70bp to 1Mbp as it is faster and more accurate that alternative algorithms `BWA-backtrack` and `BWA-SW`.

2. Aligned to the *Homo sapiens* GRCh38 genome (Ensembl release 101) using `STAR v2.7.0d` and summarised with `featureCounts` from the `Subread v1.5.2` package. These counts were used for all gene-level analysis.

## rRNA proportions

```{r rRnaProp}
rRnaProp <- read.delim(
  "/hpcfs/users/a1647910/20200310_rRNADepletion/4_T47D_ZR75_DHT_StrippedSerum/3_bwa/log/samples.mapped.all", 
  sep = ":", 
  col.names = c("sample", "proportion"), 
  header = FALSE
) %>% 
  mutate(
    sample = str_remove_all(sample, ".mapped"),
    sample = basename(sample),
    proportion = proportion/100,
    dataset = "StrippedSerum",
    organism = "human"
  ) %>%
  as_tibble()
rRnaProp$dataset %<>%
  factor(levels = c("StrippedSerum"))
```

```{r rRnaPropBar, fig.cap="*Percentages of each library that align to rRNA sequences with `bwa mem`.*"}
rRnaProp %>%
  ggplot(aes(sample, proportion)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(~dataset, scales = "free_x") +
  scale_y_continuous(labels = percent) +
  labs(x = "Sample", y = "Percent of Total", fill = "Read pair") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
```

## Gene GC content and length

```{r counts}
dgeList <- read_tsv("/hpcfs/users/a1647910/20200310_rRNADepletion/4_T47D_ZR75_DHT_StrippedSerum/4_star2pass/featureCounts/genes.out") %>%
  set_colnames(basename(colnames(.))) %>%
  set_colnames(str_remove(colnames(.), "Aligned.sortedByCoord.out.bam")) %>%
  as.data.frame() %>%
  column_to_rownames("Geneid") %>%
  DGEList() %>%
  calcNormFactors()
metaData <- read_tsv("/hpcfs/users/a1647910/20200310_rRNADepletion/4_T47D_ZR75_DHT_StrippedSerum/T47D_ZR75_DHT_StrippedSerum.tsv")
```

```{r addInfo}
dgeList$genes <- genesGR[rownames(dgeList),]
mcols(dgeList$genes) %<>% 
  as.data.frame() %>% 
  left_join(geneGcLen)
addInfo <- tibble(
  sample = rRnaProp$sample,
  dataset = "StrippedSerum",
  organism = "human",
  rRNA = rRnaProp$proportion
) %>%
  left_join(metaData)
dgeList$samples %<>%
  rownames_to_column("rowname") %>%
  mutate(sample = rowname) %>%
  left_join(addInfo) %>%
  column_to_rownames("rowname") %>%
  mutate(
    filenames = paste0(
      "/hpcfs/users/a1647910/20200310_rRNADepletion/",
      "4_T47D_ZR75_DHT_StrippedSerum/4_star2pass/bam/",
      sample,
      "Aligned.sortedByCoord.out.bam"
    )
  )
dgeList$samples$filenames <- list.files(
  "/hpcfs/users/a1647910/20200310_rRNADepletion/4_T47D_ZR75_DHT_StrippedSerum/4_star2pass/bam", 
  pattern = ".bam$", 
  full.names = TRUE
)
dgeList$samples$group <- paste0(
  dgeList$samples$cell_line, "_", dgeList$samples$treat
) %>%
  make.names() %>%
  factor(levels = unique(.))
```

```{r gcFunc}
gcInfo <- function(x) {
  x$counts %>%
    as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    as_tibble() %>%
    pivot_longer(
      cols = colnames(x), 
      names_to = "sample", 
      values_to = "counts"
    ) %>%
    dplyr::filter(
      counts > 0
    ) %>%
    left_join(
      geneGcLen
    ) %>%
    dplyr::select(
      ends_with("id"), sample, counts, aveGc, maxLen
    ) %>%
    split(f = .$sample) %>%
    lapply(
      function(x){
        DataFrame(
          gc = Rle(x$aveGc, x$counts),
          logLen = Rle(log10(x$maxLen), x$counts)
        )
      }
    ) 
}
gcSummary <- function(x) {
  x %>%
    vapply(function(x){
      c(mean(x$gc), sd(x$gc), mean(x$logLen), sd(x$logLen))
    }, numeric(4)
    ) %>%
    t() %>%
    set_colnames(
      c("mn_gc", "sd_gc", "mn_logLen", "sd_logLen")
    ) %>%
    as.data.frame() %>%
    rownames_to_column("sample") %>%
    as_tibble()
}
```

```{r rle}
rle <- gcInfo(dgeList)
```

```{r gcSummary}
sumGc <- gcSummary(rle)
```

```{r gc, fig.height=4, fig.cap="*Comparison of residual bias potentially introduced by incomplete rRNA removal. Regression lines are shown along with standard error bands for each comparison.*"}
a <- sumGc %>%
  left_join(dgeList$samples) %>%
  ggplot(aes(rRNA, mn_logLen)) +
  geom_point(aes(colour = treat, shape = cell_line), size = 3) +
  geom_smooth(method = "lm") +
  scale_x_continuous(labels = percent) +
  labs(
    x = "rRNA Proportion of Initial Library",
    y = "Mean log(Length)",
    colour = "Treatment",
    shape = "Cell line"
  ) 
b <- sumGc %>%
  left_join(dgeList$samples) %>%
  ggplot(aes(rRNA, mn_gc)) +
  geom_point(aes(colour = treat, shape = cell_line), size = 3) +
  geom_smooth(method = "lm") +
  scale_y_continuous(labels = percent) +
  scale_x_continuous(labels = percent) +
  labs(
    x = "rRNA Proportion of Initial Library",
    y = "Mean GC Content",
    colour = "Treatment",
    shape = "Cell line"
  )
ggarrange(
  a, b, ncol = 2, nrow = 1, 
  common.legend = TRUE, legend = "bottom"
) %>%
  annotate_figure("StrippedSerum Dataset (H. sapiens)")
```

## PCA 

```{r dgeFilt}
genes2keep <- dgeList %>%
  cpm() %>%
  is_greater_than(1) %>%
  rowSums() %>%
  is_weakly_greater_than(6)
dgeFilt <- dgeList[genes2keep,, keep.lib.sizes = FALSE] %>%
  calcNormFactors()
```

```{r pca}
pca <- cpm(dgeFilt, log = TRUE) %>%
  t() %>%
  prcomp()
```

```{r pcaCor}
pcaCor <- pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sumGc) %>%
  as_tibble() %>% 
  left_join(dgeList$samples) %>%
  dplyr::select(
    PC1, PC2, PC3, 
    Mean_GC = mn_gc, 
    Mean_Length = mn_logLen, 
    rRna_Proportion = rRNA
  ) %>% 
  cor()
```

```{r pcaPlots, fig.cap="*PCA plot showing rRNA proportion, mean GC content and mean log(length) after summarisation to gene-level.*"}
a <- pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(dgeList$samples) %>%
  as_tibble() %>%
  ggplot(aes(PC1, PC2)) +
  geom_point(aes(colour = treat, shape = cell_line), size = 2) +
  labs(
    x = paste0("PC1 (", percent(summary(pca)$importance["Proportion of Variance","PC1"]),")"),
    y = paste0("PC2 (", percent(summary(pca)$importance["Proportion of Variance","PC2"]),")"),
    colour = "Treatment",
    shape = "Cell line"
  )
b <- pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(dgeList$samples) %>%
  ggplot(aes(PC1, rRNA, label = rRNA)) +
  geom_point(aes(colour = treat, shape = cell_line), size = 2) +
  geom_smooth(method = "lm") +
  geom_text_repel(show.legend = FALSE) +
  scale_y_continuous(labels = percent) +
  labs(
    x = paste0("PC1 (", percent(summary(pca)$importance["Proportion of Variance","PC1"]),")"),
    y = "rRNA Proportion of Initial Library",
    colour = "Treatment",
    shape = "Cell line"
  )
c <- pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sumGc) %>%
  left_join(dgeList$samples) %>%
  as_tibble() %>%
  ggplot(aes(PC1, mn_logLen)) +
  geom_point(aes(colour = treat, shape = cell_line), size = 2) +
  geom_smooth(method = "lm") +
  labs(
    x = paste0("PC1 (", percent(summary(pca)$importance["Proportion of Variance","PC1"]),")"),
    y = "Mean log(Length)",
    colour = "Treatment",
    shape = "Cell line"
  )
d <- pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sumGc) %>%
  left_join(dgeList$samples) %>%
  as_tibble() %>%
  ggplot(aes(PC1, mn_gc)) +
  geom_point(aes(colour = treat, shape = cell_line), size = 2) +
  geom_smooth(method = "lm") +
  scale_y_continuous(labels = percent) +
  labs(
    x = paste0("PC1 (", percent(summary(pca)$importance["Proportion of Variance","PC1"]),")"),
    y = "Mean GC",
    colour = "Treatment",
    shape = "Cell line"
  )
ggarrange(
  a, b, c, d, ncol = 2, nrow = 2,
  common.legend = TRUE, legend = "bottom"
) %>%
  annotate_figure("StrippedSerum")
```

```{r corrplot, fig.cap="*Correlations between the first three principal components and measured variables: mean GC content, mean log(length) and rRNA proportion.*"}
corrplot(
  pcaCor,
  type = "lower", 
  diag = FALSE, 
  addCoef.col = 1, addCoefasPercent = TRUE
)
```

# Differential expression

```{r}
displayRes_de <- function(x){
  de <- x %>%
    dplyr::filter(DE)
  de %>%
    dplyr::slice(1:1000) %>%
    dplyr::select(-gene_biotype, -coef, -DE) %>%
    mutate(across(c("P.Value", "FDR", "Bonf"), ~ sprintf("%.2e", .x))) %>%
    datatable(
      options = list(pageLength = 10), 
      class = "striped hover condensed responsive", 
      filter = "top",
      caption = paste0(
        x$coef[1],
        ": ",
        nrow(de),
        " of ",
        nrow(x),
        " genes were classified as differentially expressed ",
        "with a FDR < 0.05. ",
        "If more than 1000 genes were classified as DE, only the top 1000 are shown."
      )
    ) %>%
    formatRound(c("logFC", "logCPM", "F"), digits = 2)
}
```

## rRNA

```{r}
design_r <- model.matrix(~rRNA, data = dgeFilt$samples)
```

```{r}
fit_r <- dgeFilt %>%
  estimateDisp(design = design_r) %>%
  glmQLFit() 
```

```{r}
res_r <- glmQLFTest(fit_r) %>%
  topTags(n = Inf) %>%
  .[["table"]] %>%
  rename_all(
    str_remove, pattern = "ID."
  ) %>%
  dplyr::select(
    Geneid = gene_id, Symbol = gene_name, gene_biotype, logFC, logCPM, F, 
    P.Value = PValue, FDR, aveLen, maxLen, aveGc, longestGc
  ) %>%
  as_tibble() %>%
  mutate(
    Bonf = p.adjust(P.Value, "bonf"),
    DE = FDR < 0.05
  )
```

```{r}
res_r %>%
  dplyr::filter(DE) %>%
  dplyr::slice(1:1000) %>%
  dplyr::select(-gene_biotype, -DE, -aveLen, -maxLen, -aveGc, -longestGc) %>%
  mutate(across(c("P.Value", "FDR", "Bonf"), ~ sprintf("%.2e", .x))) %>%
  datatable(
    options = list(pageLength = 10), 
    class = "striped hover condensed responsive", 
    filter = "top",
    caption = paste0(
      "rRNA: ",
      nrow(dplyr::filter(res_r, DE)),
      " of ",
      nrow(res_r),
      " genes were classified as differentially expressed ",
      "with a FDR < 0.05. ",
      "If more than 1000 genes were classified as DE, only the top 1000 are shown."
    )
  ) %>%
  formatRound(c("logFC", "logCPM", "F"), digits = 2)
```

## Treatment

```{r}
design_t <- model.matrix(~0+group, data = dgeFilt$samples)
cont_t <- makeContrasts(
  groupT.47D_DHT = groupT.47D_DHT - groupT.47D_Vehicle,
  groupZR.75.1_DHT = groupZR.75.1_DHT - groupZR.75.1_Vehicle,
  levels = colnames(design_t)
)
```

```{r}
fit_t <- dgeFilt %>%
  estimateDisp(design = design_t) %>%
  glmQLFit() 
```

```{r}
res_t <- colnames(cont_t) %>%
  sapply(function(x){
    glmQLFTest(fit_t, contrast = cont_t[,x]) %>%
      topTags(n = Inf) %>%
      .[["table"]] %>%
      rename_all(
        str_remove, pattern = "ID."
      ) %>%
      dplyr::select(
        Geneid = gene_id, Symbol = gene_name, gene_biotype, logFC, logCPM, F, 
        P.Value = PValue, FDR,
      ) %>%
      as_tibble() %>%
      mutate(
        Bonf = p.adjust(P.Value, "bonf"), 
        coef = x,
        DE = FDR < 0.05
      )
  },
  simplify = FALSE)
```

```{r}
displayRes_de(res_t$groupT.47D_DHT)
displayRes_de(res_t$groupZR.75.1_DHT)
```

## rRNA + treatment

```{r}
design_tr <- model.matrix(~0+rRNA+group, data = dgeFilt$samples)
cont_tr <- makeContrasts(
  groupT.47D_DHT = groupT.47D_DHT - groupT.47D_Vehicle,
  groupZR.75.1_DHT = groupZR.75.1_DHT - groupZR.75.1_Vehicle,
  rRNA = rRNA,
  levels = colnames(design_tr)
)
```

```{r}
fit_tr <- dgeFilt %>%
  estimateDisp(design = design_tr) %>%
  glmQLFit() 
```

```{r}
res_tr <- colnames(cont_tr) %>%
  sapply(function(x){
    glmQLFTest(fit_tr, contrast = cont_tr[,x]) %>%
      topTags(n = Inf) %>%
      .[["table"]] %>%
      rename_all(
        str_remove, pattern = "ID."
      ) %>%
      dplyr::select(
        Geneid = gene_id, Symbol = gene_name, gene_biotype, logFC, logCPM, F, 
        P.Value = PValue, FDR,
      ) %>%
      as_tibble() %>%
      mutate(
        Bonf = p.adjust(P.Value, "bonf"), 
        coef = x,
        DE = FDR < 0.05
      )
  },
  simplify = FALSE)
```

```{r}
displayRes_de(res_tr$rRNA)
displayRes_de(res_tr$groupT.47D_DHT)
displayRes_de(res_tr$groupZR.75.1_DHT)
```

# Enrichment

```{r}
entrezGenes <- mcols(dgeList$genes) %>%
  as.data.frame() %>%
  dplyr::filter(!is.na(entrezid)) %>%
  unnest(cols = entrezid) %>%
  dplyr::rename(entrez_gene = entrezid)
```

```{r}
ranks_r <- res_r %>%
      mutate(stat = -sign(logFC) * log10(P.Value)) %>%
      dplyr::arrange(stat) %>%
      with(structure(stat, names = Geneid))
```

```{r}
ranks_t <- res_t %>%
  lapply(function(x){
    x %>%
      mutate(stat = -sign(logFC) * log10(P.Value)) %>%
      dplyr::arrange(stat) %>%
      with(structure(stat, names = Geneid))
  })
```

```{r}
ranks_tr <- res_tr %>%
  lapply(function(x){
    x %>%
      mutate(stat = -sign(logFC) * log10(P.Value)) %>%
      dplyr::arrange(stat) %>%
      with(structure(stat, names = Geneid))
  })
```

```{r}
displayRes_enrich <- function(x, cap){
  x %>%
    unnest(cols = leadingEdge) %>%
    group_by(pathway) %>%
    mutate(
      leadingSize = n(),
      pathway = str_remove(pathway, "HALLMARK_|KEGG_|WP_"),
      pathway = str_trunc(pathway, 33)
    ) %>%
    distinct(pathway, .keep_all = TRUE) %>%
    dplyr::select(-leadingEdge) %>%
    mutate(across(c("pval", "FDR", "padj"), ~ sprintf("%.2e", .x))) %>%
    datatable(
      options = list(pageLength = 10), 
      class = "striped hover condensed responsive", 
      filter = "top",
      caption = paste(cap)
    ) %>%
    formatRound(c("log2err", "ES", "NES"), digits = 2)
}
```

## Databases

```{r}
hm <- msigdbr("Homo sapiens", category = "H")  %>% 
  left_join(entrezGenes) %>%
  dplyr::filter(!is.na(gene_id)) %>%
  distinct(gs_name, gene_id, .keep_all = TRUE)
hmByGene <- hm %>%
  split(f = .$gene_id) %>%
  lapply(extract2, "gs_name")
hmByID <- hm %>%
  split(f = .$gs_name) %>%
  lapply(extract2, "gene_id")
```

```{r}
kg <- msigdbr("Homo sapiens", category = "C2", subcategory = "CP:KEGG")  %>% 
  left_join(entrezGenes) %>%
  dplyr::filter(!is.na(gene_id)) %>%
  distinct(gs_name, gene_id, .keep_all = TRUE)
kgByGene <- kg  %>%
  split(f = .$gene_id) %>%
  lapply(extract2, "gs_name")
kgByID <- kg  %>%
  split(f = .$gs_name) %>%
  lapply(extract2, "gene_id")
```

```{r}
wk <- msigdbr("Homo sapiens", category = "C2", subcategory = "CP:WIKIPATHWAYS")  %>% 
  left_join(entrezGenes) %>%
  dplyr::filter(!is.na(gene_id)) %>%
  distinct(gs_name, gene_id, .keep_all = TRUE)
wkByGene <- wk  %>%
  split(f = .$gene_id) %>%
  lapply(extract2, "gs_name")
wkByID <- wk  %>%
  split(f = .$gs_name) %>%
  lapply(extract2, "gene_id")
```

```{r}
gsSizes <- bind_rows(hm, kg, wk) %>% 
  dplyr::select(gs_name, gene_symbol, gene_id) %>% 
  chop(c(gene_symbol, gene_id)) %>%
  mutate(
    gs_size = vapply(gene_symbol, length, integer(1))
  )
```

## Hallmark

### rRNA

```{r}
fgsea_r_hm <-fgsea(hmByID, ranks_r, eps = 0) %>%
  as_tibble() %>%
  dplyr::rename(FDR = padj) %>%
  mutate(padj = p.adjust(pval, "bonferroni")) %>%
  dplyr::arrange(pval)
```

```{r}
displayRes_enrich(fgsea_r_hm, "rRNA")
```

### Treatment

```{r}
fgsea_t_hm <- ranks_t %>%
  lapply(function(x){
    fgsea(hmByID, x, eps = 0) %>%
      as_tibble() %>%
      dplyr::rename(FDR = padj) %>%
      mutate(padj = p.adjust(pval, "bonferroni")) %>%
      dplyr::arrange(pval)
  })
```

```{r}
displayRes_enrich(fgsea_t_hm$groupT.47D_DHT, "T-47D_DHT")
displayRes_enrich(fgsea_t_hm$groupZR.75.1_DHT, "ZR-75-1_DHT")
```

### rRNA + treatment

```{r}
fgsea_tr_hm <- ranks_tr %>%
  lapply(function(x){
    fgsea(hmByID, x, eps = 0) %>%
      as_tibble() %>%
      dplyr::rename(FDR = padj) %>%
      mutate(padj = p.adjust(pval, "bonferroni")) %>%
      dplyr::arrange(pval)
  })
```

```{r}
displayRes_enrich(fgsea_tr_hm$rRNA, "rRNA")
displayRes_enrich(fgsea_tr_hm$groupT.47D_DHT, "T-47D_DHT")
displayRes_enrich(fgsea_tr_hm$groupZR.75.1_DHT, "ZR-75-1_DHT")
```

## KEGG

### rRNA

```{r}
fgsea_r_kg <-fgsea(kgByID, ranks_r, eps = 0) %>%
  as_tibble() %>%
  dplyr::rename(FDR = padj) %>%
  mutate(padj = p.adjust(pval, "bonferroni")) %>%
  dplyr::arrange(pval)
```

```{r}
displayRes_enrich(fgsea_r_kg, "rRNA")
```

### Treatment

```{r}
fgsea_t_kg <- ranks_t %>%
  lapply(function(x){
    fgsea(kgByID, x, eps = 0) %>%
      as_tibble() %>%
      dplyr::rename(FDR = padj) %>%
      mutate(padj = p.adjust(pval, "bonferroni")) %>%
      dplyr::arrange(pval)
  })
```

```{r}
displayRes_enrich(fgsea_t_kg$groupT.47D_DHT, "T-47D_DHT")
displayRes_enrich(fgsea_t_kg$groupZR.75.1_DHT, "ZR-75-1_DHT")
```

### rRNA + treatment

```{r}
fgsea_tr_kg <- ranks_tr %>%
  lapply(function(x){
    fgsea(kgByID, x, eps = 0) %>%
      as_tibble() %>%
      dplyr::rename(FDR = padj) %>%
      mutate(padj = p.adjust(pval, "bonferroni")) %>%
      dplyr::arrange(pval)
  })
```

```{r}
displayRes_enrich(fgsea_tr_kg$rRNA, "rRNA")
displayRes_enrich(fgsea_tr_kg$groupT.47D_DHT, "T-47D_DHT")
displayRes_enrich(fgsea_tr_kg$groupZR.75.1_DHT, "ZR-75-1_DHT")
```

## Wikipathways

### rRNA

```{r}
fgsea_r_wk <-fgsea(wkByID, ranks_r, eps = 0) %>%
  as_tibble() %>%
  dplyr::rename(FDR = padj) %>%
  mutate(padj = p.adjust(pval, "bonferroni")) %>%
  dplyr::arrange(pval)
```

```{r}
displayRes_enrich(fgsea_r_wk, "rRNA")
```

### Treatment

```{r}
fgsea_t_wk <- ranks_t %>%
  lapply(function(x){
    fgsea(wkByID, x, eps = 0) %>%
      as_tibble() %>%
      dplyr::rename(FDR = padj) %>%
      mutate(padj = p.adjust(pval, "bonferroni")) %>%
      dplyr::arrange(pval)
  })
```

```{r}
displayRes_enrich(fgsea_t_wk$groupT.47D_DHT, "T-47D_DHT")
displayRes_enrich(fgsea_t_wk$groupZR.75.1_DHT, "ZR-75-1_DHT")
```

### rRNA + treatment

```{r}
fgsea_tr_wk <- ranks_tr %>%
  lapply(function(x){
    fgsea(wkByID, x, eps = 0) %>%
      as_tibble() %>%
      dplyr::rename(FDR = padj) %>%
      mutate(padj = p.adjust(pval, "bonferroni")) %>%
      dplyr::arrange(pval)
  })
```

```{r}
displayRes_enrich(fgsea_tr_wk$rRNA, "rRNA")
displayRes_enrich(fgsea_tr_wk$groupT.47D_DHT, "T-47D_DHT")
displayRes_enrich(fgsea_tr_wk$groupZR.75.1_DHT, "ZR-75-1_DHT")
```

# Session info

```{r saveObjects}
save(
  addInfo,
  dgeList,
  dgeFilt,
  res_r,
  res_t,
  res_tr,
  file = here::here(
    "4_T47D_ZR75_DHT_StrippedSerum/R/output/4_1_DE.RData"
  )
)
```

```{r seshInfo}
sessionInfo()
```
