---
title: "rRNADepletion_s4ter"
author: "Lachlan Baer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    fig_width: 8
    fig_height: 6
    fig_align: "center"
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  error = FALSE, 
  fig.align = "center"
)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

# Setup 

```{r packages}
library(tidyverse)
library(magrittr)
library(ngsReports)
library(here)
library(scales)
library(ggpubr)
library(kableExtra)
library(AnnotationHub)
library(ensembldb)
library(edgeR)
library(corrplot)
library(DT)
library(pander)
library(ggrepel)
library(pheatmap)
library(ggdendro)
library(Biostrings)
library(Gviz)
library(Rsamtools)
library(viridis)
library(ggfortify)
```

```{r options}
if (interactive()) setwd(here::here())
theme_set(theme_bw())
cores <- 6
```

```{r loadObjects}
load(
  here::here("4_T47D_ZR75_DHT_StrippedSerum/R/output/4_DE.RData")
)
```

## DE overlap

### DE sequences

```{r dna}
dna <- getGenomeTwoBitFile(ensDb)
```

```{r}
# mart <- useEnsembl(
#     biomart = "genes", 
#     dataset = "hsapiens_gene_ensembl",
#     version = 101
# )
# deUp <- topTable %>%
#   dplyr::filter(DE) %>%
#   dplyr::filter(logFC > 0) %>%
#   .$Geneid
# seq <- getSequence(
#   id = names(upExons)[1], 
#   type = "ensembl_gene_id", 
#   seqType = "coding", 
#   mart = mart
#   ) %>%
#   # dplyr::filter(ensembl_gene_id == names(upExons)[1]) %>%
#   .$coding %>%
#   .[!str_detect(., "Sequence unavailable")] %>%
#   DNAStringSet() %>%
#   unlist()
```

```{r deUp}
deUp <- topTable %>%
  dplyr::filter(DE) %>%
  dplyr::filter(logFC > 0) %>%
  .$Geneid
upExons <- exonsBy(ensDb, by = "gene", filter = GeneIdFilter(deUp)) %>%
  lapply(function(x){
    GenomicRanges::reduce(x)
  }) %>%
  GRangesList()
upStrings <- lapply(
  upExons, 
  function(x){
    getSeq(dna, x) %>%
      unlist()
  }
) %>%
  DNAStringSet()
upLengths <- vapply(upStrings, length, numeric(1)) %>%
  enframe(name = "Geneid", value = "Length")
```

```{r deDown}
deDown <- topTable %>%
  dplyr::filter(DE) %>%
  dplyr::filter(logFC < 0) %>%
  .$Geneid
downExons <- exonsBy(ensDb, by = "gene", filter = GeneIdFilter(deDown)) %>%
  lapply(function(x){
    GenomicRanges::reduce(x)
  }) %>%
  GRangesList()
downStrings <- lapply(
  downExons, 
  function(x){
    getSeq(dna, x) %>%
      unlist() 
  }
) %>%
  DNAStringSet()
downLengths <- vapply(downStrings, length, numeric(1)) %>%
  enframe(name = "Geneid", value = "Length")
```

### k = 8

```{r}
k8deUp <- k8topTable %>%
  dplyr::filter(DE) %>%
  dplyr::filter(logFC > 0) %>%
  .$mer
k8deDown <- k8topTable %>%
  dplyr::filter(DE) %>%
  dplyr::filter(logFC < 0) %>%
  .$mer
```

```{r}
k8overlapUU <- mclapply(
  upStrings,
  function(x){
    vapply(
      k8deUp,
      function(y){
        countPattern(y, x) %>%
          as.logical() %>%
          sum()
      },
      numeric(1)
    ) %>%
      sum()
  },
  mc.cores = cores
) %>%
  unlist() %>%
  enframe(name = "Geneid", value = "Overlap") %>%
  dplyr::arrange(desc(Overlap)) %>%
  left_join(topTable) %>%
  left_join(upLengths) %>%
  dplyr::select(Geneid, Symbol, gene_biotype, AveExpr, logFC, P.Value, Bonf, Length, Overlap) %>%
  mutate(Norm = (Overlap / Length) * 1000) %>%
  dplyr::arrange(desc(Norm))
```

### k = 9

```{r}
k9deUp <- k9topTable %>%
  dplyr::filter(DE) %>%
  dplyr::filter(logFC > 0) %>%
  .$mer
k9deDown <- k9topTable %>%
  dplyr::filter(DE) %>%
  dplyr::filter(logFC < 0) %>%
  .$mer
```

```{r}
# k9overlapUU <- mclapply(
#   upStrings,
#   function(x){
#     vapply(
#       k9deUp,
#       function(y){
#         countPattern(y, x) %>%
#           as.logical() %>%
#           sum()
#       },
#       numeric(1)
#     ) %>%
#       sum()
#   },
#   mc.cores = cores
# ) %>%
#   unlist() %>%
#   enframe(name = "Geneid", value = "Overlap") %>%
#   dplyr::arrange(desc(Overlap)) %>%
#   left_join(topTable) %>%
#   left_join(upLengths) %>%
#   dplyr::select(Geneid, Symbol, gene_biotype, AveExpr, logFC, P.Value, Bonf, Length, Overlap) %>%
#   mutate(Norm = (Overlap / Length) * 1000) %>%
#   dplyr::arrange(desc(Norm))
```

## LM test

```{r orders}
order <- dgeList$samples %>%
  mutate(tmp = str_remove(filenames, "Aligned.sortedByCoord.out.bam")) %>%
  dplyr::arrange(filenames) %>%
  as_tibble() %>%
  dplyr::select(-tmp) %>%
  rownames_to_column("order") %>%
  dplyr::arrange(-rRNA)
```

```{r exons}
exons <- exonsBy(
  ensDb,
  by = "tx",
  columns = c("exon_id", "gene_id", "tx_id", "gene_name"),
  filter = GeneIdFilter(c(deDown, deUp))) %>%
  unlist()
colnames(mcols(exons)) <- str_remove_all(colnames(mcols(exons)), "_id") %>%
  str_replace_all("tx", "transcript")
mcols(exons) <- mcols(exons)[c("gene_name", "gene", "exon", "transcript")]
```

```{r}
lmTest <- function(goi){
  
  ## Exon info
  counts <- dgeList$counts[goi,]
  gm <- subset(exons, gene == goi)
  gmRed <- gm %>% 
    GenomicRanges::reduce()
  st <- GenomicRanges::start(gmRed)
  en <- GenomicRanges::end(gmRed)
  chr <- seqnames(gmRed)@values
  str <- gmRed %>% 
    strand() %>%
    .@values
  gmRange <- mapply(function(x, y){
    seq(x,y,1)
  }, x = st, y = en, SIMPLIFY = FALSE) %>%
    unlist() %>%
    enframe(name = NULL, value = "pos")
  
  ## Kmer info
  gmSeq <- getSeq(dna, gmRed)
  merLoc <- lapply(k8deUp, function(x){
    vmatchPattern(x, gmSeq) %>%
      mapply(function(y,z){
        start(y) + z
      }, y = ., z = st) %>%
      unlist()
  }) %>%
    unlist() %>%
    lapply(
      function(x){
        sapply(x, function(y){
          seq(y, y+7, 1)
        })
      }
    ) %>%
    unlist() %>%
    unique()
  inKmer <- gmRange$pos %in% merLoc
  
  ## Coverage info
  files <- list.files(
    "/hpcfs/users/a1647910/20200310_rRNADepletion/4_T47D_ZR75_DHT_StrippedSerum/4_star2pass/bam", 
    pattern = ".bam$", 
    full.names = TRUE
  )
  param <- ScanBamParam(
    what = c("pos", "qwidth"),
    which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
    flag = scanBamFlag(isUnmappedQuery=FALSE)
  )
  covFunc <- function(x){
    prop <- order$rRNA[order$filenames == x] %>%
      percent(accuracy = 0.1)
    reads <- counts[order$sample[order$filenames == x]]
    pileup(x, scanBamParam = param) %>%
      dplyr::filter(strand == str) %>%
      group_by(pos) %>% 
      summarise(count = sum(count)) %>%
      ungroup() %>%
      mutate(
        seqnames = chr, 
        start = pos, 
        end = pos, 
        strand = str,
        count = count/reads
      ) %>% 
      dplyr::select(seqnames, start, end, strand, count) %>%
      makeGRangesFromDataFrame(keep.extra.columns = TRUE)
  }
  covList <- mclapply(files, covFunc, mc.cores = cores)
  
  ## lm test
  names <- dgeList$samples %>% 
    dplyr::arrange(filenames) %>%
    .$sample
  test <- lapply(covList, function(x){
    tibble(
      pos = start(x),
      cov = mcols(x)$count
    )
  }) %>%
    purrr::reduce(full_join, by = "pos") %>%
    set_colnames(c("pos", names)) %>%
    right_join(gmRange) %>%
    cbind(inKmer) %>%
    as_tibble() %>%
    pivot_longer(
      cols = contains("GLL"), 
      names_to = "sample", 
      values_to = "coverage"
    ) %>%
    left_join(dgeList$samples[,c("sample", "rRNA", "lib.size")], by = "sample")
  test[is.na(test)] <- as.double(0)
  test <- test %>%
    with(
      lm(coverage ~ inKmer * rRNA)
    )

  return(
    tibble(
      Geneid = goi,
      lm = list(test)
    )
  )
  
  # if (sum(inKmer) > 0) {
  #   return(
  #     tibble(
  #       Geneid = goi,
  #       Estimate = test$coefficients["inKmerTRUE:rRNA","Estimate"],
  #       Std.error = test$coefficients["inKmerTRUE:rRNA", "Std. Error"],
  #       t = test$coefficients["inKmerTRUE:rRNA","t value"],
  #       P = test$coefficients["inKmerTRUE:rRNA","Pr(>|t|)"],
  #       adj.R.squared = test$adj.r.squared
  #     )
  #   )
  # } else {
  #   return(
  #     tibble(
  #       Geneid = goi,
  #       Estimate = NA,
  #       Std.error = NA,
  #       t = NA,
  #       P = NA,
  #       adj.R.squared = NA
  #     )
  #   )
  # }
  
}
```

```{r}
lmRes <- mclapply(deUp, lmTest, mc.cores = cores) 
lmRes %>%
  purrr::reduce(rbind) %>%
  dplyr::arrange(P) %>%
  left_join(k8overlapUU, by = "Geneid") %>%
  dplyr::select(Geneid, Symbol, Estimate, Std.error, t, -Length, -Overlap, -Norm, P) %>%
  mutate(
    adjP = p.adjust(P, "fdr"),
    sig = adjP < 0.05
  )
saveRDS(lmRes, here::here("R/files/lmRes.Rds"))
lmRes <- readRDS(here::here("R/files/lmRes.Rds"))
```

## GLM test

```{r covTest}
glmTest <- function(goi){
  
  ## Exon info
  counts <- dgeList$counts[goi,]
  gm <- subset(exons, gene == goi)
  gmRed <- gm %>% 
    GenomicRanges::reduce()
  st <- GenomicRanges::start(gmRed)
  en <- GenomicRanges::end(gmRed)
  chr <- seqnames(gmRed)@values
  str <- gmRed %>% 
    strand() %>%
    .@values
  gmRange <- mapply(function(x, y){
    seq(x,y,1)
  }, x = st, y = en, SIMPLIFY = FALSE) %>%
    unlist() %>%
    enframe(name = NULL, value = "pos")
  
  ## Kmer info
  gmSeq <- getSeq(dna, gmRed)
  merLoc <- lapply(k8deUp, function(x){
    vmatchPattern(x, gmSeq) %>%
      mapply(function(y,z){
        start(y) + z
      }, y = ., z = st) %>%
      unlist()
  }) %>%
    unlist() %>%
    lapply(
      function(x){
        sapply(x, function(y){
          seq(y, y+7, 1)
        })
      }
    ) %>%
    unlist() %>%
    unique()
  inKmer <- gmRange$pos %in% merLoc
  
  ## Coverage info
  files <- list.files(
    "/hpcfs/users/a1647910/20200310_rRNADepletion/4_T47D_ZR75_DHT_StrippedSerum/2_alignedData/bam", 
    pattern = ".bam$", 
    full.names = TRUE
  )
  param <- ScanBamParam(
    what = c("pos", "qwidth"),
    which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
    flag = scanBamFlag(isUnmappedQuery=FALSE)
  )
  covFunc <- function(x){
    prop <- order$rRNA[order$filenames == x] %>%
      percent(accuracy = 0.1)
    reads <- counts[order$sample[order$filenames == x]]
    pileup(x, scanBamParam = param) %>%
      dplyr::filter(strand == str) %>%
      group_by(pos) %>% 
      summarise(count = sum(count)) %>%
      ungroup() %>%
      mutate(
        seqnames = chr, 
        start = pos, 
        end = pos, 
        strand = str
      ) %>% 
      dplyr::select(seqnames, start, end, strand, count) %>%
      makeGRangesFromDataFrame(keep.extra.columns = TRUE)
  }
  covList <- mclapply(files, covFunc, mc.cores = cores)
  
  ## test
  reads <- dgeList$counts[goi,] %>%
    enframe(name = "sample", value = "reads")
  names <- dgeList$samples %>% 
    dplyr::arrange(filenames) %>%
    .$sample
  test <- lapply(covList, function(x){
    tibble(
      pos = start(x),
      cov = mcols(x)$count
    )
  }) %>%
    purrr::reduce(full_join, by = "pos") %>%
    set_colnames(c("pos", names)) %>%
    right_join(gmRange) %>%
    cbind(inKmer) %>%
    as_tibble() %>%
    pivot_longer(
      cols = contains("GLL"), 
      names_to = "sample", 
      values_to = "coverage"
    ) %>%
    left_join(dgeList$samples[,c("sample", "rRNA", "lib.size")], by = "sample") %>%
    left_join(reads, by = "sample")
  test[is.na(test)] <- as.double(0)
  
  test <- test %>%
    with(
      glm(coverage ~ inKmer * rRNA + offset(log(reads)), family = poisson(link = "log"))
    ) %>%
    summary()
  
  if (sum(inKmer) > 0) {
    return(
      tibble(
        Geneid = goi,
        Estimate = test$coefficients["inKmerTRUE:rRNA","Estimate"],
        Std.error = test$coefficients["inKmerTRUE:rRNA", "Std. Error"],
        z = test$coefficients["inKmerTRUE:rRNA","z value"],
        P = test$coefficients["inKmerTRUE:rRNA","Pr(>|z|)"]
      )
    )
  } else {
    return(
      tibble(
        Geneid = goi,
        Estimate = NA,
        Std.error = NA,
        z = NA,
        P = NA
      )
    )
  }
  
}
```

```{r}
glmRes <- mclapply(deUp, glmTest, mc.cores = cores) %>%
  purrr::reduce(rbind) %>%
  dplyr::arrange(P) %>%
  left_join(k8overlapUU, by = "Geneid") %>%
  dplyr::select(Geneid, Symbol, Estimate, Std.error, z, -Length, -Overlap, -Norm, P) %>%
  mutate(
    adjP = p.adjust(P, "fdr"),
    sig = adjP < 0.05
  )
saveRDS(glmRes, here::here("R/files/glmRes.Rds"))
glmRes <- readRDS(here::here("R/files/glmRes.Rds"))
```

```{r}

```

