---
title: "rRNADepletion_hypoxia"
author: "Lachlan Baer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    fig_width: 8
    fig_height: 6
    fig_align: "center"
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  error = FALSE, 
  fig.align = "center"
)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

# Setup 

```{r packages}
library(tidyverse)
library(magrittr)
library(ngsReports)
library(here)
library(scales)
library(ggpubr)
library(kableExtra)
library(AnnotationHub)
library(ensembldb)
library(edgeR)
library(corrplot)
library(DT)
library(pander)
library(ggrepel)
library(pheatmap)
library(ggdendro)
```

```{r options}
if (interactive()) setwd(here::here())
theme_set(theme_bw())
```

## Sequence information

```{r ah}
ah_Dr <- AnnotationHub() %>%
  subset(species == "Danio rerio") %>%
  subset(rdataclass == "EnsDb")
ensDb_Dr <- ah_Dr[["AH83189"]]
trEns_Dr <- transcripts(ensDb_Dr) %>%
  mcols() %>% 
  as_tibble()
trLen_Dr <- exonsBy(ensDb_Dr, "tx") %>%
  width() %>%
  vapply(sum, integer(1))
geneGcLen_Dr <- trLen_Dr %>%
  enframe() %>%
  set_colnames(c("tx_id", "length")) %>%
  left_join(trEns_Dr) %>%
  group_by(gene_id) %>% 
  summarise(
    aveLen = mean(length),
    maxLen = max(length), 
    aveGc = sum(length * gc_content) / sum(length),
    longestGc = gc_content[which.max(length)[[1]]]
  ) %>%
  mutate(
    aveGc =  aveGc / 100,
    longestGc = longestGc / 100
  )
trGcLen_Dr <- trLen_Dr %>%
  enframe() %>%
  set_colnames(c("tx_id", "length")) %>%
  left_join(trEns_Dr) %>%
  group_by(tx_id) %>% 
  summarise(
    aveLen = mean(length),
    maxLen = max(length), 
    aveGc = sum(length * gc_content) / sum(length),
    longestGc = gc_content[which.max(length)[[1]]]
  ) %>%
  mutate(
    aveGc =  aveGc / 100,
    longestGc = longestGc / 100
  )
genesGR_Dr <- genes(ensDb_Dr)
mcols(genesGR_Dr) <- mcols(genesGR_Dr)[c("gene_id", "gene_name", 
                                         "gene_biotype", "entrezid")]
txGR_Dr <- transcripts(ensDb_Dr)
mcols(txGR_Dr) <- mcols(txGR_Dr)[c("tx_id", "tx_name", 
                                   "tx_biotype", "tx_id_version", "gene_id")]
```

An `EnsDb` object was obtained for Ensembl release 101 using the `AnnotationHub` package. This provided the GC content and length for every gene and transcript in the release. For zebrafish, this consists of `r nrow(geneGcLen_Dr)` genes and `r nrow(trGcLen_Dr)` transcripts.

# Raw data

This is a polyA RNA-seq dataset generated from two families of zebrafish at different ages, 6 and 24 months. Both families consist of sibling WT zebrafish (*Danio rerio*) and heterozygous mutants (psen1^Q96K97del/+^), and samples were generated from these fish under both normal conditions and hypoxic stress. This dataset is intended to be used as a negative control for differential k-mer expression in comparison to the psen2S4ter dataset. [Previous analysis](https://github.com/UofABioinformaticsHub/20170906_Morgan_Hypoxia) showed strong DE signal between WT fish under normal conditions compared to those under hypoxic stress in both age groups. The 6-month fish from this dataset will therefore be investigated for differential expression of k-mers, to identify whether differential expression of k-mers is a function of differential gene expression. 

## Sample information

```{r files}
files <- list.files(
  path = "/hpcfs/users/a1647910/20200910_Hypoxia/0_rawData/FastQC",
  pattern = "zip",
  full.names = TRUE
)
```

```{r samples}
samples <- tibble(
  sample = str_remove(basename(files), "_fastqc.zip"),
  dataset = NA,
  organism = NA
) %>%
  mutate(
    dataset = "Hypoxia",
    organism = "zebrafish"
  )
datasets <- samples$dataset %>% 
  unique()
```

The following analysis involves `r nrow(samples)` single-end samples across `r length(datasets)` dataset(s): `r paste(datasets, sep = ", ")`.

## Library sizes

```{r rawFqcList}
rawFqc <- files %>%
  FastqcDataList()
```

```{r rawLibPlots}
data <- grep("", fqName(rawFqc))
labels <- rawFqc[data] %>%
  fqName() %>%
  str_remove("[0-9]+-MORGAN-6P-") %>%
  str_remove("\\.fastq\\.gz") %>%
  str_remove("_S[0-9]+_R1_001")
rawLib <- plotReadTotals(rawFqc[data]) +
  labs(subtitle = "Hypoxia") + 
  scale_x_discrete(labels = labels)
```

The library sizes of the unprocessed dataset(s) range between `r comma(min(readTotals(rawFqc)$Total_Sequences))` and `r comma(max(readTotals(rawFqc)$Total_Sequences))` reads.

```{r plotRawLibs}
rawLib
```

# Trimmed data

Raw libraries were trimmed using `cutadapt v1.14` to remove Illumina adapter sequences. Bases with PHRED score < 30, NextSeq-induced polyG runs and reads shorter than 35bp were also removed.

```{r trimFqc}
trimFqc <- list.files(
  path = "/hpcfs/users/a1647910/20200910_Hypoxia/1_trimmedData/FastQC",
  pattern = "zip",
  full.names = TRUE
) %>%
  FastqcDataList()
trimStats <- readTotals(rawFqc) %>%
  dplyr::rename(Raw = Total_Sequences) %>%
  left_join(readTotals(trimFqc), by = "Filename") %>%
  dplyr::rename(Trimmed = Total_Sequences) %>%
  mutate(
    Discarded = 1 - Trimmed/Raw,
    Retained = Trimmed / Raw
  )
```

After trimming of adapters between `r percent(min(trimStats$Discarded), 0.01)` and `r percent(max(trimStats$Discarded), 0.01)` of reads were discarded.

# Aligned data

Trimmed reads were:

1. Aligned to rRNA sequences using the `BWA-MEM` algorithm to estimate the proportion of reads that were of rRNA origin within each sample. `BWA-MEM` is recommended for high-quality queries of reads ranging from 70bp to 1Mbp as it is faster and more accurate that alternative algorithms `BWA-backtrack` and `BWA-SW`.

2. Aligned to the *Danio rerio* GRCz11 genome (Ensembl release 101) using `STAR v2.7.0d` and summarised with `featureCounts` from the `Subread v1.5.2` package. These counts were used for all gene-level analysis.

## rRNA proportions

```{r rRnaProp}
bwaMapped <- read.delim(
  "/hpcfs/users/a1647910/20200910_Hypoxia/3_bwa/log/samples.mapped.all", 
  sep = ":", 
  col.names = c("sample", "proportion"), 
  header = FALSE
) %>% 
  mutate(
    sample = str_remove(sample, ".mapped"),
    sample = basename(sample),
    proportion = proportion/100
  )
rRnaProp <- samples %>%
  dplyr::filter(dataset == "Hypoxia") %>%
  left_join(bwaMapped) %>%
  mutate(
    sample = str_remove(sample, "[0-9]+-MORGAN-6P-"),
    sample = str_remove(sample, "_S[0-9]+_R1_001"),
    group = str_extract(sample, "(N|H)")
  )
rRnaProp$dataset %<>%
  factor(levels = c("Hypoxia"))
```

```{r rRnaPropBar, fig.cap="*Percentages of each library that align to rRNA sequences with `bwa mem`.*"}
rRnaProp %>%
  ggplot(aes(sample, proportion)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(~dataset, scales = "free_x") +
  scale_y_continuous(labels = percent) +
  labs(x = "Sample", y = "Percent of Total")
```

```{r rRnaPropBox}
rRnaProp %>%
  ggplot(aes(group, proportion, fill = group)) +
  geom_boxplot() +
  scale_y_continuous(labels = percent) +
  labs(x = "Genotype", y = "Percent of total RNA", title = "rRNA proportions of each genotype") +
  scale_fill_discrete(
    name = "Genotype"
  )
```

## Gene GC content and length

```{r counts}
dgeList <- read_tsv("/hpcfs/users/a1647910/20200910_Hypoxia/2_alignedData/featureCounts/genes.out") %>%
  set_colnames(basename(colnames(.))) %>%
  set_colnames(str_remove(colnames(.), "Aligned.sortedByCoord.out.bam")) %>%
  set_colnames(str_remove(colnames(.), "[0-9]+-MORGAN-6P-")) %>%
  set_colnames(str_remove(colnames(.), "_S[0-9]+_R1_001")) %>%
  as.data.frame() %>%
  column_to_rownames("Geneid") %>%
  DGEList() %>%
  calcNormFactors() 
```

```{r addInfo}
dgeList$genes <- genesGR_Dr[rownames(dgeList),]
mcols(dgeList$genes) %<>% 
  as.data.frame() %>% 
  left_join(geneGcLen_Dr)
addInfo <- tibble(
  sample = rRnaProp$sample,
  dataset = "Hypoxia",
  organism = "zebrafish",
  rRNA = rRnaProp$proportion
)
dgeList$samples %<>%
  rownames_to_column("rowname") %>%
  mutate(sample = rowname) %>%
  left_join(addInfo) %>%
  column_to_rownames("rowname")
dgeList$samples$filenames <- read_tsv(
  "/hpcfs/users/a1647910/20200910_Hypoxia/2_alignedData/featureCounts/genes.out"
) %>% 
  dplyr::select(-Geneid) %>%
  colnames()
dgeList$samples$group <- colnames(dgeList) %>%
  str_extract("(N|H)") %>%
  factor(levels = c("N", "H"))
```

```{r gcFunc}
gcInfo <- function(x) {
  x$counts %>%
    as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    as_tibble() %>%
    pivot_longer(
      cols = colnames(x), 
      names_to = "sample", 
      values_to = "counts"
    ) %>%
    dplyr::filter(
      counts > 0
    ) %>%
    left_join(
      geneGcLen_Dr
    ) %>%
    dplyr::select(
      ends_with("id"), sample, counts, aveGc, maxLen
    ) %>%
    split(f = .$sample) %>%
    lapply(
      function(x){
        DataFrame(
          gc = Rle(x$aveGc, x$counts),
          logLen = Rle(log10(x$maxLen), x$counts)
        )
      }
    ) 
}
gcSummary <- function(x) {
  x %>%
    vapply(function(x){
      c(mean(x$gc), sd(x$gc), mean(x$logLen), sd(x$logLen))
    }, numeric(4)
    ) %>%
    t() %>%
    set_colnames(
      c("mn_gc", "sd_gc", "mn_logLen", "sd_logLen")
    ) %>%
    as.data.frame() %>%
    rownames_to_column("sample") %>%
    as_tibble()
}
```

```{r rle}
rle <- gcInfo(dgeList)
```

```{r gcSummary}
sumGc <- gcSummary(rle)
```

```{r gc, fig.height=4, fig.cap="*Comparison of residual bias potentially introduced by incomplete rRNA removal. Regression lines are shown along with standard error bands for each comparison.*"}
a <- sumGc %>%
  left_join(dgeList$samples) %>%
  ggplot(aes(rRNA, mn_logLen)) +
  geom_point(aes(colour = group), size = 3) +
  geom_smooth(method = "lm") +
  scale_x_continuous(labels = percent) +
  labs(
    x = "rRNA Proportion of Initial Library",
    y = "Mean log(Length)",
    colour = "Genotype"
  ) 
b <- sumGc %>%
  left_join(dgeList$samples) %>%
  ggplot(aes(rRNA, mn_gc)) +
  geom_point(aes(colour = group), size = 3) +
  geom_smooth(method = "lm") +
  scale_y_continuous(labels = percent) +
  scale_x_continuous(labels = percent) +
  labs(
    x = "rRNA Proportion of Initial Library",
    y = "Mean GC Content",
    colour = "Genotype"
  )
ggarrange(
  a, b, ncol = 2, nrow = 1, 
  common.legend = TRUE, legend = "bottom"
) %>%
  annotate_figure("Hypoxia Dataset (D. rerio)")
```

## PCA 

```{r dgeFilt}
genes2keep <- dgeList %>%
  cpm() %>%
  is_greater_than(1) %>%
  rowSums() %>%
  is_weakly_greater_than(4)
dgeFilt <- dgeList[genes2keep,, keep.lib.sizes = FALSE] %>%
  calcNormFactors()
```

```{r pca}
pca <- cpm(dgeFilt, log = TRUE) %>%
  t() %>%
  prcomp()
```

```{r pcaCor}
pcaCor <- pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sumGc) %>%
  as_tibble() %>% 
  left_join(dgeList$samples) %>%
  dplyr::select(
    PC1, PC2, PC3, 
    Mean_GC = mn_gc, 
    Mean_Length = mn_logLen, 
    rRna_Proportion = rRNA
  ) %>% 
  cor()
```

```{r pcaPlots, fig.cap="*PCA plot showing rRNA proportion, mean GC content and mean log(length) after summarisation to gene-level.*"}
a <- pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(dgeList$samples) %>%
  as_tibble() %>%
  ggplot(aes(PC1, PC2)) +
  geom_point(aes(colour = group), size = 2) +
  labs(
    x = paste0("PC1 (", percent(summary(pca)$importance["Proportion of Variance","PC1"]),")"),
    y = paste0("PC2 (", percent(summary(pca)$importance["Proportion of Variance","PC2"]),")"),
    colour = "Genotype"
  )
b <- pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(dgeList$samples) %>%
  ggplot(aes(PC1, rRNA)) +
  geom_point(aes(colour = group), size = 2) +
  geom_smooth(method = "lm") +
  scale_y_continuous(labels = percent) +
  labs(
    x = paste0("PC1 (", percent(summary(pca)$importance["Proportion of Variance","PC1"]),")"),
    y = "rRNA Proportion of Initial Library",
    colour = "Genotype"
  )
c <- pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sumGc) %>%
  left_join(dgeList$samples) %>%
  as_tibble() %>%
  ggplot(aes(PC1, mn_logLen)) +
  geom_point(aes(colour = group), size = 2) +
  geom_smooth(method = "lm") +
  labs(
    x = paste0("PC1 (", percent(summary(pca)$importance["Proportion of Variance","PC1"]),")"),
    y = "Mean log(Length)",
    colour = "Genotype"
  )
d <- pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sumGc) %>%
  left_join(dgeList$samples) %>%
  as_tibble() %>%
  ggplot(aes(PC1, mn_gc)) +
  geom_point(aes(colour = group), size = 2) +
  geom_smooth(method = "lm") +
  scale_y_continuous(labels = percent) +
  labs(
    x = paste0("PC1 (", percent(summary(pca)$importance["Proportion of Variance","PC1"]),")"),
    y = "Mean GC",
    colour = "Genotype"
  )
ggarrange(
  a, b, c, d, ncol = 2, nrow = 2,
  common.legend = TRUE, legend = "bottom"
) %>%
  annotate_figure("Hypoxia")
```

```{r corrplot, fig.cap="*Correlations between the first three principal components and measured variables: mean GC content, mean log(length) and rRNA proportion.*"}
corrplot(
  pcaCor,
  type = "lower", 
  diag = FALSE, 
  addCoef.col = 1, addCoefasPercent = TRUE
)
```

# Differential expression

```{r design}
design <- model.matrix(~group, data = dgeFilt$samples)
```

```{r voom}
voom <- voom(dgeFilt, design = design)
```

```{r fit}
fit <- lmFit(voom, design = design)
```

```{r eBayes}
eBayes <- eBayes(fit)
```

```{r topTable}
topTable <- eBayes %>%
  topTable(coef = colnames(design)[2], sort.by = "p", n = Inf) %>%
  set_colnames(str_remove(colnames(.), "ID\\.")) %>%
  mutate(DE = adj.P.Val < 0.05) %>%
  unite(Location, c(seqnames, start, end, width, strand), sep = ":") %>%
  dplyr::select(
    Geneid = gene_id,
    Symbol = gene_name,
    AveExpr,
    logFC,
    P.Value,
    FDR = adj.P.Val,
    Location,
    t,
    DE,
    everything(),
    -B
  ) %>%
  as_tibble()
```

```{r deResults}
topTable %>% 
  dplyr::select(Geneid, Symbol, AveExpr, logFC, P.Value, FDR, DE) %>%
  mutate(
    AveExpr = format(round(AveExpr, 2), nsmall = 2),
    logFC = format(round(logFC, 2), nsmall = 2),
    P.Value = sprintf("%.2e", P.Value),
    FDR = sprintf("%.2e", FDR)
  ) %>%
  dplyr::slice(1:200) %>%
  datatable(
    options = list(pageLength = 20), 
    class = "striped hover condensed responsive", 
    filter = "top",
    caption = paste(
      "The top 100 differentially expressed genes.",
      nrow(dplyr::filter(topTable, DE)),
      "of",
      nrow(topTable),
      "genes were classified as DE with an FDR < 0.05."
    )
  )
```

# k-mer analysis

## k = 5

#### Counts

```{r k5counts}
k5files <- list.files("/hpcfs/users/a1647910/20200910_Hypoxia/5_jellyfishFq/k5", pattern = "_dumps.txt", full.names = TRUE)
k5counts <- lapply(k5files, function(x){
  read_delim(x, col_names = c("mer", basename(x)), delim = " ") %>%
    set_colnames(str_remove_all(colnames(.), "_S[0-9]+_R1_001|[0-9]+-MORGAN-6P-|_dumps\\.txt"))
}) %>%
  purrr::reduce(full_join)
```

```{r k5dge}
k5dge <- k5counts %>%
  as.data.frame() %>%
  column_to_rownames("mer") %>%
  DGEList() %>%
  calcNormFactors()
k5dge$samples %<>%
  rownames_to_column("rowname") %>%
  mutate(sample = rowname) %>%
  left_join(addInfo) %>%
  column_to_rownames("rowname")
k5dge$samples$group <- colnames(k5dge) %>%
  str_extract("(N|H)") %>%
  factor(levels = c("N", "H"))
```

#### Properties

```{r k5dist}
k5dist <- k5dge %>%
  cpm(log = TRUE) %>%
  as.data.frame() %>%
  pivot_longer(everything(), names_to = "sample", values_to = "count") %>%
  ggplot(aes(x=count, colour = sample)) +
  geom_density() +
  labs(x = "intensity", title = "Distribution of 5-mers")
```

```{r k5den, results="hide"}
k5labels <- k5dge$samples %>% 
  mutate(label = paste0(sample, "\n", percent(rRNA, accuracy = 0.01), " rRNA")) %>% 
  .$label
k5heat <- k5dge %>% 
  cpm(log = TRUE) %>%
  as.data.frame() %>%
  t() %>%
  pheatmap(silent = TRUE, cluster_cols = FALSE,
           show_colnames = FALSE, fontsize = 9,
           fontsize_row = 10, border_color = NA,
           main = "5-mer counts heatmap", labels_row = k5labels)
k5heat$tree_row$labels <- k5labels
k5den <- ggdendrogram(k5heat$tree_row, rotate = TRUE) +
  labs(title = "Hierarchical clustering of 5-mer counts") +
  theme(plot.title = element_text(size = 12))
```

```{r k5pca}
# Assess cpm values to make sure PCA results are not heavily skewed by highly expressed genes
k5pca <- k5dge %>%
  cpm(log = TRUE) %>%
  t() %>%
  prcomp()
```

```{r k5pcaSum, results="hide"}
# Quick inspection to check whether first two PCA components capture most of the variability
summary(k5pca)$importance %>% pander(split.tables = Inf)
```

```{r k5pcaPlot, fig.align = "center"}
# Plot PCA
k5pcaPlot <- k5pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  as_tibble() %>%
  dplyr::select(sample, PC1, PC2) %>%
  left_join(k5dge$samples) %>%
  ggplot(aes(PC1, PC2, colour = group, label = sample)) +
  geom_point(alpha = 0.8, size = 3) +
  geom_text_repel(show.legend = FALSE) +
  labs(
    x = paste0("PC1 (", percent(summary(k5pca)$importance[2, "PC1"]), ")"),
    y = paste0("PC2 (", percent(summary(k5pca)$importance[2, "PC2"]), ")"),
    colour = "Genotype",
    title = "k = 5"
  ) +
  scale_colour_discrete(labels = c("Wildtype", "Mutant"))
```

#### Differential expression

```{r k5DE}
k5design <- model.matrix(~group, data = k5dge$samples)
k5voom <- voom(k5dge, design = k5design)
k5fit <- lmFit(k5voom, design = k5design)
k5eBayes <- eBayes(k5fit)
```

```{r k5topTable}
k5topTable <- k5eBayes %>%
  topTable(coef = colnames(k5design)[2], sort.by = "p", n = Inf) %>%
  set_colnames(str_remove(colnames(.), "ID\\.")) %>%
  rownames_to_column("mer") %>%
  mutate(DE = adj.P.Val < 0.05) %>%
  dplyr::select(
    mer,
    AveExpr,
    logFC,
    P.Value,
    FDR = adj.P.Val,
    t,
    DE,
    everything(),
    -B
  ) %>%
  as_tibble()
```

## k = 6

#### Counts

```{r k6counts}
k6files <- list.files("/hpcfs/users/a1647910/20200910_Hypoxia/5_jellyfishFq/k6", pattern = "_dumps.txt", full.names = TRUE)
k6counts <- lapply(k6files, function(x){
  read_delim(x, col_names = c("mer", basename(x)), delim = " ") %>%
    set_colnames(str_remove_all(colnames(.), "_S[0-9]+_R1_001|[0-9]+-MORGAN-6P-|_dumps\\.txt"))
}) %>%
  purrr::reduce(full_join)
```

```{r k6dge}
k6dge <- k6counts %>%
  as.data.frame() %>%
  column_to_rownames("mer") %>%
  DGEList() %>%
  calcNormFactors()
k6dge$samples %<>%
  rownames_to_column("rowname") %>%
  mutate(sample = rowname) %>%
  left_join(addInfo) %>%
  column_to_rownames("rowname")
k6dge$samples$group <- colnames(k6dge) %>%
  str_extract("(N|H)") %>%
  factor(levels = c("N", "H"))
```

#### Properties

```{r k6dist}
k6dist <- k6dge %>%
  cpm(log = TRUE) %>%
  as.data.frame() %>%
  pivot_longer(everything(), names_to = "sample", values_to = "count") %>%
  ggplot(aes(x=count, colour = sample)) +
  geom_density() +
  labs(x = "intensity", title = "Distribution of 6-mers")
```

```{r k6den, results="hide"}
k6labels <- k6dge$samples %>% 
  mutate(label = paste0(sample, "\n", percent(rRNA, accuracy = 0.01), " rRNA")) %>% 
  .$label
k6heat <- k6dge %>% 
  cpm(log = TRUE) %>%
  as.data.frame() %>%
  t() %>%
  pheatmap(silent = TRUE, cluster_cols = FALSE,
           show_colnames = FALSE, fontsize = 9,
           fontsize_row = 10, border_color = NA,
           main = "6-mer counts heatmap", labels_row = k6labels)
k6heat$tree_row$labels <- k6labels
k6den <- ggdendrogram(k6heat$tree_row, rotate = TRUE) +
  labs(title = "Hierarchical clustering of 6-mer counts") +
  theme(plot.title = element_text(size = 12))
```

```{r k6pca}
# Assess cpm values to make sure PCA results are not heavily skewed by highly expressed genes
k6pca <- k6dge %>%
  cpm(log = TRUE) %>%
  t() %>%
  prcomp()
```

```{r k6pcaSum, results="hide"}
# Quick inspection to check whether first two PCA components capture most of the variability
summary(k6pca)$importance %>% pander(split.tables = Inf)
```

```{r k6pcaPlot, fig.align = "center"}
# Plot PCA
k6pcaPlot <- k6pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  as_tibble() %>%
  dplyr::select(sample, PC1, PC2) %>%
  left_join(k6dge$samples) %>%
  ggplot(aes(PC1, PC2, colour = group, label = sample)) +
  geom_point(alpha = 0.8, size = 3) +
  geom_text_repel(show.legend = FALSE) +
  labs(
    x = paste0("PC1 (", percent(summary(k6pca)$importance[2, "PC1"]), ")"),
    y = paste0("PC2 (", percent(summary(k6pca)$importance[2, "PC2"]), ")"),
    colour = "Genotype",
    title = "k = 6"
  ) +
  scale_colour_discrete(labels = c("Wildtype", "Mutant"))
```

#### Differential expression

```{r k6DE}
k6design <- model.matrix(~group, data = k6dge$samples)
k6voom <- voom(k6dge, design = k6design)
k6fit <- lmFit(k6voom, design = k6design)
k6eBayes <- eBayes(k6fit)
```

```{r k6topTable}
k6topTable <- k6eBayes %>%
  topTable(coef = colnames(k6design)[2], sort.by = "p", n = Inf) %>%
  set_colnames(str_remove(colnames(.), "ID\\.")) %>%
  rownames_to_column("mer") %>%
  mutate(DE = adj.P.Val < 0.05) %>%
  dplyr::select(
    mer,
    AveExpr,
    logFC,
    P.Value,
    FDR = adj.P.Val,
    t,
    DE,
    everything(),
    -B
  ) %>%
  as_tibble()
```

## k = 7

#### Counts

```{r k7counts}
k7files <- list.files("/hpcfs/users/a1647910/20200910_Hypoxia/5_jellyfishFq/k7", pattern = "_dumps.txt", full.names = TRUE)
k7counts <- lapply(k7files, function(x){
  read_delim(x, col_names = c("mer", basename(x)), delim = " ") %>%
    set_colnames(str_remove_all(colnames(.), "_S[0-9]+_R1_001|[0-9]+-MORGAN-6P-|_dumps\\.txt"))
}) %>%
  purrr::reduce(full_join)
```

```{r k7dge}
k7dge <- k7counts %>%
  as.data.frame() %>%
  column_to_rownames("mer") %>%
  DGEList() %>%
  calcNormFactors()
k7dge$samples %<>%
  rownames_to_column("rowname") %>%
  mutate(sample = rowname) %>%
  left_join(addInfo) %>%
  column_to_rownames("rowname")
k7dge$samples$group <- colnames(k7dge) %>%
  str_extract("(N|H)") %>%
  factor(levels = c("N", "H"))
```

#### Properties

```{r k7dist}
k7dist <- k7dge %>%
  cpm(log = TRUE) %>%
  as.data.frame() %>%
  pivot_longer(everything(), names_to = "sample", values_to = "count") %>%
  ggplot(aes(x=count, colour = sample)) +
  geom_density() +
  labs(x = "intensity", title = "Distribution of 7-mers")
```

```{r k7den, results="hide"}
k7labels <- k7dge$samples %>% 
  mutate(label = paste0(sample, "\n", percent(rRNA, accuracy = 0.01), " rRNA")) %>% 
  .$label
k7heat <- k7dge %>% 
  cpm(log = TRUE) %>%
  as.data.frame() %>%
  t() %>%
  pheatmap(silent = TRUE, cluster_cols = FALSE,
           show_colnames = FALSE, fontsize = 9,
           fontsize_row = 10, border_color = NA,
           main = "7-mer counts heatmap", labels_row = k7labels)
k7heat$tree_row$labels <- k7labels
k7den <- ggdendrogram(k7heat$tree_row, rotate = TRUE) +
  labs(title = "Hierarchical clustering of 7-mer counts") +
  theme(plot.title = element_text(size = 12))
```

```{r k7pca}
# Assess cpm values to make sure PCA results are not heavily skewed by highly expressed genes
k7pca <- k7dge %>%
  cpm(log = TRUE) %>%
  t() %>%
  prcomp()
```

```{r k7pcaSum, results="hide"}
# Quick inspection to check whether first two PCA components capture most of the variability
summary(k7pca)$importance %>% pander(split.tables = Inf)
```

```{r k7pcaPlot, fig.align = "center"}
# Plot PCA
k7pcaPlot <- k7pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  as_tibble() %>%
  dplyr::select(sample, PC1, PC2) %>%
  left_join(k7dge$samples) %>%
  ggplot(aes(PC1, PC2, colour = group, label = sample)) +
  geom_point(alpha = 0.8, size = 3) +
  geom_text_repel(show.legend = FALSE) +
  labs(
    x = paste0("PC1 (", percent(summary(k7pca)$importance[2, "PC1"]), ")"),
    y = paste0("PC2 (", percent(summary(k7pca)$importance[2, "PC2"]), ")"),
    colour = "Genotype",
    title = "k = 7"
  ) +
  scale_colour_discrete(labels = c("Wildtype", "Mutant"))
```

#### Differential expression

```{r k7DE}
k7design <- model.matrix(~group, data = k7dge$samples)
k7voom <- voom(k7dge, design = k7design)
k7fit <- lmFit(k7voom, design = k7design)
k7eBayes <- eBayes(k7fit)
```

```{r k7topTable}
k7topTable <- k7eBayes %>%
  topTable(coef = colnames(k7design)[2], sort.by = "p", n = Inf) %>%
  set_colnames(str_remove(colnames(.), "ID\\.")) %>%
  rownames_to_column("mer") %>%
  mutate(DE = adj.P.Val < 0.05) %>%
  dplyr::select(
    mer,
    AveExpr,
    logFC,
    P.Value,
    FDR = adj.P.Val,
    t,
    DE,
    everything(),
    -B
  ) %>%
  as_tibble()
```

## k = 8

#### Counts

```{r k8counts}
k8files <- list.files("/hpcfs/users/a1647910/20200910_Hypoxia/5_jellyfishFq/k8", pattern = "_dumps.txt", full.names = TRUE)
k8counts <- lapply(k8files, function(x){
  read_delim(x, col_names = c("mer", basename(x)), delim = " ") %>%
    set_colnames(str_remove_all(colnames(.), "_S[0-9]+_R1_001|[0-9]+-MORGAN-6P-|_dumps\\.txt"))
}) %>%
  purrr::reduce(full_join)
```

```{r k8dge}
k8dge <- k8counts %>%
  as.data.frame() %>%
  column_to_rownames("mer") %>%
  DGEList() %>%
  calcNormFactors()
k8dge$samples %<>%
  rownames_to_column("rowname") %>%
  mutate(sample = rowname) %>%
  left_join(addInfo) %>%
  column_to_rownames("rowname")
k8dge$samples$group <- colnames(k8dge) %>%
  str_extract("(N|H)") %>%
  factor(levels = c("N", "H"))
```

#### Properties

```{r k8dist}
k8dist <- k8dge %>%
  cpm(log = TRUE) %>%
  as.data.frame() %>%
  pivot_longer(everything(), names_to = "sample", values_to = "count") %>%
  ggplot(aes(x=count, colour = sample)) +
  geom_density() +
  labs(x = "intensity", title = "Distribution of 8-mers")
```

```{r k8den, results="hide"}
k8labels <- k8dge$samples %>% 
  mutate(label = paste0(sample, "\n", percent(rRNA, accuracy = 0.01), " rRNA")) %>% 
  .$label
k8heat <- k8dge %>% 
  cpm(log = TRUE) %>%
  as.data.frame() %>%
  t() %>%
  pheatmap(silent = TRUE, cluster_cols = FALSE,
           show_colnames = FALSE, fontsize = 9,
           fontsize_row = 10, border_color = NA,
           main = "8-mer counts heatmap", labels_row = k8labels)
k8heat$tree_row$labels <- k8labels
k8den <- ggdendrogram(k8heat$tree_row, rotate = TRUE) +
  labs(title = "Hierarchical clustering of 8-mer counts") +
  theme(plot.title = element_text(size = 12))
```

```{r k8pca}
# Assess cpm values to make sure PCA results are not heavily skewed by highly expressed genes
k8pca <- k8dge %>%
  cpm(log = TRUE) %>%
  t() %>%
  prcomp()
```

```{r k8pcaSum, results="hide"}
# Quick inspection to check whether first two PCA components capture most of the variability
summary(k8pca)$importance %>% pander(split.tables = Inf)
```

```{r k8pcaPlot, fig.align = "center"}
# Plot PCA
k8pcaPlot <- k8pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  as_tibble() %>%
  dplyr::select(sample, PC1, PC2) %>%
  left_join(k8dge$samples) %>%
  ggplot(aes(PC1, PC2, colour = group, label = sample)) +
  geom_point(alpha = 0.8, size = 3) +
  geom_text_repel(show.legend = FALSE) +
  labs(
    x = paste0("PC1 (", percent(summary(k8pca)$importance[2, "PC1"]), ")"),
    y = paste0("PC2 (", percent(summary(k8pca)$importance[2, "PC2"]), ")"),
    colour = "Genotype",
    title = "k = 8"
  ) +
  scale_colour_discrete(labels = c("Wildtype", "Mutant"))
```

#### Differential expression

```{r k8DE}
k8design <- model.matrix(~group, data = k8dge$samples)
k8voom <- voom(k8dge, design = k8design)
k8fit <- lmFit(k8voom, design = k8design)
k8eBayes <- eBayes(k8fit)
```

```{r k8topTable}
k8topTable <- k8eBayes %>%
  topTable(coef = colnames(k8design)[2], sort.by = "p", n = Inf) %>%
  set_colnames(str_remove(colnames(.), "ID\\.")) %>%
  rownames_to_column("mer") %>%
  mutate(DE = adj.P.Val < 0.05) %>%
  dplyr::select(
    mer,
    AveExpr,
    logFC,
    P.Value,
    FDR = adj.P.Val,
    t,
    DE,
    everything(),
    -B
  ) %>%
  as_tibble()
```

## k = 9

#### Counts

```{r k9counts}
k9files <- list.files("/hpcfs/users/a1647910/20200910_Hypoxia/5_jellyfishFq/k9", pattern = "_dumps.txt", full.names = TRUE)
k9counts <- lapply(k9files, function(x){
  read_delim(x, col_names = c("mer", basename(x)), delim = " ") %>%
    set_colnames(str_remove_all(colnames(.), "_S[0-9]+_R1_001|[0-9]+-MORGAN-6P-|_dumps\\.txt"))
}) %>%
  purrr::reduce(full_join)
```

```{r k9dge}
k9dge <- k9counts %>%
  as.data.frame() %>%
  column_to_rownames("mer") %>%
  DGEList() %>%
  calcNormFactors()
k9dge$samples %<>%
  rownames_to_column("rowname") %>%
  mutate(sample = rowname) %>%
  left_join(addInfo) %>%
  column_to_rownames("rowname")
k9dge$samples$group <- colnames(k9dge) %>%
  str_extract("(N|H)") %>%
  factor(levels = c("N", "H"))
```

#### Properties

```{r k9dist}
k9dist <- k9dge %>%
  cpm(log = TRUE) %>%
  as.data.frame() %>%
  pivot_longer(everything(), names_to = "sample", values_to = "count") %>%
  ggplot(aes(x=count, colour = sample)) +
  geom_density() +
  labs(x = "intensity", title = "Distribution of 9-mers")
```

```{r k9den, results="hide"}
k9labels <- k9dge$samples %>% 
  mutate(label = paste0(sample, "\n", percent(rRNA, accuracy = 0.01), " rRNA")) %>% 
  .$label
k9heat <- k9dge %>% 
  cpm(log = TRUE) %>%
  as.data.frame() %>%
  t() %>%
  pheatmap(silent = TRUE, cluster_cols = FALSE,
           show_colnames = FALSE, fontsize = 9,
           fontsize_row = 10, border_color = NA,
           main = "9-mer counts heatmap", labels_row = k9labels)
k9heat$tree_row$labels <- k9labels
k9den <- ggdendrogram(k9heat$tree_row, rotate = TRUE) +
  labs(title = "Hierarchical clustering of 9-mer counts") +
  theme(plot.title = element_text(size = 12))
```

```{r k9pca}
# Assess cpm values to make sure PCA results are not heavily skewed by highly expressed genes
k9pca <- k9dge %>%
  cpm(log = TRUE) %>%
  t() %>%
  prcomp()
```

```{r k9pcaSum, results="hide"}
# Quick inspection to check whether first two PCA components capture most of the variability
summary(k9pca)$importance %>% pander(split.tables = Inf)
```

```{r k9pcaPlot, fig.align = "center"}
# Plot PCA
k9pcaPlot <- k9pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  as_tibble() %>%
  dplyr::select(sample, PC1, PC2) %>%
  left_join(k9dge$samples) %>%
  ggplot(aes(PC1, PC2, colour = group, label = sample)) +
  geom_point(alpha = 0.8, size = 3) +
  geom_text_repel(show.legend = FALSE) +
  labs(
    x = paste0("PC1 (", percent(summary(k9pca)$importance[2, "PC1"]), ")"),
    y = paste0("PC2 (", percent(summary(k9pca)$importance[2, "PC2"]), ")"),
    colour = "Genotype",
    title = "k = 9"
  ) +
  scale_colour_discrete(labels = c("Wildtype", "Mutant"))
```

#### Differential expression

```{r k9DE}
k9design <- model.matrix(~group, data = k9dge$samples)
k9voom <- voom(k9dge, design = k9design)
k9fit <- lmFit(k9voom, design = k9design)
k9eBayes <- eBayes(k9fit)
```

```{r k9topTable}
k9topTable <- k9eBayes %>%
  topTable(coef = colnames(k9design)[2], sort.by = "p", n = Inf) %>%
  set_colnames(str_remove(colnames(.), "ID\\.")) %>%
  rownames_to_column("mer") %>%
  mutate(DE = adj.P.Val < 0.05) %>%
  dplyr::select(
    mer,
    AveExpr,
    logFC,
    P.Value,
    FDR = adj.P.Val,
    t,
    DE,
    everything(),
    -B
  ) %>%
  as_tibble()
```

## k = 10

#### Counts

```{r k10counts}
k10files <- list.files("/hpcfs/users/a1647910/20200910_Hypoxia/5_jellyfishFq/k10", pattern = "_dumps.txt", full.names = TRUE)
k10counts <- lapply(k10files, function(x){
  read_delim(x, col_names = c("mer", basename(x)), delim = " ") %>%
    set_colnames(str_remove_all(colnames(.), "_S[0-9]+_R1_001|[0-9]+-MORGAN-6P-|_dumps\\.txt"))
}) %>%
  purrr::reduce(full_join)
```

```{r k10dge}
k10dge <- k10counts %>%
  as.data.frame() %>%
  column_to_rownames("mer") %>%
  DGEList() %>%
  calcNormFactors()
k10dge$samples %<>%
  rownames_to_column("rowname") %>%
  mutate(sample = rowname) %>%
  left_join(addInfo) %>%
  column_to_rownames("rowname")
k10dge$samples$group <- colnames(k10dge) %>%
  str_extract("(N|H)") %>%
  factor(levels = c("N", "H"))
```

#### Properties

```{r k10dist}
k10dist <- k10dge %>%
  cpm(log = TRUE) %>%
  as.data.frame() %>%
  pivot_longer(everything(), names_to = "sample", values_to = "count") %>%
  ggplot(aes(x=count, colour = sample)) +
  geom_density() +
  labs(x = "intensity", title = "Distribution of 10-mers")
```

```{r k10den, results="hide"}
k10labels <- k10dge$samples %>% 
  mutate(label = paste0(sample, "\n", percent(rRNA, accuracy = 0.01), " rRNA")) %>% 
  .$label
k10heat <- k10dge %>% 
  cpm(log = TRUE) %>%
  as.data.frame() %>%
  t() %>%
  pheatmap(silent = TRUE, cluster_cols = FALSE,
           show_colnames = FALSE, fontsize = 9,
           fontsize_row = 10, border_color = NA,
           main = "10-mer counts heatmap", labels_row = k10labels)
k10heat$tree_row$labels <- k10labels
k10den <- ggdendrogram(k10heat$tree_row, rotate = TRUE) +
  labs(title = "Hierarchical clustering of 10-mer counts") +
  theme(plot.title = element_text(size = 12))
```

```{r k10pca}
# Assess cpm values to make sure PCA results are not heavily skewed by highly expressed genes
k10pca <- k10dge %>%
  cpm(log = TRUE) %>%
  t() %>%
  prcomp()
```

```{r k10pcaSum, results="hide"}
# Quick inspection to check whether first two PCA components capture most of the variability
summary(k10pca)$importance %>% pander(split.tables = Inf)
```

```{r k10pcaPlot, fig.align = "center"}
# Plot PCA
k10pcaPlot <- k10pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  as_tibble() %>%
  dplyr::select(sample, PC1, PC2) %>%
  left_join(k10dge$samples) %>%
  ggplot(aes(PC1, PC2, colour = group, label = sample)) +
  geom_point(alpha = 0.8, size = 3) +
  geom_text_repel(show.legend = FALSE) +
  labs(
    x = paste0("PC1 (", percent(summary(k10pca)$importance[2, "PC1"]), ")"),
    y = paste0("PC2 (", percent(summary(k10pca)$importance[2, "PC2"]), ")"),
    colour = "Genotype",
    title = "k = 10"
  ) +
  scale_colour_discrete(labels = c("Wildtype", "Mutant"))
```

#### Differential expression

```{r k10DE}
k10design <- model.matrix(~group, data = k10dge$samples)
k10voom <- voom(k10dge, design = k10design)
k10fit <- lmFit(k10voom, design = k10design)
k10eBayes <- eBayes(k10fit)
```

```{r k10topTable}
k10topTable <- k10eBayes %>%
  topTable(coef = colnames(k10design)[2], sort.by = "p", n = Inf) %>%
  set_colnames(str_remove(colnames(.), "ID\\.")) %>%
  rownames_to_column("mer") %>%
  mutate(DE = adj.P.Val < 0.05) %>%
  dplyr::select(
    mer,
    AveExpr,
    logFC,
    P.Value,
    FDR = adj.P.Val,
    t,
    DE,
    everything(),
    -B
  ) %>%
  as_tibble()
```

## Results

### Distributions

```{r allDists, fig.height=12}
ggarrange(
  k5dist, k6dist, k7dist, k8dist, k9dist, k10dist,
  ncol = 2, nrow = 3, common.legend = TRUE, legend = "bottom"
)
```

### PCA

```{r allPcas, fig.height=12}
ggarrange(
  k5pcaPlot, k6pcaPlot, k7pcaPlot, k8pcaPlot, k9pcaPlot, k10pcaPlot,
  ncol = 2, nrow = 3, common.legend = TRUE, legend = "bottom"
)
```

### Clustering

```{r allDens, fig.height=12}
ggarrange(
  k5den, k6den, k7den, k8den, k9den, k10den,
  ncol = 2, nrow = 3, common.legend = TRUE, legend = "bottom"
)
```

### Differential expression

```{r topResFunc}
topRes <- function(x, cap){
  x %>% 
    dplyr::select(mer, AveExpr, logFC, P.Value, FDR, DE) %>%
    mutate(
      AveExpr = format(round(AveExpr, 2), nsmall = 2),
      logFC = format(round(logFC, 2), nsmall = 2),
      P.Value = sprintf("%.2e", P.Value),
      FDR = sprintf("%.2e", FDR)
    ) %>%
    dplyr::slice(1:100) %>%
    datatable(
      options = list(pageLength = 10),
      class = "striped hover condensed responsive",
      filter = "top",
      caption = cap
    )
}
```

```{r k5top}
topRes(k5topTable,
       cap = paste(
         "The top 100 differentially expressed 5-mers.",
         nrow(dplyr::filter(k5topTable, DE)),
         "of",
         nrow(k5topTable),
         "detected sequences were classified as DE with an FDR < 0.05."
       )
)
```

```{r k6top}
topRes(k6topTable,
       cap = paste(
         "The top 100 differentially expressed 6-mers.",
         nrow(dplyr::filter(k6topTable, DE)),
         "of",
         nrow(k6topTable),
         "detected sequences were classified as DE with an FDR < 0.05."
       )
)
```

```{r k7top}
topRes(k7topTable,
       cap = paste(
         "The top 100 differentially expressed 7-mers.",
         nrow(dplyr::filter(k7topTable, DE)),
         "of",
         nrow(k7topTable),
         "detected sequences were classified as DE with an FDR < 0.05."
       )
)
```

```{r k8top}
topRes(k8topTable,
       cap = paste(
         "The top 100 differentially expressed 8-mers.",
         nrow(dplyr::filter(k8topTable, DE)),
         "of",
         nrow(k8topTable),
         "detected sequences were classified as DE with an FDR < 0.05."
       )
)
```

```{r k9top}
topRes(k9topTable,
       cap = paste(
         "The top 100 differentially expressed 9-mers.",
         nrow(dplyr::filter(k9topTable, DE)),
         "of",
         nrow(k9topTable),
         "detected sequences were classified as DE with an FDR < 0.05."
       )
)
```

```{r k10top}
topRes(k10topTable,
       cap = paste(
         "The top 100 differentially expressed 10-mers.",
         nrow(dplyr::filter(k10topTable, DE)),
         "of",
         nrow(k10topTable),
         "detected sequences were classified as DE with an FDR < 0.05."
       )
)
```