---
title: "rRNA Depletion"
author: "Lachlan Baer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    fig_width: 8
    fig_height: 6
    fig_align: "center"
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  error = FALSE, 
  fig.align = "center"
)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

# Setup 

```{r packages}
library(tidyverse)
library(magrittr)
library(ngsReports)
library(here)
library(scales)
library(ggpubr)
library(kableExtra)
library(AnnotationHub)
library(ensembldb)
library(edgeR)
library(corrplot)
library(cqn)
library(DT)
library(Gviz)
library(AnnotationFilter)
library(Rsamtools)
```

```{r options}
if (interactive()) setwd(here::here())
theme_set(theme_bw())
```

## Sample information

```{r files}
files1 <- list.files(
  path = "0_rawData/FastQC",
  pattern = "zip",
  full.names = TRUE
)
files2 <- list.files(
  path = "../20170327_Psen2S4Ter_RNASeq/data/0_rawData/FastQC",
  pattern = "zip",
  full.names = TRUE
) %>%
  str_subset("R1_fastqc.zip")
filesFull <- c(files1, files2)
files <- filesFull %>%
  basename()
```

```{r samples}
samples <- tibble(
  sample = str_remove(files, "_fastqc.zip"),
  dataset = NA,
  organism = NA
) %>%
  mutate(
    dataset = ifelse(
      str_detect(sample, "SRR213"), "E-GEOD-71609", dataset
    ),
    dataset = ifelse(
      str_detect(sample, "SRR218"), "E-GEOD-72322", dataset
    ),
    dataset = ifelse(
      str_detect(sample, "Ps2Ex"), "Psen2S4Ter", dataset
    ),
    sample = ifelse(
      str_detect(sample, "Ps2Ex"), str_remove(sample, "_R1"), sample
    ),
    organism = ifelse(
      str_detect(sample, "(SRR213|SRR218|Ps2Ex)"), "zebrafish", organism
    ),
    dataset = ifelse(
      str_detect(sample, "ERR313"), "E-MTAB-7636", dataset
    ),
    dataset = ifelse(
      str_detect(sample, "ERR268"), "E-MTAB-6972", dataset
    ),
    organism = ifelse(
      str_detect(sample, "(ERR313|ERR268)"), "mouse", organism
    )
  )
datasets <- samples$dataset %>% 
  unique()
```

## Sequence information

```{r ah_Dr, cache=TRUE}
ah_Dr <- AnnotationHub() %>%
  subset(species == "Danio rerio") %>%
  subset(rdataclass == "EnsDb")
ensDb_Dr <- ah_Dr[["AH74989"]]
trEns_Dr <- transcripts(ensDb_Dr) %>%
  mcols() %>% 
  as_tibble()
trLen_Dr <- exonsBy(ensDb_Dr, "tx") %>%
  width() %>%
  vapply(sum, integer(1))
geneGcLen_Dr <- trLen_Dr %>%
  enframe() %>%
  set_colnames(c("tx_id", "length")) %>%
  left_join(trEns_Dr) %>%
  group_by(gene_id) %>% 
  summarise(
    aveLen = mean(length),
    maxLen = max(length), 
    aveGc = sum(length * gc_content) / sum(length),
    longestGc = gc_content[which.max(length)[[1]]]
  ) %>%
  mutate(
    aveGc =  aveGc / 100,
    longestGc = longestGc / 100
  )
trGcLen_Dr <- trLen_Dr %>%
  enframe() %>%
  set_colnames(c("tx_id", "length")) %>%
  left_join(trEns_Dr) %>%
  group_by(tx_id) %>% 
  summarise(
    aveLen = mean(length),
    maxLen = max(length), 
    aveGc = sum(length * gc_content) / sum(length),
    longestGc = gc_content[which.max(length)[[1]]]
  ) %>%
  mutate(
    aveGc =  aveGc / 100,
    longestGc = longestGc / 100
  )
genesGR_Dr <- genes(ensDb_Dr)
mcols(genesGR_Dr) <- mcols(genesGR_Dr)[c("gene_id", "gene_name", 
                                         "gene_biotype", "entrezid")]
txGR_Dr <- transcripts(ensDb_Dr)
mcols(txGR_Dr) <- mcols(txGR_Dr)[c("tx_id", "tx_name", 
                                   "tx_biotype", "tx_id_version", "gene_id")]
```

```{r ah_Mm, cache=TRUE}
ah_Mm <- AnnotationHub() %>%
  subset(species == "Mus musculus") %>%
  subset(rdataclass == "EnsDb")
ensDb_Mm <- ah_Mm[["AH75036"]]
trEns_Mm <- transcripts(ensDb_Mm) %>%
  mcols() %>% 
  as_tibble()
trLen_Mm <- exonsBy(ensDb_Mm, "tx") %>%
  width() %>%
  vapply(sum, integer(1))
geneGcLen_Mm <- trLen_Mm %>%
  enframe() %>%
  set_colnames(c("tx_id", "length")) %>%
  left_join(trEns_Mm) %>%
  group_by(gene_id) %>% 
  summarise(
    aveLen = mean(length),
    maxLen = max(length), 
    aveGc = sum(length * gc_content) / sum(length),
    longestGc = gc_content[which.max(length)[[1]]]
  ) %>%
  mutate(
    aveGc =  aveGc / 100,
    longestGc = longestGc / 100
  )
trGcLen_Mm <- trLen_Mm %>%
  enframe() %>%
  set_colnames(c("tx_id", "length")) %>%
  left_join(trEns_Mm) %>%
  group_by(tx_id) %>% 
  summarise(
    aveLen = mean(length),
    maxLen = max(length), 
    aveGc = sum(length * gc_content) / sum(length),
    longestGc = gc_content[which.max(length)[[1]]]
  ) %>%
  mutate(
    aveGc =  aveGc / 100,
    longestGc = longestGc / 100
  )
genesGR_Mm <- genes(ensDb_Mm)
mcols(genesGR_Mm) <- mcols(genesGR_Mm)[c("gene_id", "gene_name", 
                                         "gene_biotype", "entrezid")]
```

```{r ah_Hs, cache=TRUE}
ah_Hs <- AnnotationHub() %>%
  subset(species == "Homo sapiens") %>%
  subset(rdataclass == "EnsDb")
ensDb_Hs <- ah_Hs[["AH75011"]]
trEns_Hs <- transcripts(ensDb_Hs) %>%
  mcols() %>% 
  as_tibble()
trLen_Hs <- exonsBy(ensDb_Hs, "tx") %>%
  width() %>%
  vapply(sum, integer(1))
geneGcLen_Hs <- trLen_Hs %>%
  enframe() %>%
  set_colnames(c("tx_id", "length")) %>%
  left_join(trEns_Hs) %>%
  group_by(gene_id) %>% 
  summarise(
    aveLen = mean(length),
    maxLen = max(length), 
    aveGc = sum(length * gc_content) / sum(length),
    longestGc = gc_content[which.max(length)[[1]]]
  ) %>%
  mutate(
    aveGc =  aveGc / 100,
    longestGc = longestGc / 100
  )
trGcLen_Hs <- trLen_Hs %>%
  enframe() %>%
  set_colnames(c("tx_id", "length")) %>%
  left_join(trEns_Hs) %>%
  group_by(tx_id) %>% 
  summarise(
    aveLen = mean(length),
    maxLen = max(length), 
    aveGc = sum(length * gc_content) / sum(length),
    longestGc = gc_content[which.max(length)[[1]]]
  ) %>%
  mutate(
    aveGc =  aveGc / 100,
    longestGc = longestGc / 100
  )
genesGR_Hs <- genes(ensDb_Hs)
mcols(genesGR_Hs) <- mcols(genesGR_Hs)[c("gene_id", "gene_name", 
                                         "gene_biotype", "entrezid")]
```

An `EnsDb` object was obtained for Ensembl release 98 using the `AnnotationHub` package. This provided the GC content and length for every gene and transcript in the release. 

- For *zebrafish*, this consisted of `r nrow(geneGcLen_Dr)` genes and `r nrow(trGcLen_Dr)` transcripts.
- For *mouse*, `r nrow(geneGcLen_Mm)` genes and `r nrow(trGcLen_Mm)` transcripts.
- For *human*, `r nrow(geneGcLen_Hs)` genes and `r nrow(trGcLen_Hs)` transcripts.

# Raw data

Raw data was sourced from publically available datasets on ArrayExpress. Samples from zebrafish (*Danio rerio*) were chosen for an initial inspection as it is expected that rRNA sequences are more divergent from commonly chosen model organisms such as mice. Hence, inefficient ribosomal RNA (rRNA) depletion by kits that are generally optimised for these common model organisms may produce a stronger signal in zebrafish. 

Following an initial inspection, further analysis on human and mouse datasets was performed.

The following analysis involves `r nrow(samples)` samples across `r length(datasets)` datasets: `r paste(datasets, sep = ", ")`.

## Library sizes

```{r rawFqcList}
rawFqc <- filesFull %>%
  FastqcDataList()
```

```{r rawLibPlots}
data1 <- grepl("SRR213", fqName(rawFqc))
rawLib1 <- plotReadTotals(rawFqc[data1]) +
  labs(subtitle = "E-GEOD-71609")
data2 <- grepl("SRR218", fqName(rawFqc))
rawLib2 <- plotReadTotals(rawFqc[data2]) +
  labs(subtitle = "E-GEOD-72322")
data3 <- grepl("Ps2Ex", fqName(rawFqc))
labels3 <- rawFqc[data3] %>%
  fqName() %>%
  str_remove("_6month_07_07_2016_F3") %>%
  str_remove("\\.fastq\\.gz") %>%
  str_remove("Ps2Ex3M1_")
rawLib3 <- plotReadTotals(rawFqc[data3]) +
  labs(subtitle = "Psen2S4Ter") + 
  scale_x_discrete(labels = labels3)
data4 <- grepl("ERR313", fqName(rawFqc))
rawLib4 <- plotReadTotals(rawFqc[data4]) +
  labs(subtitle = "E-MTAB-7636") 
data5 <- grepl("ERR268", fqName(rawFqc))
rawLib5 <- plotReadTotals(rawFqc[data5]) +
  labs(subtitle = "E-MTAB-6972")
```

The library sizes of unprocessed datasets ranged between `r comma(min(readTotals(rawFqc)$Total_Sequences))` and `r comma(max(readTotals(rawFqc)$Total_Sequences))` reads.

```{r plotRawLibs, fig.height=15}
ggarrange(rawLib1, rawLib2, rawLib3, rawLib4, rawLib5, ncol = 1, nrow = 5)
```

## GC content

rRNA transcripts are known to have high GC content. Therefore, surveying the GC content of the raw reads serves as a good starting point for detecting incomplete rRNA removal. A spike in GC content at > 70% is expected if this is the case.

```{r gcDist1, fig.cap="*GC content of reads in the dataset. No obvious spikes above 70% GC are observed, however unusual distibution shapes in some samples are worth further exploration.*"}
plotly::ggplotly(
  plotGcContent(
    x = rawFqc[data1], 
    plotType = "line",
    gcType = "Transcriptome",
    species = "Drerio"
  ) +
    labs(title = "Dataset 1, E-GEOD-71609 (D. rerio)") + 
    theme(legend.position="none")
)
```

```{r gcDist2, fig.cap="*GC content of reads in the dataset. No obvious spikes above 70% GC are observed, however unusual distibution shapes in some samples are worth further exploration.*"}
plotly::ggplotly(
  plotGcContent(
    x = rawFqc[data2], 
    plotType = "line",
    gcType = "Transcriptome",
    species = "Drerio"
  ) +
    labs(title = "Dataset 2, E-GEOD-72322 (D. rerio)") + 
    theme(legend.position="none")
)
```

```{r gcDist3, fig.cap="*GC content of reads in the dataset. No obvious spikes above 70% GC are observed, however unusual distibution shapes in some samples are worth further exploration.*"}
plotly::ggplotly(
  plotGcContent(
    x = rawFqc[data3], 
    plotType = "line",
    gcType = "Transcriptome",
    species = "Drerio"
  ) +
    labs(title = "Dataset 3, Psen2S4Ter (D. rerio)") + 
    theme(legend.position="none")
) 
```

```{r gcDist4, fig.cap="*GC content of reads in the dataset. No obvious spikes above 70% GC are observed, however unusual distibution shapes in some samples are worth further exploration.*"}
plotly::ggplotly(
  plotGcContent(
    x = rawFqc[data4], 
    plotType = "line",
    gcType = "Transcriptome",
    species = "Mmusculus"
  ) +
    labs(title = "Dataset 4, E-MTAB-7636 (M. musculus)") + 
    theme(legend.position="none")
)
```

```{r gcDist5, fig.cap="*GC content of reads in the dataset. No obvious spikes above 70% GC are observed, however unusual distibution shapes in some samples are worth further exploration.*"}
plotly::ggplotly(
  plotGcContent(
    x = rawFqc[data5], 
    plotType = "line",
    gcType = "Transcriptome",
    species = "Mmusculus"
  ) +
    labs(title = "Dataset 5, E-MTAB-6972 (M. musuculus)") + 
    theme(legend.position="none")
)
```
## Overrepresented sequences

The top 30 overrepresented sequences were analysed using `blastn` and  were found to be predominantly rRNA sequences.

```{r overrep}
getModule(rawFqc, "Overrep") %>% 
  group_by(Sequence, Possible_Source) %>% 
  summarise(`Found In` = n(), `Highest Percentage` = max(Percentage)) %>% 
  arrange(desc(`Highest Percentage`), desc(`Found In`)) %>% 
  ungroup() %>% 
  dplyr::slice(1:30) %>%
  mutate(`Highest Percentage` = percent_format(0.01)(`Highest Percentage`/100)) %>%
  kable(
    align = "llrr", 
    caption = paste(
      "Top", nrow(.),"Overrepresented sequences.",
      "The number of samples they were found in is shown,",
      "along with the percentage of the most 'contaminated' sample."
    )
  ) %>%
  kable_styling(
    bootstrap_options = c("striped", "hover", "condensed", "responsive")
  )
```

## Trimming

```{r trimFqc}
trimFqc1 <- list.files(
  path = "1_trimmedData/FastQC", 
  pattern = "zip", 
  full.names = TRUE
) 
trimFqc2 <- list.files(
  path = "../20170327_Psen2S4Ter_RNASeq/data/1_trimmedData/FastQC",
  pattern = "zip",
  full.names = TRUE
)
trimFqc <- c(trimFqc1, trimFqc2) %>%
  FastqcDataList()
trimStats <- readTotals(rawFqc) %>%
  dplyr::rename(Raw = Total_Sequences) %>%
  left_join(readTotals(trimFqc), by = "Filename") %>%
  dplyr::rename(Trimmed = Total_Sequences) %>%
  mutate(
    Discarded = 1 - Trimmed/Raw,
    Retained = Trimmed / Raw
  )
```

After trimming of adapters between `r percent(min(trimStats$Discarded), 0.01)` and `r percent(max(trimStats$Discarded), 0.01)` of reads were discarded.

# Aligned data

Trimmed reads were firstly aligned to rRNA sequences using the `BWA-MEM` algorithm to calculate the proportion of reads that were of rRNA origin within each sample. `BWA-MEM` is recommended for high-quality queries of reads ranging from 70bp to 1Mbp as it is faster and more accurate that alternative algorithms `BWA-backtrack` and `BWA-SW`.

# rRNA proportions

```{r rRnaProp}
prop1 <- samples %>%
  dplyr::filter(dataset == "E-GEOD-71609") %>%
  # cbind(tibble(proportion = c(0.0344, 0.0339, 0.0409, 0.0347, 0.0362, 0.0397, 0.0288, 0.0346, 0.0429, 0.0297))) 
  cbind(tibble(proportion = c(0.0363, 0.0367, 0.0451, 0.0367, 0.0388, 0.0427, 0.0307, 0.0369, 0.0468, 0.0312)))
prop2 <- samples %>%
  dplyr::filter(dataset == "E-GEOD-72322") %>%
  # cbind(tibble(proportion = c(0.1465, 0.1048, 0.1304, 0.1498, 0.1136, 0.1274, 0.1206, 0.1318, 0.1351, 0.1385, 0.1456, 0.1353, 0.1634, 0.1480, 0.1492, 0.1361, 0.1683, 0.1709, 0.1639, 0.1654, 0.1940, 0.1782, 0.1837, 0.1666)))
  cbind(tibble(proportion = c(0.1673, 0.1060, 0.1410, 0.1758, 0.1154, 0.1292, 0.1213, 0.1361, 0.1368, 0.1449, 0.1510, 0.1370, 0.1745, 0.1580, 0.1562, 0.1441, 0.1743, 0.1749, 0.1715, 0.1713, 0.2110, 0.1887, 0.1895, 0.1739)))
prop3 <- samples %>%
  dplyr::filter(dataset == "Psen2S4Ter") %>%
  cbind(tibble(proportion = c(0.2291, 0.2504, 0.2615, 0.2260, 0.2215, 0.2128, 0.2468, 0.2284, 0.1366, 0.1188, 0.1892, 0.1608))) %>%
  mutate(
    sample = str_remove(sample, "_6month_07_07_2016_F3"),
    sample = str_remove(sample, "Ps2Ex3M1_")
  )
prop4 <- samples %>%
  dplyr::filter(dataset == "E-MTAB-7636") %>%
  cbind(tibble(proportion = c(0.0614, 0.0517, 0.0458, 0.0633, 0.0547, 0.0564, 0.0525, 0.0520, 0.0551, 0.0513, 0.0473, 0.0448, 0.0407, 0.0238, 0.0499, 0.0307, 0.0383, 0.0178, 0.0232, 0.0384, 0.0461, 0.0545, 0.0389, 0.0353, 0.0866, 0.0763)))
prop5 <- samples %>%
  dplyr::filter(dataset == "E-MTAB-6972") %>%
  cbind(tibble(proportion = c(0.0123, 0.0122, 0.0121, 0.0121, 0.0121, 0.0131, 0.0131, 0.0131, 0.0130, 0.0131, 0.0929, 0.0924, 0.0931, 0.0929, 0.0927, 0.0224, 0.0223, 0.0222, 0.0223, 0.0222)))
rRnaProp <- rbind(prop1, prop2, prop3, prop4, prop5) 
rRnaProp$dataset %<>%
  factor(levels = c("E-GEOD-71609", "E-GEOD-72322", "Psen2S4Ter", "E-MTAB-7636", "E-MTAB-6972"))
```

```{r plotRRnaProp, fig.cap="*Percentages of each library that align to rRNA sequences with `bwa mem`.*"}
rRnaProp %>%
  ggplot(aes(sample, proportion)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(~dataset, scales = "free_x") +
  scale_y_continuous(labels = percent) +
  labs(x = "Sample", y = "Percent of Total") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
```

Following alignment to rRNA sequences, reads that mapped to rRNA sequences were removed and a separate `fastq` file was created.

These sequences were aligned to the *Danio rerio* GRCz11 genome (Ensembl release 98) using `STAR 2.7.0d` and summarised with `featureCounts`.

# Gene GC content and length

```{r counts}
dgeList_1 <- read_tsv("3_alignedDataStar/featureCounts/genes_Dr.out") %>%
  set_colnames(basename(colnames(.))) %>%
  set_colnames(str_remove(colnames(.), "Aligned.sortedByCoord.out.bam")) %>%
  as.data.frame() %>%
  column_to_rownames("Geneid") %>%
  dplyr::select(str_subset(colnames(.), "SRR213")) %>%
  DGEList() %>%
  calcNormFactors()
dgeList_2 <- read_tsv("3_alignedDataStar/featureCounts/genes_Dr.out") %>%
  set_colnames(basename(colnames(.))) %>%
  set_colnames(str_remove(colnames(.), "Aligned.sortedByCoord.out.bam")) %>%
  as.data.frame() %>%
  column_to_rownames("Geneid") %>%
  dplyr::select(str_subset(colnames(.), "SRR218")) %>%
  DGEList() %>%
  calcNormFactors()
dgeList_3 <- read_tsv("../20170327_Psen2S4Ter_RNASeq/data/2_alignedData/featureCounts/genes.out") %>%
  set_colnames(basename(colnames(.))) %>%
  set_colnames(str_remove(colnames(.), "Aligned.sortedByCoord.out.bam")) %>%
  set_colnames(str_remove(colnames(.), "_6month_07_07_2016_F3")) %>%
  set_colnames(str_remove(colnames(.), "Ps2Ex3M1_")) %>%
  as.data.frame() %>%
  column_to_rownames("Geneid") %>%
  DGEList() %>%
  calcNormFactors()
dgeList_4 <- read_tsv("3_alignedDataStar/featureCounts/genes_Mm.out") %>%
  set_colnames(basename(colnames(.))) %>%
  set_colnames(str_remove(colnames(.), "Aligned.sortedByCoord.out.bam")) %>%
  as.data.frame() %>%
  column_to_rownames("Geneid") %>%
  dplyr::select(str_subset(colnames(.), "ERR313")) %>%
  DGEList() %>%
  calcNormFactors()
dgeList_5 <- read_tsv("3_alignedDataStar/featureCounts/genes_Mm.out") %>%
  set_colnames(basename(colnames(.))) %>%
  set_colnames(str_remove(colnames(.), "Aligned.sortedByCoord.out.bam")) %>%
  as.data.frame() %>%
  column_to_rownames("Geneid") %>%
  dplyr::select(str_subset(colnames(.), "ERR268")) %>%
  DGEList() %>%
  calcNormFactors()
```

```{r addInfo}
dgeList_1$genes <- genesGR_Dr[rownames(dgeList_1),]
mcols(dgeList_1$genes) %<>% 
  as.data.frame() %>% 
  left_join(geneGcLen_Dr) 
dgeList_1$samples %<>%
  rownames_to_column("rowname") %>%
  mutate(sample = rowname) %>%
  left_join(rRnaProp) %>%
  column_to_rownames("rowname")
dgeList_1$samples$filenames <- read_tsv(
  "3_alignedDataStar/featureCounts/genes_Dr.out"
) %>% 
  dplyr::select(str_subset(colnames(.), "SRR213")) %>%
  colnames() 

dgeList_2$genes <- genesGR_Dr[rownames(dgeList_2),]
mcols(dgeList_2$genes) %<>% 
  as.data.frame() %>% 
  left_join(geneGcLen_Dr)
dgeList_2$samples %<>%
  rownames_to_column("rowname") %>%
  mutate(sample = rowname) %>%
  left_join(rRnaProp) %>%
  column_to_rownames("rowname")
dgeList_2$samples$filenames <- read_tsv(
  "3_alignedDataStar/featureCounts/genes_Dr.out"
) %>% 
  dplyr::select(str_subset(colnames(.), "SRR218")) %>%
  colnames()

dgeList_3$genes <- genesGR_Dr[rownames(dgeList_3),]
mcols(dgeList_3$genes) %<>% 
  as.data.frame() %>% 
  left_join(geneGcLen_Dr)
dgeList_3$samples %<>%
  rownames_to_column("rowname") %>%
  mutate(sample = rowname) %>%
  left_join(rRnaProp) %>%
  column_to_rownames("rowname")
## file paths need to be changed to current filepath 
## the original filepath (that featureCounts used) is different to the current
dgeList_3$samples$filenames <- read_tsv(
  "../20170327_Psen2S4Ter_RNASeq/data/2_alignedData/featureCounts/genes.out"
) %>% 
  dplyr::select(str_subset(colnames(.), "Ps2Ex3M1_")) %>%
  colnames() %>%
  str_replace(., "/2_al", "/data/2_al")

dgeList_4$genes <- genesGR_Mm[rownames(dgeList_4),]
mcols(dgeList_4$genes) %<>% 
  as.data.frame() %>% 
  left_join(geneGcLen_Mm)
dgeList_4$samples %<>%
  rownames_to_column("rowname") %>%
  mutate(sample = rowname) %>%
  left_join(rRnaProp) %>%
  column_to_rownames("rowname")
dgeList_4$samples$filenames <- read_tsv(
  "3_alignedDataStar/featureCounts/genes_Mm.out"
) %>% 
  dplyr::select(str_subset(colnames(.), "ERR313")) %>%
  colnames()

dgeList_5$genes <- genesGR_Mm[rownames(dgeList_5),]
mcols(dgeList_5$genes) %<>% 
  as.data.frame() %>% 
  left_join(geneGcLen_Mm)
dgeList_5$samples %<>%
  rownames_to_column("rowname") %>%
  mutate(sample = rowname) %>%
  left_join(rRnaProp) %>%
  column_to_rownames("rowname")
dgeList_5$samples$filenames <- read_tsv(
  "3_alignedDataStar/featureCounts/genes_Mm.out"
) %>% 
  dplyr::select(str_subset(colnames(.), "ERR268")) %>%
  colnames()
```

```{r gcFunc}
gcInfo_Dr <- function(x) {
  x$counts %>%
    as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    as_tibble() %>%
    pivot_longer(
      cols = colnames(x), 
      names_to = "sample", 
      values_to = "counts"
    ) %>%
    dplyr::filter(
      counts > 0
    ) %>%
    left_join(
      geneGcLen_Dr
    ) %>%
    dplyr::select(
      ends_with("id"), sample, counts, aveGc, maxLen
    ) %>%
    split(f = .$sample) %>%
    lapply(
      function(x){
        DataFrame(
          gc = Rle(x$aveGc, x$counts),
          logLen = Rle(log10(x$maxLen), x$counts)
        )
      }
    ) 
}
gcInfo_Mm <- function(x) {
  x$counts %>%
    as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    as_tibble() %>%
    pivot_longer(
      cols = colnames(x), 
      names_to = "sample", 
      values_to = "counts"
    ) %>%
    dplyr::filter(
      counts > 0
    ) %>%
    left_join(
      geneGcLen_Mm
    ) %>%
    dplyr::select(
      ends_with("id"), sample, counts, aveGc, maxLen
    ) %>%
    split(f = .$sample) %>%
    lapply(
      function(x){
        DataFrame(
          gc = Rle(x$aveGc, x$counts),
          logLen = Rle(log10(x$maxLen), x$counts)
        )
      }
    ) 
}
gcSummary <- function(x) {
  x %>%
    vapply(function(x){
      c(mean(x$gc), sd(x$gc), mean(x$logLen), sd(x$logLen))
    }, numeric(4)
    ) %>%
    t() %>%
    set_colnames(
      c("mn_gc", "sd_gc", "mn_logLen", "sd_logLen")
    ) %>%
    as.data.frame() %>%
    rownames_to_column("sample") %>%
    as_tibble()
}
```

```{r rle}
rle_1 <- gcInfo_Dr(dgeList_1)
rle_2 <- gcInfo_Dr(dgeList_2)
rle_3 <- gcInfo_Dr(dgeList_3)
rle_4 <- gcInfo_Mm(dgeList_4)
rle_5 <- gcInfo_Mm(dgeList_5)
```

```{r gcSummary}
sum_1 <- gcSummary(rle_1)
sum_2 <- gcSummary(rle_2)
sum_3 <- gcSummary(rle_3)
sum_4 <- gcSummary(rle_4)
sum_5 <- gcSummary(rle_5)
```

```{r gc_1, fig.height=4, fig.cap="*Comparison of residual bias potentially introduced by incomplete rRNA removal. Regression lines are shown along with standard error bands for each comparison.*"}
a_1 <- sum_1 %>%
  left_join(rRnaProp) %>%
  ggplot(aes(proportion, mn_logLen)) +
  geom_point(size = 3) +
  geom_smooth(method = "lm") +
  scale_x_continuous(labels = percent) +
  labs(
    x = "rRNA Proportion of Initial Library",
    y = "Mean log(Length)"
  ) 
b_1 <- sum_1 %>%
  left_join(rRnaProp) %>%
  ggplot(aes(proportion, mn_gc)) +
  geom_point(size = 3) +
  geom_smooth(method = "lm") +
  scale_y_continuous(labels = percent) +
  scale_x_continuous(labels = percent) +
  labs(
    x = "rRNA Proportion of Initial Library",
    y = "Mean GC Content"
  ) 
ggarrange(a_1, b_1, ncol = 2, nrow = 1) %>%
  annotate_figure("Dataset 1, E-GEOD-71609 (D. rerio)")
```

```{r gc_2, fig.height=4, fig.cap="*Comparison of residual bias potentially introduced by incomplete rRNA removal. Regression lines are shown along with standard error bands for each comparison.*"}
a_2 <- sum_2 %>%
  left_join(rRnaProp) %>%
  ggplot(aes(proportion, mn_logLen)) +
  geom_point(size = 3) +
  geom_smooth(method = "lm") +
  scale_x_continuous(labels = percent) +
  labs(
    x = "rRNA Proportion of Initial Library",
    y = "Mean log(Length)"
  ) 
b_2 <- sum_2 %>%
  left_join(rRnaProp) %>%
  ggplot(aes(proportion, mn_gc)) +
  geom_point(size = 3) +
  geom_smooth(method = "lm") +
  scale_y_continuous(labels = percent) +
  scale_x_continuous(labels = percent) +
  labs(
    x = "rRNA Proportion of Initial Library",
    y = "Mean GC Content"
  )
ggarrange(a_2, b_2, ncol = 2, nrow = 1) %>%
  annotate_figure("Dataset 2, E-GEOD-72322 (D. rerio)")
```

```{r gc_3, fig.height=4, fig.cap="*Comparison of residual bias potentially introduced by incomplete rRNA removal. Regression lines are shown along with standard error bands for each comparison.*"}
a_3 <- sum_3 %>%
  left_join(rRnaProp) %>%
  ggplot(aes(proportion, mn_logLen)) +
  geom_point(size = 3) +
  geom_smooth(method = "lm") +
  scale_x_continuous(labels = percent) +
  labs(
    x = "rRNA Proportion of Initial Library",
    y = "Mean log(Length)"
  ) 
b_3 <- sum_3 %>%
  left_join(rRnaProp) %>%
  ggplot(aes(proportion, mn_gc)) +
  geom_point(size = 3) +
  geom_smooth(method = "lm") +
  scale_y_continuous(labels = percent) +
  scale_x_continuous(labels = percent) +
  labs(
    x = "rRNA Proportion of Initial Library",
    y = "Mean GC Content"
  )
ggarrange(a_3, b_3, ncol = 2, nrow = 1) %>%
  annotate_figure("Dataset 3, PsenS4Ter (D. rerio)")
```

```{r gc_4, fig.height=4, fig.cap="*Comparison of residual bias potentially introduced by incomplete rRNA removal. Regression lines are shown along with standard error bands for each comparison.*"}
a_4 <- sum_4 %>%
  left_join(rRnaProp) %>%
  ggplot(aes(proportion, mn_logLen)) +
  geom_point(size = 3) +
  geom_smooth(method = "lm") +
  scale_x_continuous(labels = percent) +
  labs(
    x = "rRNA Proportion of Initial Library",
    y = "Mean log(Length)"
  ) 
b_4 <- sum_4 %>%
  left_join(rRnaProp) %>%
  ggplot(aes(proportion, mn_gc)) +
  geom_point(size = 3) +
  geom_smooth(method = "lm") +
  scale_y_continuous(labels = percent) +
  scale_x_continuous(labels = percent) +
  labs(
    x = "rRNA Proportion of Initial Library",
    y = "Mean GC Content"
  )
ggarrange(a_4, b_4, ncol = 2, nrow = 1) %>%
  annotate_figure("Dataset 4, E-MTAB-7636 (M. musculus)")
```

```{r gc_5, fig.height=4, fig.cap="*Comparison of residual bias potentially introduced by incomplete rRNA removal. Regression lines are shown along with standard error bands for each comparison.*"}
a_5 <- sum_5 %>%
  left_join(rRnaProp) %>%
  ggplot(aes(proportion, mn_logLen)) +
  geom_point(size = 3) +
  geom_smooth(method = "lm") +
  scale_x_continuous(labels = percent) +
  labs(
    x = "rRNA Proportion of Initial Library",
    y = "Mean log(Length)"
  ) 
b_5 <- sum_5 %>%
  left_join(rRnaProp) %>%
  ggplot(aes(proportion, mn_gc)) +
  geom_point(size = 3) +
  geom_smooth(method = "lm") +
  scale_y_continuous(labels = percent) +
  scale_x_continuous(labels = percent) +
  labs(
    x = "rRNA Proportion of Initial Library",
    y = "Mean GC Content"
  )
ggarrange(a_5, b_5, ncol = 2, nrow = 1) %>%
  annotate_figure("Dataset 5, E-MTAB-6972 (M. musculus)")
```

# PCA 

```{r dgeFilt}
genes2keep_1 <- dgeList_1 %>%
  cpm() %>%
  is_greater_than(1) %>%
  rowSums() %>%
  is_weakly_greater_than(5)
dgeFilt_1 <- dgeList_1[genes2keep_1,, keep.lib.sizes = FALSE] %>%
  calcNormFactors()
genes2keep_2 <- dgeList_2 %>%
  cpm() %>%
  is_greater_than(1) %>%
  rowSums() %>%
  is_weakly_greater_than(12)
dgeFilt_2 <- dgeList_2[genes2keep_2,, keep.lib.sizes = FALSE] %>%
  calcNormFactors()
genes2keep_3 <- dgeList_3 %>%
  cpm() %>%
  is_greater_than(1) %>%
  rowSums() %>%
  is_weakly_greater_than(6)
dgeFilt_3 <- dgeList_3[genes2keep_3,, keep.lib.sizes = FALSE] %>%
  calcNormFactors()
genes2keep_4 <- dgeList_4 %>%
  cpm() %>%
  is_greater_than(1) %>%
  rowSums() %>%
  is_weakly_greater_than(13)
dgeFilt_4 <- dgeList_4[genes2keep_4,, keep.lib.sizes = FALSE] %>%
  calcNormFactors()
genes2keep_5 <- dgeList_5 %>%
  cpm() %>%
  is_greater_than(1) %>%
  rowSums() %>%
  is_weakly_greater_than(10)
dgeFilt_5 <- dgeList_5[genes2keep_5,, keep.lib.sizes = FALSE] %>%
  calcNormFactors()
```

```{r pca}
pca_1 <- cpm(dgeFilt_1, log = TRUE) %>%
  t() %>%
  prcomp()
pca_2 <- cpm(dgeFilt_2, log = TRUE) %>%
  t() %>%
  prcomp()
pca_3 <- cpm(dgeFilt_3, log = TRUE) %>%
  t() %>%
  prcomp()
pca_4 <- cpm(dgeFilt_4, log = TRUE) %>%
  t() %>%
  prcomp()
pca_5 <- cpm(dgeFilt_5, log = TRUE) %>%
  t() %>%
  prcomp()
```

```{r pcaCor}
pcaCor_1 <- pca_1$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sum_1) %>%
  as_tibble() %>% 
  left_join(rRnaProp) %>%
  dplyr::select(
    PC1, PC2, PC3, 
    Mean_GC = mn_gc, 
    Mean_Length = mn_logLen, 
    rRna_Proportion = proportion
  ) %>% 
  cor()
pcaCor_2 <- pca_2$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sum_2) %>%
  as_tibble() %>% 
  left_join(rRnaProp) %>%
  dplyr::select(
    PC1, PC2, PC3, 
    Mean_GC = mn_gc, 
    Mean_Length = mn_logLen, 
    rRna_Proportion = proportion
  ) %>% 
  cor()
pcaCor_3 <- pca_3$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sum_3) %>%
  as_tibble() %>% 
  left_join(rRnaProp) %>%
  dplyr::select(
    PC1, PC2, PC3, 
    Mean_GC = mn_gc, 
    Mean_Length = mn_logLen, 
    rRna_Proportion = proportion
  ) %>% 
  cor()
pcaCor_4 <- pca_4$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sum_4) %>%
  as_tibble() %>% 
  left_join(rRnaProp) %>%
  dplyr::select(
    PC1, PC2, PC3, 
    Mean_GC = mn_gc, 
    Mean_Length = mn_logLen, 
    rRna_Proportion = proportion
  ) %>% 
  cor()
pcaCor_5 <- pca_5$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sum_5) %>%
  as_tibble() %>% 
  left_join(rRnaProp) %>%
  dplyr::select(
    PC1, PC2, PC3, 
    Mean_GC = mn_gc, 
    Mean_Length = mn_logLen, 
    rRna_Proportion = proportion
  ) %>% 
  cor()
```

### Dataset 1 (*D. rerio*)

```{r pcaPlots_1, fig.cap="*PCA plot showing rRNA proportion, mean GC content and mean log(length) after summarisation to gene-level.*"}
a_1 <- pca_1$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  as_tibble() %>%
  ggplot(aes(PC1, PC2)) +
  geom_point(size = 2) +
  labs(
    x = paste0("PC1 (", percent(summary(pca_1)$importance["Proportion of Variance","PC1"]),")"),
    y = paste0("PC2 (", percent(summary(pca_1)$importance["Proportion of Variance","PC2"]),")")
  )
b_1 <- pca_1$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(rRnaProp) %>%
  ggplot(aes(PC1, proportion)) +
  geom_point(size = 2) +
  geom_smooth(method = "lm") +
  scale_y_continuous(labels = percent) +
  labs(
    x = paste0("PC1 (", percent(summary(pca_1)$importance["Proportion of Variance","PC1"]),")"),
    y = "rRNA Proportion of Initial Library"
  )
c_1 <- pca_1$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sum_1) %>%
  as_tibble() %>%
  ggplot(aes(PC1, mn_logLen)) +
  geom_point(size = 2) +
  geom_smooth(method = "lm") +
  labs(
    x = paste0("PC1 (", percent(summary(pca_1)$importance["Proportion of Variance","PC1"]),")"),
    y = "Mean log(Length)"
  )
d_1 <- pca_1$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sum_1) %>%
  as_tibble() %>%
  ggplot(aes(PC1, mn_gc)) +
  geom_point(size = 2) +
  geom_smooth(method = "lm") +
  scale_y_continuous(labels = percent) +
  labs(
    x = paste0("PC1 (", percent(summary(pca_1)$importance["Proportion of Variance","PC1"]),")"),
    y = "Mean GC"
  )
ggarrange(a_1, b_1, c_1, d_1, ncol = 2, nrow = 2) %>%
  annotate_figure("E-GEOD-71609")
```

```{r corrplot_1, fig.cap="*Correlations between the first three principal components and measured variables: mean GC content, mean log(length) and rRNA proportion.*"}
corrplot(
  pcaCor_1,
  type = "lower", 
  diag = FALSE, 
  addCoef.col = 1, addCoefasPercent = TRUE
)
```

### Dataset 2 (*D. rerio*)

```{r pcaPlots_2, fig.cap="*PCA plot showing rRNA proportion, mean GC content and mean log(length) after summarisation to gene-level.*"}
a_2 <- pca_2$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  as_tibble() %>%
  ggplot(aes(PC1, PC2)) +
  geom_point(size = 2) +
  labs(
    x = paste0("PC1 (", percent(summary(pca_2)$importance["Proportion of Variance","PC1"]),")"),
    y = paste0("PC2 (", percent(summary(pca_2)$importance["Proportion of Variance","PC2"]),")")
  )
b_2 <- pca_2$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(rRnaProp) %>%
  ggplot(aes(PC1, proportion)) +
  geom_point(size = 2) +
  geom_smooth(method = "lm") +
  scale_y_continuous(labels = percent) +
  labs(
    x = paste0("PC1 (", percent(summary(pca_2)$importance["Proportion of Variance","PC1"]),")"),
    y = "rRNA Proportion of Initial Library"
  )
c_2 <- pca_2$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sum_2) %>%
  as_tibble() %>%
  ggplot(aes(PC1, mn_logLen)) +
  geom_point(size = 2) +
  geom_smooth(method = "lm") +
  labs(
    x = paste0("PC1 (", percent(summary(pca_2)$importance["Proportion of Variance","PC1"]),")"),
    y = "Mean log(Length)"
  )
d_2 <- pca_2$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sum_2) %>%
  as_tibble() %>%
  ggplot(aes(PC1, mn_gc)) +
  geom_point(size = 2) +
  geom_smooth(method = "lm") +
  scale_y_continuous(labels = percent) +
  labs(
    x = paste0("PC1 (", percent(summary(pca_2)$importance["Proportion of Variance","PC1"]),")"),
    y = "Mean GC"
  )
ggarrange(a_2, b_2, c_2, d_2, ncol = 2, nrow = 2) %>%
  annotate_figure("E-GEOD-72322")
```

```{r corrplot_2, fig.cap="*Correlations between the first three principal components and measured variables: mean GC content, mean log(length) and rRNA proportion.*"}
corrplot(
  pcaCor_2,
  type = "lower", 
  diag = FALSE, 
  addCoef.col = 1, addCoefasPercent = TRUE
)
```

### Dataset 3 (*D. rerio*)

```{r pcaPlots_3, fig.cap="*PCA plot showing rRNA proportion, mean GC content and mean log(length) after summarisation to gene-level.*"}
a_3 <- pca_3$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  as_tibble() %>%
  ggplot(aes(PC1, PC2)) +
  geom_point(size = 2) +
  labs(
    x = paste0("PC1 (", percent(summary(pca_3)$importance["Proportion of Variance","PC1"]),")"),
    y = paste0("PC2 (", percent(summary(pca_3)$importance["Proportion of Variance","PC2"]),")")
  )
b_3 <- pca_3$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(rRnaProp) %>%
  ggplot(aes(PC1, proportion)) +
  geom_point(size = 2) +
  geom_smooth(method = "lm") +
  scale_y_continuous(labels = percent) +
  labs(
    x = paste0("PC1 (", percent(summary(pca_3)$importance["Proportion of Variance","PC1"]),")"),
    y = "rRNA Proportion of Initial Library"
  )
c_3 <- pca_3$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sum_3) %>%
  as_tibble() %>%
  ggplot(aes(PC1, mn_logLen)) +
  geom_point(size = 2) +
  geom_smooth(method = "lm") +
  labs(
    x = paste0("PC1 (", percent(summary(pca_3)$importance["Proportion of Variance","PC1"]),")"),
    y = "Mean log(Length)"
  )
d_3 <- pca_3$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sum_3) %>%
  as_tibble() %>%
  ggplot(aes(PC1, mn_gc)) +
  geom_point(size = 2) +
  geom_smooth(method = "lm") +
  scale_y_continuous(labels = percent) +
  labs(
    x = paste0("PC1 (", percent(summary(pca_3)$importance["Proportion of Variance","PC1"]),")"),
    y = "Mean GC"
  )
ggarrange(a_3, b_3, c_3, d_3, ncol = 2, nrow = 2) %>%
  annotate_figure("Psen2S4Ter")
```

```{r corrplot_3, fig.cap="*Correlations between the first three principal components and measured variables: mean GC content, mean log(length) and rRNA proportion.*"}
corrplot(
  pcaCor_3,
  type = "lower", 
  diag = FALSE, 
  addCoef.col = 1, addCoefasPercent = TRUE
)
```

### Dataset 4 (*M. musculus*)

```{r pcaPlots_4, fig.cap="*PCA plot showing rRNA proportion, mean GC content and mean log(length) after summarisation to gene-level.*"}
a_4 <- pca_4$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  as_tibble() %>%
  ggplot(aes(PC1, PC2)) +
  geom_point(size = 2) +
  labs(
    x = paste0("PC1 (", percent(summary(pca_4)$importance["Proportion of Variance","PC1"]),")"),
    y = paste0("PC2 (", percent(summary(pca_4)$importance["Proportion of Variance","PC2"]),")")
  )
b_4 <- pca_4$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(rRnaProp) %>%
  ggplot(aes(PC1, proportion)) +
  geom_point(size = 2) +
  geom_smooth(method = "lm") +
  scale_y_continuous(labels = percent) +
  labs(
    x = paste0("PC1 (", percent(summary(pca_4)$importance["Proportion of Variance","PC1"]),")"),
    y = "rRNA Proportion of Initial Library"
  )
c_4 <- pca_4$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sum_4) %>%
  as_tibble() %>%
  ggplot(aes(PC1, mn_logLen)) +
  geom_point(size = 2) +
  geom_smooth(method = "lm") +
  labs(
    x = paste0("PC1 (", percent(summary(pca_4)$importance["Proportion of Variance","PC1"]),")"),
    y = "Mean log(Length)"
  )
d_4 <- pca_4$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sum_4) %>%
  as_tibble() %>%
  ggplot(aes(PC1, mn_gc)) +
  geom_point(size = 2) +
  geom_smooth(method = "lm") +
  scale_y_continuous(labels = percent) +
  labs(
    x = paste0("PC1 (", percent(summary(pca_4)$importance["Proportion of Variance","PC1"]),")"),
    y = "Mean GC"
  )
ggarrange(a_4, b_4, c_4, d_4, ncol = 2, nrow = 2) %>%
  annotate_figure("E-MTAB-7636")
```

```{r corrplot_4, fig.cap="*Correlations between the first three principal components and measured variables: mean GC content, mean log(length) and rRNA proportion.*"}
corrplot(
  pcaCor_4,
  type = "lower", 
  diag = FALSE, 
  addCoef.col = 1, addCoefasPercent = TRUE
)
```

### Dataset 5 (*M. musculus*)

```{r pcaPlots_5, fig.cap="*PCA plot showing rRNA proportion, mean GC content and mean log(length) after summarisation to gene-level.*"}
a_5 <- pca_5$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  as_tibble() %>%
  ggplot(aes(PC1, PC2)) +
  geom_point(size = 2) +
  labs(
    x = paste0("PC1 (", percent(summary(pca_5)$importance["Proportion of Variance","PC1"]),")"),
    y = paste0("PC2 (", percent(summary(pca_5)$importance["Proportion of Variance","PC2"]),")")
  )
b_5 <- pca_5$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(rRnaProp) %>%
  ggplot(aes(PC1, proportion)) +
  geom_point(size = 2) +
  geom_smooth(method = "lm") +
  scale_y_continuous(labels = percent) +
  labs(
    x = paste0("PC1 (", percent(summary(pca_5)$importance["Proportion of Variance","PC1"]),")"),
    y = "rRNA Proportion of Initial Library"
  )
c_5 <- pca_5$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sum_5) %>%
  as_tibble() %>%
  ggplot(aes(PC1, mn_logLen)) +
  geom_point(size = 2) +
  geom_smooth(method = "lm") +
  labs(
    x = paste0("PC1 (", percent(summary(pca_5)$importance["Proportion of Variance","PC1"]),")"),
    y = "Mean log(Length)"
  )
d_5 <- pca_5$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sum_5) %>%
  as_tibble() %>%
  ggplot(aes(PC1, mn_gc)) +
  geom_point(size = 2) +
  geom_smooth(method = "lm") +
  scale_y_continuous(labels = percent) +
  labs(
    x = paste0("PC1 (", percent(summary(pca_5)$importance["Proportion of Variance","PC1"]),")"),
    y = "Mean GC"
  )
ggarrange(a_5, b_5, c_5, d_5, ncol = 2, nrow = 2) %>%
  annotate_figure("E-MTAB-7636")
```

```{r corrplot_5, fig.cap="*Correlations between the first three principal components and measured variables: mean GC content, mean log(length) and rRNA proportion.*"}
corrplot(
  pcaCor_5,
  type = "lower", 
  diag = FALSE, 
  addCoef.col = 1, addCoefasPercent = TRUE
)
```

# CQN

```{r cqn}
cqn_1 <- cqn(
  dgeFilt_1$counts, 
  x = mcols(dgeFilt_1$genes)$aveGc,
  lengths = mcols(dgeFilt_1$genes)$aveLen,
  sizeFactors = dgeFilt_1$samples$lib.size
)
cqn_2 <- cqn(
  dgeFilt_2$counts, 
  x = mcols(dgeFilt_2$genes)$aveGc,
  lengths = mcols(dgeFilt_2$genes)$aveLen,
  sizeFactors = dgeFilt_2$samples$lib.size
)
```

```{r pcaCqn}
pcaCqn_1 <- add(cqn_1$y, cqn_1$offset) %>%
  t() %>%
  prcomp()
pcaCqn_2 <- add(cqn_2$y, cqn_2$offset) %>%
  t() %>%
  prcomp()
```

```{r pcaCorcCqn}
pcaCorCqn_1 <- pcaCqn_1$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sum_1) %>%
  as_tibble() %>% 
  left_join(rRnaProp) %>%
  dplyr::select(
    PC1, PC2, PC3, 
    Mean_GC = mn_gc, 
    Mean_Length = mn_logLen, 
    rRna_Proportion = proportion
  ) %>% 
  cor()
pcaCorCqn_2 <- pcaCqn_2$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sum_2) %>%
  as_tibble() %>% 
  left_join(rRnaProp) %>%
  dplyr::select(
    PC1, PC2, PC3, 
    Mean_GC = mn_gc, 
    Mean_Length = mn_logLen, 
    rRna_Proportion = proportion
  ) %>% 
  cor()
```

### Dataset 1 (*D. rerio*)

```{r pcaCqnPlots_1, fig.cap="*PCA plot after CQN showing rRNA proportion, mean GC content and mean log(length) after summarisation to gene-level.*"}
a_1 <- pcaCqn_1$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  as_tibble() %>%
  ggplot(aes(PC1, PC2)) +
  geom_point(size = 2) +
  labs(
    x = paste0("PC1 (", percent(summary(pcaCqn_1)$importance["Proportion of Variance","PC1"]),")"),
    y = paste0("PC2 (", percent(summary(pcaCqn_1)$importance["Proportion of Variance","PC2"]),")")
  )
b_1 <- pcaCqn_1$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(rRnaProp) %>%
  ggplot(aes(PC1, proportion)) +
  geom_point(size = 2) +
  geom_smooth(method = "lm") +
  scale_y_continuous(labels = percent) +
  labs(
    x = paste0("PC1 (", percent(summary(pcaCqn_1)$importance["Proportion of Variance","PC1"]),")"),
    y = "rRNA Proportion of Initial Library"
  )
c_1 <- pcaCqn_1$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sum_1) %>%
  as_tibble() %>%
  ggplot(aes(PC1, mn_logLen)) +
  geom_point(size = 2) +
  geom_smooth(method = "lm") +
  labs(
    x = paste0("PC1 (", percent(summary(pcaCqn_1)$importance["Proportion of Variance","PC1"]),")"),
    y = "Mean log(Length)"
  )
d_1 <- pcaCqn_1$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sum_1) %>%
  as_tibble() %>%
  ggplot(aes(PC1, mn_gc)) +
  geom_point(size = 2) +
  geom_smooth(method = "lm") +
  scale_y_continuous(labels = percent) +
  labs(
    x = paste0("PC1 (", percent(summary(pcaCqn_1)$importance["Proportion of Variance","PC1"]),")"),
    y = "Mean GC"
  )
ggarrange(a_1, b_1, c_1, d_1, ncol = 2, nrow = 2) %>%
  annotate_figure("E-GEOD-71609")
```

```{r corrplotCqn_1, fig.cap="*Correlations between the first three principal components and measured variables: mean GC content, mean log(length) and rRNA proportion.*"}
corrplot(
  pcaCorCqn_1,
  type = "lower", 
  diag = FALSE, 
  addCoef.col = 1, addCoefasPercent = TRUE
)
```

### Dataset 2 (*D. rerio*)

```{r pcaCqnPlots_2, fig.cap="*PCA plot after CQN showing rRNA proportion, mean GC content and mean log(length) after summarisation to gene-level.*"}
a_2 <- pcaCqn_2$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  as_tibble() %>%
  ggplot(aes(PC1, PC2)) +
  geom_point(size = 2) +
  labs(
    x = paste0("PC1 (", percent(summary(pcaCqn_2)$importance["Proportion of Variance","PC1"]),")"),
    y = paste0("PC2 (", percent(summary(pcaCqn_2)$importance["Proportion of Variance","PC2"]),")")
  )
b_2 <- pcaCqn_2$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(rRnaProp) %>%
  ggplot(aes(PC1, proportion)) +
  geom_point(size = 2) +
  geom_smooth(method = "lm") +
  scale_y_continuous(labels = percent) +
  labs(
    x = paste0("PC1 (", percent(summary(pcaCqn_2)$importance["Proportion of Variance","PC1"]),")"),
    y = "rRNA Proportion of Initial Library"
  )
c_2 <- pcaCqn_2$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sum_2) %>%
  as_tibble() %>%
  ggplot(aes(PC1, mn_logLen)) +
  geom_point(size = 2) +
  geom_smooth(method = "lm") +
  labs(
    x = paste0("PC1 (", percent(summary(pcaCqn_2)$importance["Proportion of Variance","PC1"]),")"),
    y = "Mean log(Length)"
  )
d_2 <- pcaCqn_2$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(sum_2) %>%
  as_tibble() %>%
  ggplot(aes(PC1, mn_gc)) +
  geom_point(size = 2) +
  geom_smooth(method = "lm") +
  scale_y_continuous(labels = percent) +
  labs(
    x = paste0("PC1 (", percent(summary(pcaCqn_2)$importance["Proportion of Variance","PC1"]),")"),
    y = "Mean GC"
  )
ggarrange(a_2, b_2, c_2, d_2, ncol = 2, nrow = 2) %>%
  annotate_figure("E-GEOD-72322")
```

```{r corrplotCqn_2, fig.cap="*Correlations between the first three principal components and measured variables: mean GC content, mean log(length) and rRNA proportion.*"}
corrplot(
  pcaCorCqn_2,
  type = "lower", 
  diag = FALSE, 
  addCoef.col = 1, addCoefasPercent = TRUE
)
```

# Differential abundance

```{r design}
design_1 <- model.matrix(~proportion, data = dgeFilt_1$samples)
design_2 <- model.matrix(~proportion, data = dgeFilt_2$samples)
design_3 <- model.matrix(~proportion, data = dgeFilt_3$samples)
design_4 <- model.matrix(~proportion, data = dgeFilt_4$samples)
design_5 <- model.matrix(~proportion, data = dgeFilt_5$samples)
```

```{r voom}
voom_1 <- voom(dgeFilt_1, design = design_1)
voom_2 <- voom(dgeFilt_2, design = design_2)
voom_3 <- voom(dgeFilt_3, design = design_3)
voom_4 <- voom(dgeFilt_4, design = design_4)
voom_5 <- voom(dgeFilt_5, design = design_5)
```

```{r fit}
fit_1 <- lmFit(voom_1, design = design_1)
fit_2 <- lmFit(voom_2, design = design_2)
fit_3 <- lmFit(voom_3, design = design_3)
fit_4 <- lmFit(voom_4, design = design_4)
fit_5 <- lmFit(voom_5, design = design_5)
```

```{r eBayes}
eBayes_1 <- eBayes(fit_1)
eBayes_2 <- eBayes(fit_2)
eBayes_3 <- eBayes(fit_3)
eBayes_4 <- eBayes(fit_4)
eBayes_5 <- eBayes(fit_5)
```

```{r topTable}
topTable_1 <- eBayes_1 %>%
  topTable(coef = colnames(design_1)[2], sort.by = "p", n = Inf) %>%
  set_colnames(str_remove(colnames(.), "ID\\.")) %>%
  mutate(DE = adj.P.Val < 0.05) %>%
  unite(Location, c(seqnames, start, end, width, strand), sep = ":") %>%
  dplyr::select(
    Geneid = gene_id,
    Symbol = gene_name,
    AveExpr,
    logFC,
    P.Value,
    FDR = adj.P.Val,
    Location,
    t,
    DE,
    everything(),
    -B
  ) %>%
  as_tibble()
topTable_2 <- eBayes_2 %>%
  topTable(coef = colnames(design_2)[2], sort.by = "p", n = Inf) %>%
  set_colnames(str_remove(colnames(.), "ID\\.")) %>%
  mutate(DE = adj.P.Val < 0.05) %>%
  unite(Location, c(seqnames, start, end, width, strand), sep = ":") %>%
  dplyr::select(
    Geneid = gene_id,
    Symbol = gene_name,
    AveExpr,
    logFC,
    P.Value,
    FDR = adj.P.Val,
    Location,
    t,
    DE,
    everything(),
    -B
  ) %>%
  as_tibble()
topTable_3 <- eBayes_3 %>%
  topTable(coef = colnames(design_3)[2], sort.by = "p", n = Inf) %>%
  set_colnames(str_remove(colnames(.), "ID\\.")) %>%
  mutate(DE = adj.P.Val < 0.05) %>%
  unite(Location, c(seqnames, start, end, width, strand), sep = ":") %>%
  dplyr::select(
    Geneid = gene_id,
    Symbol = gene_name,
    AveExpr,
    logFC,
    P.Value,
    FDR = adj.P.Val,
    Location,
    t,
    DE,
    everything(),
    -B
  ) %>%
  as_tibble()
topTable_4 <- eBayes_4 %>%
  topTable(coef = colnames(design_4)[2], sort.by = "p", n = Inf) %>%
  set_colnames(str_remove(colnames(.), "ID\\.")) %>%
  mutate(DE = adj.P.Val < 0.05) %>%
  unite(Location, c(seqnames, start, end, width, strand), sep = ":") %>%
  dplyr::select(
    Geneid = gene_id,
    Symbol = gene_name,
    AveExpr,
    logFC,
    P.Value,
    FDR = adj.P.Val,
    Location,
    t,
    DE,
    everything(),
    -B
  ) %>%
  as_tibble()
topTable_5 <- eBayes_5 %>%
  topTable(coef = colnames(design_5)[2], sort.by = "p", n = Inf) %>%
  set_colnames(str_remove(colnames(.), "ID\\.")) %>%
  mutate(DE = adj.P.Val < 0.05) %>%
  unite(Location, c(seqnames, start, end, width, strand), sep = ":") %>%
  dplyr::select(
    Geneid = gene_id,
    Symbol = gene_name,
    AveExpr,
    logFC,
    P.Value,
    FDR = adj.P.Val,
    Location,
    t,
    DE,
    everything(),
    -B
  ) %>%
  as_tibble()
```

### Dataset 1 (*D. rerio*)

```{r difResults_1}
topTable_1 %>% 
  dplyr::select(Geneid, Symbol, AveExpr, logFC, P.Value, FDR, DE) %>%
  mutate(
    AveExpr = format(round(AveExpr, 2), nsmall = 2),
    logFC = format(round(logFC, 2), nsmall = 2),
    P.Value = sprintf("%.2e", P.Value),
    FDR = sprintf("%.2e", FDR)
  ) %>%
  dplyr::slice(1:200) %>%
  datatable(options = list(pageLength = 20), class = "striped hover condensed responsive", filter = "top")
```

### Dataset 2 (*D. rerio*)

```{r difResults_2}
topTable_2 %>% 
  dplyr::select(Geneid, Symbol, AveExpr, logFC, P.Value, FDR, DE) %>%
  mutate(
    AveExpr = format(round(AveExpr, 2), nsmall = 2),
    logFC = format(round(logFC, 2), nsmall = 2),
    P.Value = sprintf("%.2e", P.Value),
    FDR = sprintf("%.2e", FDR)
  ) %>%
  dplyr::slice(1:200) %>%
  datatable(options = list(pageLength = 20), class = "striped hover condensed responsive", filter = "top")
```

### Dataset 3 (*D. rerio*)

```{r difResults_3}
topTable_3 %>% 
  dplyr::select(Geneid, Symbol, AveExpr, logFC, P.Value, FDR, DE) %>%
  mutate(
    AveExpr = format(round(AveExpr, 2), nsmall = 2),
    logFC = format(round(logFC, 2), nsmall = 2),
    P.Value = sprintf("%.2e", P.Value),
    FDR = sprintf("%.2e", FDR)
  ) %>%
  dplyr::slice(1:200) %>%
  datatable(options = list(pageLength = 20), class = "striped hover condensed responsive", filter = "top")
```

### Dataset 4 (*M. musculus*)

```{r difResults_4}
topTable_4 %>% 
  dplyr::select(Geneid, Symbol, AveExpr, logFC, P.Value, FDR, DE) %>%
  mutate(
    AveExpr = format(round(AveExpr, 2), nsmall = 2),
    logFC = format(round(logFC, 2), nsmall = 2),
    P.Value = sprintf("%.2e", P.Value),
    FDR = sprintf("%.2e", FDR)
  ) %>%
  dplyr::slice(1:200) %>%
  datatable(options = list(pageLength = 20), class = "striped hover condensed responsive", filter = "top")
```

### Dataset 5 (*M. musculus*)

```{r difResults_5}
topTable_5 %>% 
  dplyr::select(Geneid, Symbol, AveExpr, logFC, P.Value, FDR, DE) %>%
  mutate(
    AveExpr = format(round(AveExpr, 2), nsmall = 2),
    logFC = format(round(logFC, 2), nsmall = 2),
    P.Value = sprintf("%.2e", P.Value),
    FDR = sprintf("%.2e", FDR)
  ) %>%
  dplyr::slice(1:200) %>%
  datatable(options = list(pageLength = 20), class = "striped hover condensed responsive", filter = "top")
```

# Transcript counts

```{r catchSalmon}
dirs_3 <- list.dirs(
  "~/Documents/Bioinformatics/Projects/20170327_Psen2S4Ter_RNASeq/data/5_salmon/quant/", 
  recursive = FALSE
)
salmon_3 <- catchSalmon(paths = dirs_3)
```

```{r counts_tx}
dgeListTx_3 <- salmon_3$counts %>%
  set_colnames(basename(colnames(.))) %>%
  set_colnames(str_remove(colnames(.), "_6month_07_07_2016_F3")) %>%
  set_colnames(str_remove(colnames(.), "Ps2Ex3M1_")) %>%
  as.data.frame() %>%
  rownames_to_column("tx_id") %>%
  mutate(tx_id = str_remove(tx_id, "\\.[0-9]+$")) %>%
  column_to_rownames("tx_id") %>%
  DGEList() %>%
  calcNormFactors()
```

```{r dgeFilt_tx}
genes2keep_3 <- dgeListTx_3 %>%
  cpm() %>%
  is_greater_than(1) %>%
  rowSums() %>%
  is_weakly_greater_than(6)
dgeFiltTx_3 <- dgeListTx_3[genes2keep_3,, keep.lib.sizes = FALSE] %>%
  calcNormFactors()
```

# k-mer analysis 

```{r topIds}
topIds <- topTable_3 %>%
  head(n = 2000) %>%
  .$Geneid
```

```{r conIds}
conIds <- topTable_3 %>%
  tail(n = 2000) %>%
  .$Geneid
```

```{r}
loadMer <- function(x){
  read.table(
    paste0(
      "/data/biohub/20170327_Psen2S4Ter_RNASeq/data/6_jellyfish/", 
      x, 
      "_dumps.txt"
    ),
    col.names = c("mer", "count")
  ) %>%
    dplyr::arrange(-count) %>%
    as_tibble()
}
```

```{r}
dna <- ensembldb::getGenomeTwoBitFile(ensDb_Dr)
```

## Test genes

```{r}
# topGenes <- genes(ensDb_Dr) %>%
#   .[topIds]
# geneStringSet <- getSeq(dna, topGenes)
# writeXStringSet(
#   geneStringSet,
#   "/data/biohub/20170327_Psen2S4Ter_RNASeq/data/6_jellyfish/test2000.fa"
# )
```

```{r}
testMer500k6 <- loadMer("test500k6")
testMer500k7 <- loadMer("test500k7")
testMer500k8 <- loadMer("test500k8")
testMer500k9 <- loadMer("test500k9")
testMer500k10 <- loadMer("test500k10")
testMer500k15 <- loadMer("test500k15")
# testMer500k21 <- loadMer("test500k21")
```

```{r}
testMer1000k6 <- loadMer("test1000k6")
testMer1000k7 <- loadMer("test1000k7")
testMer1000k8 <- loadMer("test1000k8")
testMer1000k9 <- loadMer("test1000k9")
# testMer1000k10 <- loadMer("test1000k10")
# testMer1000k15 <- loadMer("test1000k15")
# testMer1000k21 <- loadMer("test1000k21")
```

## Control genes

```{r}
# conGenes <- genes(ensDb_Dr) %>%
#   .[conIds]
# conGeneStringSet <- getSeq(dna, conGenes)
# writeXStringSet(
#   conGeneStringSet,
#   "/data/biohub/20170327_Psen2S4Ter_RNASeq/data/6_jellyfish/control2000.fa"
# )
```

```{r}
conMer500k6 <- loadMer("control500k6")
conMer500k7 <- loadMer("control500k7")
conMer500k8 <- loadMer("control500k8")
conMer500k9 <- loadMer("control500k9")
conMer500k10 <- loadMer("control500k10")
conMer500k15 <- loadMer("control500k15")
# conMer500k21 <- loadMer("control500k21")
```

```{r}
conMer1000k6 <- loadMer("control1000k6")
conMer1000k7 <- loadMer("control1000k7")
conMer1000k8 <- loadMer("control1000k8")
conMer1000k9 <- loadMer("control1000k9")
# conMer1000k10 <- loadMer("control1000k10")
# conMer1000k15 <- loadMer("control1000k15")
# conMer1000k21 <- loadMer("control1000k21")
```

## Enrichment testing

```{r}
merRatio <- function(x, y) {
  x %>% 
    inner_join(y, by = "mer") %>% 
    mutate(
      p.x = count.x / sum(count.x), 
      p.y = count.y / sum(count.y), 
      ratio = p.x / p.y
    ) %>% 
    arrange(desc(ratio))
}
```

### Test vs control, 500 genes

```{r}
merRat500k6 <- merRatio(testMer500k6, conMer500k6)
merRat500k7 <- merRatio(testMer500k7, conMer500k7)
merRat500k8 <- merRatio(testMer500k8, conMer500k8)
merRat500k9 <- merRatio(testMer500k9, conMer500k9)
merRat500k10 <- merRatio(testMer500k10, conMer500k10)
merRat500k15 <- merRatio(testMer500k15, conMer500k15)
```

```{r}
plotRat500k6 <- merRat500k6 %>%
  ggplot(aes(-log10(p.x), -log10(p.y))) +
  geom_point() +
  labs(x = "test", y = "control", title = "k = 6") +
  geom_abline(slope = 1, intercept = 0, col = "blue")
plotRat500k7 <- merRat500k7 %>%
  ggplot(aes(-log10(p.x), -log10(p.y))) +
  geom_point() +
  labs(x = "test", y = "control", title = "k = 7") +
  geom_abline(slope = 1, intercept = 0, col = "blue")
plotRat500k8 <- merRat500k8 %>%
  ggplot(aes(-log10(p.x), -log10(p.y))) +
  geom_point() +
  labs(x = "test", y = "control", title = "k = 8") +
  geom_abline(slope = 1, intercept = 0, col = "blue")
plotRat500k9 <- merRat500k9 %>%
  ggplot(aes(-log10(p.x), -log10(p.y))) +
  geom_point() +
  labs(x = "test", y = "control", title = "k = 9") +
  geom_abline(slope = 1, intercept = 0, col = "blue")
plotRat500k10 <- merRat500k10 %>%
  ggplot(aes(-log10(p.x), -log10(p.y))) +
  geom_point() +
  labs(x = "test", y = "control", title = "k = 10") +
  geom_abline(slope = 1, intercept = 0, col = "blue")
plotRat500k15 <- merRat500k15 %>%
  ggplot(aes(-log10(p.x), -log10(p.y))) +
  geom_point() +
  labs(x = "test", y = "control", title = "k = 15") +
  geom_abline(slope = 1, intercept = 0, col = "blue")
plotRat500 <- ggarrange(plotRat500k6, plotRat500k7, plotRat500k8, plotRat500k9,
                        plotRat500k10, plotRat500k15, nrow = 3, ncol = 3)
annotate_figure(
  plotRat500, 
  bottom = "Log ratio of top 500 test gene kmer counts vs 500 control gene kmer counts"
)
```

```{r}
merP500k6 <- merRat500k6 %>%
  dplyr::slice(1100:1200) %>%
  mutate(
    n = sum(testMer500k6$count),
    binom.test = pmap(list(count.x, n, p.y), binom.test),
    p.val = vapply(binom.test, function(x){x[["p.value"]]}, numeric(1)),
    p.adj = p.adjust(p.val, "bonf")
  ) %>%
dplyr::arrange(p.val)
# saveRDS(merP500k6, paste0(here::here(), "/R/merP500k6.rds"))
# merP500k7 <- merRat500k7 %>%
#   mutate(
#     n = sum(testMer500k7$count),
#     binom.test = pmap(list(count.x, n, p.y), binom.test),
#     p.val = vapply(binom.test, function(x){x[["p.value"]]}, numeric(1)),
#     p.adj = p.adjust(p.val, "bonf")
#   ) %>%
# dplyr::arrange(p.val)
# saveRDS(merP500k7, paste0(here::here(), "/R/merP500k7.rds"))
# merP500k8 <- merRat500k8 %>%
#   mutate(
#     n = sum(testMer500k8$count),
#     binom.test = pmap(list(count.x, n, p.y), binom.test),
#     p.val = vapply(binom.test, function(x){x[["p.value"]]}, numeric(1)),
#     p.adj = p.adjust(p.val, "bonf")
#   ) %>%
# dplyr::arrange(p.val)
# saveRDS(merP500k8, paste0(here::here(), "/R/merP500k8.rds"))
# merP500k9 <- merRat500k9 %>%
#   mutate(
#     n = sum(testMer500k9$count),
#     binom.test = pmap(list(count.x, n, p.y), binom.test),
#     p.val = vapply(binom.test, function(x){x[["p.value"]]}, numeric(1)),
#     p.adj = p.adjust(p.val, "bonf")
#   ) %>%
# dplyr::arrange(p.val)
# saveRDS(merP500k9, paste0(here::here(), "/R/merP500k9.rds"))
```

```{r}
merP500k6 <- readRDS(paste0(here::here(), "/R/merP500k6.rds"))
merP500k7 <- readRDS(paste0(here::here(), "/R/merP500k7.rds"))
merP500k8 <- readRDS(paste0(here::here(), "/R/merP500k8.rds"))
```

### Test vs control, 1000 genes

```{r}
merRat1000k6 <- merRatio(testMer1000k6, conMer1000k6)
merRat1000k7 <- merRatio(testMer1000k7, conMer1000k7)
merRat1000k8 <- merRatio(testMer1000k8, conMer1000k8)
merRat1000k9 <- merRatio(testMer1000k9, conMer1000k9)
```

```{r}
plotRat1000k6 <- merRat1000k6 %>%
  ggplot(aes(-log10(p.x), -log10(p.y))) +
  geom_point() +
  labs(x = "test", y = "control", title = "k = 6") +
  geom_abline(slope = 1, intercept = 0, col = "blue")
plotRat1000k7 <- merRat1000k7 %>%
  ggplot(aes(-log10(p.x), -log10(p.y))) +
  geom_point() +
  labs(x = "test", y = "control", title = "k = 7") +
  geom_abline(slope = 1, intercept = 0, col = "blue")
plotRat1000k8 <- merRat1000k8 %>%
  ggplot(aes(-log10(p.x), -log10(p.y))) +
  geom_point() +
  labs(x = "test", y = "control", title = "k = 8") +
  geom_abline(slope = 1, intercept = 0, col = "blue")
plotRat1000k9 <- merRat1000k9 %>%
  ggplot(aes(-log10(p.x), -log10(p.y))) +
  geom_point() +
  labs(x = "test", y = "control", title = "k = 9") +
  geom_abline(slope = 1, intercept = 0, col = "blue")
plotRat1000 <- ggarrange(plotRat1000k6, plotRat1000k7, plotRat1000k8, plotRat1000k9, 
          nrow = 2, ncol = 2)
annotate_figure(
  plotRat1000, 
  bottom = "Log ratio of top 1000 test gene kmer counts vs 1000 control gene kmer counts"
)
```

```{r}
merP500k8 %>%
  mutate(sig = p.adj < 0.05 & ratio > 1) %>%
  ggplot(aes(-log10(p.x), -log10(p.y))) +
  geom_point(aes(colour = sig)) +
  labs(x = "test", y = "control", title = "k = 8") +
  geom_abline(slope = 1, intercept = 0, col = "blue") +
  scale_colour_manual(values = c("grey", "red"))
```

```{r}
# merP1000k6 <- merRat1000k6 %>%
#   mutate(
#     n = sum(testMer1000k6$count),
#     binom.test = pmap(list(count.x, n, p.y), binom.test),
#     p.val = vapply(binom.test, function(x){x[["p.value"]]}, numeric(1)),
#     p.val = sprintf("%.2e", p.val),
#     p.adj = p.adjust(p.val, "bonf")
#   ) %>%
# dplyr::arrange(p.val)
```

### 500 genes vs 1000 genes, test

```{r}
merRatTestk6 <- merRatio(testMer500k6, testMer1000k6)
merRatTestk7 <- merRatio(testMer500k7, testMer1000k7)
merRatTestk8 <- merRatio(testMer500k8, testMer1000k8)
merRatTestk9 <- merRatio(testMer500k9, testMer1000k9)
```

```{r}
plotRatTestk6 <- merRatTestk6 %>%
  ggplot(aes(-log10(p.x), -log10(p.y))) +
  geom_point() +
  labs(x = "500 test genes", y = "1000 test genes", title = "k = 6") +
  geom_abline(slope = 1, intercept = 0, col = "blue")
plotRatTestk7 <- merRatTestk7 %>%
  ggplot(aes(-log10(p.x), -log10(p.y))) +
  geom_point() +
  labs(x = "500 test genes", y = "1000 test genes", title = "k = 7") +
  geom_abline(slope = 1, intercept = 0, col = "blue")
plotRatTestk8 <- merRatTestk8 %>%
  ggplot(aes(-log10(p.x), -log10(p.y))) +
  geom_point() +
  labs(x = "500 test genes", y = "1000 test genes", title = "k = 8") +
  geom_abline(slope = 1, intercept = 0, col = "blue")
plotRatTestk9 <- merRatTestk9 %>%
  ggplot(aes(-log10(p.x), -log10(p.y))) +
  geom_point() +
  labs(x = "500 test genes", y = "1000 test genes", title = "k = 9") +
  geom_abline(slope = 1, intercept = 0, col = "blue")
plotRatTest <- ggarrange(plotRatTestk6, plotRatTestk7, plotRatTestk8, plotRatTestk9, 
          nrow = 2, ncol = 2)
annotate_figure(
  plotRatTest, 
  bottom = "Log ratio of top 500 test gene kmer counts vs top 1000 gene kmer counts"
)
```

### 500 genes vs 1000 genes, control

```{r}
merRatConk6 <- merRatio(conMer500k6, conMer1000k6)
merRatConk7 <- merRatio(conMer500k7, conMer1000k7)
merRatConk8 <- merRatio(conMer500k8, conMer1000k8)
merRatConk9 <- merRatio(conMer500k9, conMer1000k9)
```

```{r}
plotRatConk6 <- merRatConk6 %>%
  ggplot(aes(-log10(p.x), -log10(p.y))) +
  geom_point() +
  labs(x = "500 control genes", y = "1000 control genes", title = "k = 6") +
  geom_abline(slope = 1, intercept = 0, col = "blue")
plotRatConk7 <- merRatConk7 %>%
  ggplot(aes(-log10(p.x), -log10(p.y))) +
  geom_point() +
  labs(x = "500 control genes", y = "1000 control genes", title = "k = 7") +
  geom_abline(slope = 1, intercept = 0, col = "blue")
plotRatConk8 <- merRatConk8 %>%
  ggplot(aes(-log10(p.x), -log10(p.y))) +
  geom_point() +
  labs(x = "500 control genes", y = "1000 control genes", title = "k = 8") +
  geom_abline(slope = 1, intercept = 0, col = "blue")
plotRatConk9 <- merRatConk9 %>%
  ggplot(aes(-log10(p.x), -log10(p.y))) +
  geom_point() +
  labs(x = "500 control genes", y = "1000 control genes", title = "k = 9") +
  geom_abline(slope = 1, intercept = 0, col = "blue")
plotRatCon <- ggarrange(plotRatConk6, plotRatConk7, plotRatConk8, plotRatConk9, 
          nrow = 2, ncol = 2)
annotate_figure(
  plotRatCon, 
  bottom = "Log ratio of 500 control gene kmer counts vs 1000 control kmer counts"
)
```

## Transcript-level

```{r}
# topTx <- transcriptsBy(ensDb_Dr, by = "gene") %>%
#   .[topIds] %>%
#   unlist() %>%
#   subset(tx_id %in% rownames(dgeFiltTx_3)) %>%
#   mcols() %>%
#   .$tx_id
# topExons <- exonsBy(ensDb_Dr, by = "tx") %>%
#   .[topTx]
# txStrings <- lapply(topExons, function(x){
#   getSeq(dna, x) %>%
#     unlist()
# }
# )
# txStringSet <- DNAStringSet(txStrings)
# writeXStringSet(
#   txStringSet,
#   "/data/biohub/20170327_Psen2S4Ter_RNASeq/data/6_jellyfish/top100Tx.fa"
# )
```

```{r}
# conTx <- transcriptsBy(ensDb_Dr, by = "gene") %>%
#   .[conIds] %>%
#   unlist() %>%
#   subset(tx_id %in% rownames(dgeFiltTx_3)) %>%
#   mcols() %>%
#   .$tx_id
# conExons <- exonsBy(ensDb_Dr, by = "tx") %>%
#   .[conTx]
# conTxStrings <- lapply(conExons, function(x){
#   getSeq(dna, x) %>%
#     unlist()
# }
# )
# conTxStringSet <- DNAStringSet(conTxStrings)
# writeXStringSet(
#   conTxStringSet,
#   "/data/biohub/20170327_Psen2S4Ter_RNASeq/data/6_jellyfish/con100Tx.fa"
# )
```

```{r}
# mer_tx100 <- read.table(
#   "/data/biohub/20170327_Psen2S4Ter_RNASeq/data/6_jellyfish/top100Tx_dumps.txt",
#   col.names = c("mer", "count")
# ) %>%
#   dplyr::arrange(-count) %>%
#   as_tibble()
# mer_conTx100 <- read.table(
#   "/data/biohub/20170327_Psen2S4Ter_RNASeq/data/6_jellyfish/con100Tx_dumps.txt",
#   col.names = c("mer", "count")
# ) %>%
#   dplyr::arrange(-count) %>%
#   as_tibble()
# mer_tx100 %>% 
#   inner_join(mer_conTx100, by = "mer") %>% 
#   mutate(
#     p.x = count.x / sum(count.x), 
#     p.y = count.y / sum(count.y), 
#     ratio = log2(p.x / p.y)
#   ) %>% 
#   arrange(desc(ratio))
```

# Coverage patterns

## Setup

```{r GvizOpts}
options(ucscChromosomeNames = FALSE)
```

```{r exons}
ah_Dr <- AnnotationHub() %>%
  subset(species == "Danio rerio") %>%
  subset(rdataclass == "EnsDb")
ensDb_Dr <- ah_Dr[["AH74989"]]
exons_Dr <- exonsBy(
  ensDb_Dr,
  by = "tx",
  columns = c("exon_id", "gene_id", "tx_id", "gene_name"),
  filter = AnnotationFilterList(
    SeqNameFilter(c(1:25)),
    GeneIdFilter("ENSD", "startsWith")
  )
) %>% 
  unlist()
colnames(mcols(exons_Dr)) <- str_remove_all(colnames(mcols(exons_Dr)), "_id") %>%
  str_replace_all("tx", "transcript")
mcols(exons_Dr) <- mcols(exons_Dr)[c("gene_name", "gene", "exon", "transcript")]
```

```{r genesFilt}
genesFilt_Dr <- genes(
  ensDb_Dr,
  filter = AnnotationFilterList(
    SeqNameFilter(c(1:25)),
    GeneIdFilter("ENSD", "startsWith") 
  )
)
```

```{r orders}
order_3 <- dgeList_3$samples %>%
  dplyr::arrange(filenames) %>%
  as_tibble() %>% 
  rownames_to_column("order") %>% 
  dplyr::arrange(-proportion)
```

## Dataset 3 (*D. rerio*)

### *rps9*

```{r}
## Gene region & genome axis tracks
sym <- "rps9"
goi <- genesFilt_Dr[mcols(genesFilt_Dr)$gene_name == sym,]$gene_id
counts <- dgeList_3$counts[goi,]
gm <- subset(exons_Dr, gene == goi) %>%
  subset(transcript %in% rownames(dgeFiltTx_3))
gmRed <- gm %>% 
  GenomicRanges::reduce()
st <- GenomicRanges::start(gmRed) %>% 
  min()
en <- GenomicRanges::end(gmRed) %>% 
  max()
chr <- seqnames(gmRed)@values
str <- gmRed %>% 
  strand() %>%
  .@values
gTrack <- GenomeAxisTrack()
rrTrack <- GeneRegionTrack(
  gmRed,
  name = NULL,
  fill = "darkorange"
)
rTrack <- GeneRegionTrack(
  gm,
  name = NULL
)
```

```{r}
## GC content track
win <- slidingWindows(gmRed, width = 1) %>% 
  unlist()
start <- win %>%
  start()
end <- win %>%
  end()
seq <- getSeq(dna, win)
gc <- seq %>%
  letterFrequency(letters = "GC", OR = "") %>%
  .[,1]
tbl <- tibble(
  start = start,
  end = end,
  gc = gc
) %>% 
  mutate(
    seqnames = chr,
    strand = str,
    cumsum = cumsum(gc),
    cummean = cummean(gc),
    lag = dplyr::lag(cumsum, n = 50, default = NA),
    lead = dplyr::lead(cumsum, n = 50, default = NA),
    gcsum = lead - lag,
    gcprop = gcsum / 101
  ) 
gcTrack <- tbl %>% 
  dplyr::select(seqnames, start, end, strand, gcprop) %>%
  makeGRangesFromDataFrame(
    keep.extra.columns = TRUE
  ) %>%
  DataTrack(
    type = c("b"),
    ylim = c(0, 1), 
    cex = 0.2, 
    name = "GC %",
    col.title="black",
    fontface.title=2,
    fontsize=10,
    grid = TRUE,
    lwd.grid = 0.5,
    lty.grid = "dashed",
    v = 0
  )
```

```{r}
## Coverage tracks
files <- list.files(
  "/data/biohub/20170327_Psen2S4Ter_RNASeq/data/2_alignedData/bam", 
  pattern = ".bam$", 
  full.names = TRUE
)
param <- ScanBamParam(
  what = c("pos", "qwidth"),
  which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
  flag=scanBamFlag(isUnmappedQuery=FALSE)
)
covFunc <- function(x){
  prop <- order_3$proportion[order_3$filenames == x] %>%
    percent(accuracy = 0.01)
  reads <- counts[order_3$sample[order_3$filenames == x]]
  pileup(x, scanBamParam = param) %>%
    dplyr::filter(strand == str) %>%
    mutate(start = pos, end = pos, count = count/reads) %>% 
    dplyr::select(seqnames, start, end, strand, count) %>%
    makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>%
    DataTrack(
      type = c("heatmap"),
      cex = 0.2, 
      name = prop,
      col.title="black",
      fontface.title=2,
      fontsize=10,
    )
}
covList <- lapply(files, covFunc) 
covOrd <- covList[as.integer(order_3$order)]
```

```{r}
## k-mer track
geneMod <- subset(genesGR_Dr, gene_id == goi)
geneSeq <- getSeq(dna, geneMod)
merTop <- merP500k8$mer[1:50]
# merLoc <- vmatchPattern(merTop[25], geneSeq)
merLoc <- lapply(merTop, function(x){
  vmatchPattern(x, geneSeq) %>%
    unlist()
})
merStart <- rapply(merLoc, function(x){
  start(x)
})
merWidth <- rapply(merLoc, function(x){
  width(x)
})
hTrack <- HighlightTrack(
  trackList = gcTrack, 
  start = start(geneMod) + merStart,
  width = merWidth,
  chromosome = chr,
  inBackground = TRUE,
  col = "white",
  fill = "red"
)
```

```{r, fig.height=9}
## Plot the tracks
plotTracks(
  c(gTrack, rrTrack, rTrack, covOrd, hTrack),
  from = st - 100,
  to = en + 100,
  chromosome = chr,
  sizes = c(1,0.5,1,rep(0.5, length(covOrd)),1), 
  main = sym,
  cex.main = 1.5,
  fontface.main = 1
)
```

#### Zoomed-in view

```{r, fig.height=9}
plotTracks(
  c(gTrack, rrTrack, rTrack, covOrd, gcTrack),
  from = 31923000,
  to = 31923150,
  chromosome = chr,
  sizes = c(1,0.5,1,rep(0.5, length(covOrd)),1), 
  main = "exon 2",
  cex.main = 1.5,
  fontface.main = 1
)
```

```{r, fig.height=9}
plotTracks(
  c(gTrack, rrTrack, rTrack, covOrd, gcTrack),
  from = 31923300,
  to = 31923520,
  chromosome = chr,
  sizes = c(1,0.5,1,rep(0.5, length(covOrd)),1), 
  main = "exon 3",
  cex.main = 1.5,
  fontface.main = 1
)
```

```{r, fig.height=9}
plotTracks(
  c(gTrack, rrTrack, rTrack, covOrd, gcTrack),
  from = 31924360,
  to = 31924600,
  chromosome = chr,
  sizes = c(1,0.5,1,rep(0.5, length(covOrd)),1), 
  main = "exon 4",
  cex.main = 1.5,
  fontface.main = 1
)
```

```{r, fig.height=9}
# alFunc <- function(x){
#   prop <- order_3$proportion[order_3$filenames == x] %>%
#     percent(accuracy = 0.01)
#   count <- dgeList_3$counts[goi,] %>%
#     .[order_3$sample[order_3$filenames == x]] %>%
#     unname()
#   AlignmentsTrack(
#     range = x, 
#     isPaired = TRUE,
#     name = prop,
#     col.title="black",
#     fontface.title=2,
#     fontsize=10,
#     transformation = function(x) {(x/counts)*1000},
#     showAxis = FALSE,
#     type = "coverage"
#   )
# }
# alList <- lapply(files, alFunc) 
# ordList <- alList[as.integer(order_3$order)]
# hList <- HighlightTrack(
#   ordList, 
#   start = c(31923289, 31924367),
#   end = c(31923524, 31924591),
#   chromosome = chr)
```

```{r, fig.height=3}
# lTrack <- order_3 %>%
#   tail(n = 1) %>% 
#   .$filenames %>%
#   AlignmentsTrack(alpha = 0.85, col.coverage = "blue")
# hTrack <- order_3 %>%
#   head(n = 1) %>% 
#   .$filenames %>%
#   AlignmentsTrack(alpha = 0.85, col.coverage = "red")
# oTrack <- OverlayTrack(c(lTrack, hTrack))
```

***

### *rps2*

```{r}
sym <- "rps2"
goi <- genesFilt_Dr[mcols(genesFilt_Dr)$gene_name == sym,]$gene_id
counts <- dgeList_3$counts[goi,]
gm <- subset(exons_Dr, gene == goi) %>%
  subset(transcript %in% rownames(dgeFiltTx_3))
gmRed <- gm %>% 
  GenomicRanges::reduce()
st <- GenomicRanges::start(gmRed) %>% 
  min()
en <- GenomicRanges::end(gmRed) %>% 
  max()
chr <- seqnames(gmRed)@values
str <- gmRed %>% 
  strand() %>%
  .@values
gTrack <- GenomeAxisTrack()
rrTrack <- GeneRegionTrack(
  gmRed,
  name = NULL,
  fill = "darkorange"
)
rTrack <- GeneRegionTrack(
  gm,
  name = NULL
)
```

```{r}
## GC content track
win <- slidingWindows(gmRed, width = 1) %>% 
  unlist()
start <- win %>%
  start()
end <- win %>%
  end()
seq <- getSeq(dna, win)
gc <- seq %>%
  letterFrequency(letters = "GC", OR = "") %>%
  .[,1]
tbl <- tibble(
  start = start,
  end = end,
  gc = gc
) %>% 
  mutate(
    seqnames = chr,
    strand = str,
    cumsum = cumsum(gc),
    cummean = cummean(gc),
    lag = dplyr::lag(cumsum, n = 50, default = NA),
    lead = dplyr::lead(cumsum, n = 50, default = NA),
    gcsum = lead - lag,
    gcprop = gcsum / 101
  ) 
gcTrack <- tbl %>% 
  dplyr::select(seqnames, start, end, strand, gcprop) %>%
  makeGRangesFromDataFrame(
    keep.extra.columns = TRUE
  ) %>%
  DataTrack(
    type = c("b"),
    ylim = c(0, 1), 
    cex = 0.2, 
    name = "GC %",
    col.title="black",
    fontface.title=2,
    fontsize=10,
    grid = TRUE,
    lwd.grid = 0.5,
    lty.grid = "dashed",
    v = 0
  )
```

```{r}
## Coverage tracks
files <- list.files(
  "/data/biohub/20170327_Psen2S4Ter_RNASeq/data/2_alignedData/bam", 
  pattern = ".bam$", 
  full.names = TRUE
)
param <- ScanBamParam(
  what = c("pos", "qwidth"),
  which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
  flag=scanBamFlag(isUnmappedQuery=FALSE)
)
covFunc <- function(x){
  prop <- order_3$proportion[order_3$filenames == x] %>%
    percent(accuracy = 0.01)
  reads <- counts[order_3$sample[order_3$filenames == x]]
  pileup(x, scanBamParam = param) %>%
    dplyr::filter(strand == str) %>%
    mutate(start = pos, end = pos, count = count/reads) %>% 
    dplyr::select(seqnames, start, end, strand, count) %>%
    makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>%
    DataTrack(
      type = c("heatmap"),
      cex = 0.2, 
      name = prop,
      col.title="black",
      fontface.title=2,
      fontsize=10,
    )
}
covList <- lapply(files, covFunc) 
covOrd <- covList[as.integer(order_3$order)]
```

```{r}
## k-mer track
geneMod <- subset(genesGR_Dr, gene_id == goi)
geneSeq <- getSeq(dna, geneMod)
merTop <- merP500k8$mer[1:50]
# merLoc <- vmatchPattern(merTop[25], geneSeq)
merLoc <- lapply(merTop, function(x){
  vmatchPattern(x, geneSeq) %>%
    unlist()
})
merStart <- rapply(merLoc, function(x){
  start(x)
})
merWidth <- rapply(merLoc, function(x){
  width(x)
})
hTrack <- HighlightTrack(
  trackList = gcTrack, 
  start = start(geneMod) + merStart,
  width = merWidth,
  chromosome = chr,
  inBackground = TRUE,
  col = "white",
  fill = "red"
)
```

```{r, fig.height=9}
## Plot the tracks
plotTracks(
  c(gTrack, rrTrack, rTrack, covOrd, hTrack),
  from = st - 100,
  to = en + 100,
  chromosome = chr,
  sizes = c(1,0.5,1,rep(0.5, length(covOrd)),1), 
  main = sym,
  cex.main = 1.5,
  fontface.main = 1
)
```

***

### *rps13*

```{r}
sym <- "rps13"
goi <- genesFilt_Dr[mcols(genesFilt_Dr)$gene_name == sym,]$gene_id
counts <- dgeList_3$counts[goi,]
gm <- subset(exons_Dr, gene == goi) %>%
  subset(transcript %in% rownames(dgeFiltTx_3))
gmRed <- gm %>% 
  GenomicRanges::reduce()
st <- GenomicRanges::start(gmRed) %>% 
  min()
en <- GenomicRanges::end(gmRed) %>% 
  max()
chr <- seqnames(gmRed)@values
str <- gmRed %>% 
  strand() %>%
  .@values
gTrack <- GenomeAxisTrack()
rrTrack <- GeneRegionTrack(
  gmRed,
  name = NULL,
  fill = "darkorange"
)
rTrack <- GeneRegionTrack(
  gm,
  name = NULL
)
```

```{r}
## GC content track
win <- slidingWindows(gmRed, width = 1) %>% 
  unlist()
start <- win %>%
  start()
end <- win %>%
  end()
seq <- getSeq(dna, win)
gc <- seq %>%
  letterFrequency(letters = "GC", OR = "") %>%
  .[,1]
tbl <- tibble(
  start = start,
  end = end,
  gc = gc
) %>% 
  mutate(
    seqnames = chr,
    strand = str,
    cumsum = cumsum(gc),
    cummean = cummean(gc),
    lag = dplyr::lag(cumsum, n = 50, default = NA),
    lead = dplyr::lead(cumsum, n = 50, default = NA),
    gcsum = lead - lag,
    gcprop = gcsum / 101
  ) 
gcTrack <- tbl %>% 
  dplyr::select(seqnames, start, end, strand, gcprop) %>%
  makeGRangesFromDataFrame(
    keep.extra.columns = TRUE
  ) %>%
  DataTrack(
    type = c("b"),
    ylim = c(0, 1), 
    cex = 0.2, 
    name = "GC %",
    col.title="black",
    fontface.title=2,
    fontsize=10,
    grid = TRUE,
    lwd.grid = 0.5,
    lty.grid = "dashed",
    v = 0
  )
```

```{r}
## Coverage tracks
files <- list.files(
  "/data/biohub/20170327_Psen2S4Ter_RNASeq/data/2_alignedData/bam", 
  pattern = ".bam$", 
  full.names = TRUE
)
param <- ScanBamParam(
  what = c("pos", "qwidth"),
  which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
  flag=scanBamFlag(isUnmappedQuery=FALSE)
)
covFunc <- function(x){
  prop <- order_3$proportion[order_3$filenames == x] %>%
    percent(accuracy = 0.01)
  reads <- counts[order_3$sample[order_3$filenames == x]]
  pileup(x, scanBamParam = param) %>%
    dplyr::filter(strand == str) %>%
    mutate(start = pos, end = pos, count = count/reads) %>% 
    dplyr::select(seqnames, start, end, strand, count) %>%
    makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>%
    DataTrack(
      type = c("heatmap"),
      cex = 0.2, 
      name = prop,
      col.title="black",
      fontface.title=2,
      fontsize=10,
    )
}
covList <- lapply(files, covFunc) 
covOrd <- covList[as.integer(order_3$order)]
```

```{r}
## k-mer track
geneMod <- subset(genesGR_Dr, gene_id == goi)
geneSeq <- getSeq(dna, geneMod)
merTop <- merP500k7$mer[1:50]
# merLoc <- vmatchPattern(merTop[25], geneSeq)
merLoc <- lapply(merTop, function(x){
  vmatchPattern(x, geneSeq) %>%
    unlist()
})
merStart <- rapply(merLoc, function(x){
  start(x)
})
merWidth <- rapply(merLoc, function(x){
  width(x)
})
hTrack <- HighlightTrack(
  trackList = gcTrack, 
  start = start(geneMod) + merStart,
  width = merWidth,
  chromosome = chr,
  inBackground = TRUE,
  col = "white",
  fill = "red"
)
```

```{r, fig.height=9}
## Plot the tracks
plotTracks(
  c(gTrack, rrTrack, rTrack, covOrd, hTrack),
  from = st - 100,
  to = en + 100,
  chromosome = chr,
  sizes = c(1,0.5,1,rep(0.5, length(covOrd)),1), 
  main = sym,
  cex.main = 1.5,
  fontface.main = 1
)
```

***

### *rpl8*

```{r}
sym <- "rpl8"
goi <- genesFilt_Dr[mcols(genesFilt_Dr)$gene_name == sym,]$gene_id
counts <- dgeList_3$counts[goi,]
gm <- subset(exons_Dr, gene == goi) %>%
  subset(transcript %in% rownames(dgeFiltTx_3))
gmRed <- gm %>% 
  GenomicRanges::reduce()
st <- GenomicRanges::start(gmRed) %>% 
  min()
en <- GenomicRanges::end(gmRed) %>% 
  max()
chr <- seqnames(gmRed)@values
str <- gmRed %>% 
  strand() %>%
  .@values
gTrack <- GenomeAxisTrack()
rrTrack <- GeneRegionTrack(
  gmRed,
  name = NULL,
  fill = "darkorange"
)
rTrack <- GeneRegionTrack(
  gm,
  name = NULL
)
```

```{r}
## GC content track
win <- slidingWindows(gmRed, width = 1) %>% 
  unlist()
start <- win %>%
  start()
end <- win %>%
  end()
seq <- getSeq(dna, win)
gc <- seq %>%
  letterFrequency(letters = "GC", OR = "") %>%
  .[,1]
tbl <- tibble(
  start = start,
  end = end,
  gc = gc
) %>% 
  mutate(
    seqnames = chr,
    strand = str,
    cumsum = cumsum(gc),
    cummean = cummean(gc),
    lag = dplyr::lag(cumsum, n = 50, default = NA),
    lead = dplyr::lead(cumsum, n = 50, default = NA),
    gcsum = lead - lag,
    gcprop = gcsum / 101
  ) 
gcTrack <- tbl %>% 
  dplyr::select(seqnames, start, end, strand, gcprop) %>%
  makeGRangesFromDataFrame(
    keep.extra.columns = TRUE
  ) %>%
  DataTrack(
    type = c("b"),
    ylim = c(0, 1), 
    cex = 0.2, 
    name = "GC %",
    col.title="black",
    fontface.title=2,
    fontsize=10,
    grid = TRUE,
    lwd.grid = 0.5,
    lty.grid = "dashed",
    v = 0
  )
```

```{r}
## Coverage tracks
files <- list.files(
  "/data/biohub/20170327_Psen2S4Ter_RNASeq/data/2_alignedData/bam", 
  pattern = ".bam$", 
  full.names = TRUE
)
param <- ScanBamParam(
  what = c("pos", "qwidth"),
  which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
  flag=scanBamFlag(isUnmappedQuery=FALSE)
)
covFunc <- function(x){
  prop <- order_3$proportion[order_3$filenames == x] %>%
    percent(accuracy = 0.01)
  reads <- counts[order_3$sample[order_3$filenames == x]]
  pileup(x, scanBamParam = param) %>%
    dplyr::filter(strand == str) %>%
    mutate(start = pos, end = pos, count = count/reads) %>% 
    dplyr::select(seqnames, start, end, strand, count) %>%
    makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>%
    DataTrack(
      type = c("heatmap"),
      cex = 0.2, 
      name = prop,
      col.title="black",
      fontface.title=2,
      fontsize=10,
    )
}
covList <- lapply(files, covFunc) 
covOrd <- covList[as.integer(order_3$order)]
```

```{r}
## k-mer track
geneMod <- subset(genesGR_Dr, gene_id == goi)
geneSeq <- getSeq(dna, geneMod)
merTop <- merP500k8$mer[1:50]
# merLoc <- vmatchPattern(merTop[25], geneSeq)
merLoc <- lapply(merTop, function(x){
  vmatchPattern(x, geneSeq) %>%
    unlist()
})
merStart <- rapply(merLoc, function(x){
  start(x)
})
merWidth <- rapply(merLoc, function(x){
  width(x)
})
hTrack <- HighlightTrack(
  trackList = gcTrack, 
  start = start(geneMod) + merStart,
  width = merWidth,
  chromosome = chr,
  inBackground = TRUE,
  col = "white",
  fill = "red"
)
```

```{r, fig.height=9}
## Plot the tracks
plotTracks(
  c(gTrack, rrTrack, rTrack, covOrd, hTrack),
  from = st - 100,
  to = en + 100,
  chromosome = chr,
  sizes = c(1,0.5,1,rep(0.5, length(covOrd)),1), 
  main = sym,
  cex.main = 1.5,
  fontface.main = 1
)
```

***

### *ppiab*

```{r}
sym <- "ppiab"
goi <- genesFilt_Dr[mcols(genesFilt_Dr)$gene_name == sym,]$gene_id
counts <- dgeList_3$counts[goi,]
gm <- subset(exons_Dr, gene == goi) %>%
  subset(transcript %in% rownames(dgeFiltTx_3))
gmRed <- gm %>% 
  GenomicRanges::reduce()
st <- GenomicRanges::start(gmRed) %>% 
  min()
en <- GenomicRanges::end(gmRed) %>% 
  max()
chr <- seqnames(gmRed)@values
str <- gmRed %>% 
  strand() %>%
  .@values
gTrack <- GenomeAxisTrack()
rrTrack <- GeneRegionTrack(
  gmRed,
  name = NULL,
  fill = "darkorange"
)
rTrack <- GeneRegionTrack(
  gm,
  name = NULL
)
```

```{r}
## GC content track
win <- slidingWindows(gmRed, width = 1) %>% 
  unlist()
start <- win %>%
  start()
end <- win %>%
  end()
seq <- getSeq(dna, win)
gc <- seq %>%
  letterFrequency(letters = "GC", OR = "") %>%
  .[,1]
tbl <- tibble(
  start = start,
  end = end,
  gc = gc
) %>% 
  mutate(
    seqnames = chr,
    strand = str,
    cumsum = cumsum(gc),
    cummean = cummean(gc),
    lag = dplyr::lag(cumsum, n = 50, default = NA),
    lead = dplyr::lead(cumsum, n = 50, default = NA),
    gcsum = lead - lag,
    gcprop = gcsum / 101
  ) 
gcTrack <- tbl %>% 
  dplyr::select(seqnames, start, end, strand, gcprop) %>%
  makeGRangesFromDataFrame(
    keep.extra.columns = TRUE
  ) %>%
  DataTrack(
    type = c("b"),
    ylim = c(0, 1), 
    cex = 0.2, 
    name = "GC %",
    col.title="black",
    fontface.title=2,
    fontsize=10,
    grid = TRUE,
    lwd.grid = 0.5,
    lty.grid = "dashed",
    v = 0
  )
```

```{r}
## Coverage tracks
files <- list.files(
  "/data/biohub/20170327_Psen2S4Ter_RNASeq/data/2_alignedData/bam", 
  pattern = ".bam$", 
  full.names = TRUE
)
param <- ScanBamParam(
  what = c("pos", "qwidth"),
  which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
  flag=scanBamFlag(isUnmappedQuery=FALSE)
)
covFunc <- function(x){
  prop <- order_3$proportion[order_3$filenames == x] %>%
    percent(accuracy = 0.01)
  reads <- counts[order_3$sample[order_3$filenames == x]]
  pileup(x, scanBamParam = param) %>%
    dplyr::filter(strand == str) %>%
    mutate(start = pos, end = pos, count = count/reads) %>% 
    dplyr::select(seqnames, start, end, strand, count) %>%
    makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>%
    DataTrack(
      type = c("heatmap"),
      cex = 0.2, 
      name = prop,
      col.title="black",
      fontface.title=2,
      fontsize=10,
    )
}
covList <- lapply(files, covFunc) 
covOrd <- covList[as.integer(order_3$order)]
```

```{r}
## k-mer track
geneMod <- subset(genesGR_Dr, gene_id == goi)
geneSeq <- getSeq(dna, geneMod)
merTop <- merP500k8$mer[1:50]
# merLoc <- vmatchPattern(merTop[25], geneSeq)
merLoc <- lapply(merTop, function(x){
  vmatchPattern(x, geneSeq) %>%
    unlist()
})
merStart <- rapply(merLoc, function(x){
  start(x)
})
merWidth <- rapply(merLoc, function(x){
  width(x)
})
hTrack <- HighlightTrack(
  trackList = gcTrack, 
  start = start(geneMod) + merStart,
  width = merWidth,
  chromosome = chr,
  inBackground = TRUE,
  col = "white",
  fill = "red"
)
```

```{r, fig.height=9}
## Plot the tracks
plotTracks(
  c(gTrack, rrTrack, rTrack, covOrd, hTrack),
  from = st - 100,
  to = en + 100,
  chromosome = chr,
  sizes = c(1,0.5,1,rep(0.5, length(covOrd)),1), 
  main = sym,
  cex.main = 1.5,
  fontface.main = 1
)
```

***

### *h2afx1*

```{r}
sym <- "h2afx1"
goi <- genesFilt_Dr[mcols(genesFilt_Dr)$gene_name == sym,]$gene_id
counts <- dgeList_3$counts[goi,]
gm <- subset(exons_Dr, gene == goi) %>%
  subset(transcript %in% rownames(dgeFiltTx_3))
gmRed <- gm %>% 
  GenomicRanges::reduce()
st <- GenomicRanges::start(gmRed) %>% 
  min()
en <- GenomicRanges::end(gmRed) %>% 
  max()
chr <- seqnames(gmRed)@values
str <- gmRed %>% 
  strand() %>%
  .@values
gTrack <- GenomeAxisTrack()
rrTrack <- GeneRegionTrack(
  gmRed,
  name = NULL,
  fill = "darkorange"
)
rTrack <- GeneRegionTrack(
  gm,
  name = NULL
)
```

```{r}
## GC content track
win <- slidingWindows(gmRed, width = 1) %>% 
  unlist()
start <- win %>%
  start()
end <- win %>%
  end()
seq <- getSeq(dna, win)
gc <- seq %>%
  letterFrequency(letters = "GC", OR = "") %>%
  .[,1]
tbl <- tibble(
  start = start,
  end = end,
  gc = gc
) %>% 
  mutate(
    seqnames = chr,
    strand = str,
    cumsum = cumsum(gc),
    cummean = cummean(gc),
    lag = dplyr::lag(cumsum, n = 50, default = NA),
    lead = dplyr::lead(cumsum, n = 50, default = NA),
    gcsum = lead - lag,
    gcprop = gcsum / 101
  ) 
gcTrack <- tbl %>% 
  dplyr::select(seqnames, start, end, strand, gcprop) %>%
  makeGRangesFromDataFrame(
    keep.extra.columns = TRUE
  ) %>%
  DataTrack(
    type = c("b"),
    ylim = c(0, 1), 
    cex = 0.2, 
    name = "GC %",
    col.title="black",
    fontface.title=2,
    fontsize=10,
    grid = TRUE,
    lwd.grid = 0.5,
    lty.grid = "dashed",
    v = 0
  )
```

```{r}
## Coverage tracks
files <- list.files(
  "/data/biohub/20170327_Psen2S4Ter_RNASeq/data/2_alignedData/bam", 
  pattern = ".bam$", 
  full.names = TRUE
)
param <- ScanBamParam(
  what = c("pos", "qwidth"),
  which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
  flag=scanBamFlag(isUnmappedQuery=FALSE)
)
covFunc <- function(x){
  prop <- order_3$proportion[order_3$filenames == x] %>%
    percent(accuracy = 0.01)
  reads <- counts[order_3$sample[order_3$filenames == x]]
  pileup(x, scanBamParam = param) %>%
    dplyr::filter(strand == str) %>%
    mutate(start = pos, end = pos, count = count/reads) %>% 
    dplyr::select(seqnames, start, end, strand, count) %>%
    makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>%
    DataTrack(
      type = c("heatmap"),
      cex = 0.2, 
      name = prop,
      col.title="black",
      fontface.title=2,
      fontsize=10,
    )
}
covList <- lapply(files, covFunc) 
covOrd <- covList[as.integer(order_3$order)]
```

```{r}
## k-mer track
geneMod <- subset(genesGR_Dr, gene_id == goi)
geneSeq <- getSeq(dna, geneMod)
merTop <- merP500k8$mer[1:50]
# merLoc <- vmatchPattern(merTop[25], geneSeq)
merLoc <- lapply(merTop, function(x){
  vmatchPattern(x, geneSeq) %>%
    unlist()
})
merStart <- rapply(merLoc, function(x){
  start(x)
})
merWidth <- rapply(merLoc, function(x){
  width(x)
})
hTrack <- HighlightTrack(
  trackList = gcTrack, 
  start = start(geneMod) + merStart,
  width = merWidth,
  chromosome = chr,
  inBackground = TRUE,
  col = "white",
  fill = "red"
)
```

```{r, fig.height=9}
## Plot the tracks
plotTracks(
  c(gTrack, rrTrack, rTrack, covOrd, hTrack),
  from = st - 100,
  to = en + 100,
  chromosome = chr,
  sizes = c(1,0.5,1,rep(0.5, length(covOrd)),1), 
  main = sym,
  cex.main = 1.5,
  fontface.main = 1
)
```

***

### *ran*

```{r}
sym <- "ran"
goi <- genesFilt_Dr[mcols(genesFilt_Dr)$gene_name == sym,]$gene_id
counts <- dgeList_3$counts[goi,]
gm <- subset(exons_Dr, gene == goi) %>%
  subset(transcript %in% rownames(dgeFiltTx_3))
gmRed <- gm %>% 
  GenomicRanges::reduce()
st <- GenomicRanges::start(gmRed) %>% 
  min()
en <- GenomicRanges::end(gmRed) %>% 
  max()
chr <- seqnames(gmRed)@values
str <- gmRed %>% 
  strand() %>%
  .@values
gTrack <- GenomeAxisTrack()
rrTrack <- GeneRegionTrack(
  gmRed,
  name = NULL,
  fill = "darkorange"
)
rTrack <- GeneRegionTrack(
  gm,
  name = NULL
)
```

```{r}
## GC content track
win <- slidingWindows(gmRed, width = 1) %>% 
  unlist()
start <- win %>%
  start()
end <- win %>%
  end()
seq <- getSeq(dna, win)
gc <- seq %>%
  letterFrequency(letters = "GC", OR = "") %>%
  .[,1]
tbl <- tibble(
  start = start,
  end = end,
  gc = gc
) %>% 
  mutate(
    seqnames = chr,
    strand = str,
    cumsum = cumsum(gc),
    cummean = cummean(gc),
    lag = dplyr::lag(cumsum, n = 50, default = NA),
    lead = dplyr::lead(cumsum, n = 50, default = NA),
    gcsum = lead - lag,
    gcprop = gcsum / 101
  ) 
gcTrack <- tbl %>% 
  dplyr::select(seqnames, start, end, strand, gcprop) %>%
  makeGRangesFromDataFrame(
    keep.extra.columns = TRUE
  ) %>%
  DataTrack(
    type = c("b"),
    ylim = c(0, 1), 
    cex = 0.2, 
    name = "GC %",
    col.title="black",
    fontface.title=2,
    fontsize=10,
    grid = TRUE,
    lwd.grid = 0.5,
    lty.grid = "dashed",
    v = 0
  )
```

```{r}
## Coverage tracks
files <- list.files(
  "/data/biohub/20170327_Psen2S4Ter_RNASeq/data/2_alignedData/bam", 
  pattern = ".bam$", 
  full.names = TRUE
)
param <- ScanBamParam(
  what = c("pos", "qwidth"),
  which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
  flag=scanBamFlag(isUnmappedQuery=FALSE)
)
covFunc <- function(x){
  prop <- order_3$proportion[order_3$filenames == x] %>%
    percent(accuracy = 0.01)
  reads <- counts[order_3$sample[order_3$filenames == x]]
  pileup(x, scanBamParam = param) %>%
    dplyr::filter(strand == str) %>%
    mutate(start = pos, end = pos, count = count/reads) %>% 
    dplyr::select(seqnames, start, end, strand, count) %>%
    makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>%
    DataTrack(
      type = c("heatmap"),
      cex = 0.2, 
      name = prop,
      col.title="black",
      fontface.title=2,
      fontsize=10,
    )
}
covList <- lapply(files, covFunc) 
covOrd <- covList[as.integer(order_3$order)]
```

```{r}
## k-mer track
geneMod <- subset(genesGR_Dr, gene_id == goi)
geneSeq <- getSeq(dna, geneMod)
merTop <- merP500k8$mer[1:50]
# merLoc <- vmatchPattern(merTop[25], geneSeq)
merLoc <- lapply(merTop, function(x){
  vmatchPattern(x, geneSeq) %>%
    unlist()
})
merStart <- rapply(merLoc, function(x){
  start(x)
})
merWidth <- rapply(merLoc, function(x){
  width(x)
})
hTrack <- HighlightTrack(
  trackList = gcTrack, 
  start = start(geneMod) + merStart,
  width = merWidth,
  chromosome = chr,
  inBackground = TRUE,
  col = "white",
  fill = "red"
)
```

```{r, fig.height=9}
## Plot the tracks
plotTracks(
  c(gTrack, rrTrack, rTrack, covOrd, hTrack),
  from = st - 100,
  to = en + 100,
  chromosome = chr,
  sizes = c(1,0.5,1,rep(0.5, length(covOrd)),1), 
  main = sym,
  cex.main = 1.5,
  fontface.main = 1
)
```

*** 

### *rps16*

```{r}
sym <- "rps16"
goi <- genesFilt_Dr[mcols(genesFilt_Dr)$gene_name == sym,]$gene_id
counts <- dgeList_3$counts[goi,]
gm <- subset(exons_Dr, gene == goi) %>%
  subset(transcript %in% rownames(dgeFiltTx_3))
gmRed <- gm %>% 
  GenomicRanges::reduce()
st <- GenomicRanges::start(gmRed) %>% 
  min()
en <- GenomicRanges::end(gmRed) %>% 
  max()
chr <- seqnames(gmRed)@values
str <- gmRed %>% 
  strand() %>%
  .@values
gTrack <- GenomeAxisTrack()
rrTrack <- GeneRegionTrack(
  gmRed,
  name = NULL,
  fill = "darkorange"
)
rTrack <- GeneRegionTrack(
  gm,
  name = NULL
)
```

```{r}
win <- slidingWindows(gmRed, width = 1) %>% 
  unlist()
start <- win %>%
  start()
end <- win %>%
  end()
seq <- getSeq(dna, win)
gc <- seq %>%
  letterFrequency(letters = "GC", OR = "") %>%
  .[,1]
tbl <- tibble(
  start = start,
  end = end,
  gc = gc
) %>% 
  mutate(
    seqnames = chr,
    strand = str,
    cumsum = cumsum(gc),
    cummean = cummean(gc),
    lag = dplyr::lag(cumsum, n = 50, default = NA),
    lead = dplyr::lead(cumsum, n = 50, default = NA),
    gcsum = lead - lag,
    gcprop = gcsum / 101
  ) 
gcTrack <- tbl %>% 
  dplyr::select(seqnames, start, end, strand, gcprop) %>%
  makeGRangesFromDataFrame(
    keep.extra.columns = TRUE
  ) %>%
  DataTrack(
    type = c("b"),
    ylim = c(0, 1), 
    cex = 0.2, 
    name = "GC %",
    col.title="black",
    fontface.title=2,
    fontsize=10,
    grid = TRUE,
    lwd.grid = 0.5,
    lty.grid = "dashed",
    v = 0
  )
```

```{r}
files <- list.files(
  "/data/biohub/20170327_Psen2S4Ter_RNASeq/data/2_alignedData/bam", 
  pattern = ".bam$", 
  full.names = TRUE
)
param <- ScanBamParam(
  what = c("pos", "qwidth"),
  which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
  flag=scanBamFlag(isUnmappedQuery=FALSE)
)
covFunc <- function(x){
  prop <- order_3$proportion[order_3$filenames == x] %>%
    percent(accuracy = 0.01)
  reads <- counts[order_3$sample[order_3$filenames == x]]
  pileup(x, scanBamParam = param) %>%
    dplyr::filter(strand == str) %>%
    mutate(start = pos, end = pos, count = count/reads) %>% 
    dplyr::select(seqnames, start, end, strand, count) %>%
    makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>%
    DataTrack(
      type = c("heatmap"),
      cex = 0.2, 
      name = prop,
      col.title="black",
      fontface.title=2,
      fontsize=10,
    )
}
covList <- lapply(files, covFunc) 
covOrd <- covList[as.integer(order_3$order)]
```

```{r, fig.height=9}
plotTracks(
  c(gTrack, rrTrack, rTrack, covOrd, gcTrack),
  from = st - 100,
  to = en + 100,
  chromosome = chr,
  sizes = c(1,0.5,1,rep(0.5, length(covOrd)),1), 
  main = sym,
  cex.main = 1.5,
  fontface.main = 1
)
```

*** 

### *pnn*

```{r}
sym <- "pnn"
goi <- genesFilt_Dr[mcols(genesFilt_Dr)$gene_name == sym,]$gene_id
counts <- dgeList_3$counts[goi,]
gm <- subset(exons_Dr, gene == goi) %>%
  subset(transcript %in% rownames(dgeFiltTx_3))
gmRed <- gm %>% 
  GenomicRanges::reduce()
st <- GenomicRanges::start(gmRed) %>% 
  min()
en <- GenomicRanges::end(gmRed) %>% 
  max()
chr <- seqnames(gmRed)@values
str <- gmRed %>% 
  strand() %>%
  .@values
gTrack <- GenomeAxisTrack()
rrTrack <- GeneRegionTrack(
  gmRed,
  name = NULL,
  fill = "darkorange"
)
rTrack <- GeneRegionTrack(
  gm,
  name = NULL
)
```

```{r}
win <- slidingWindows(gmRed, width = 1) %>% 
  unlist()
start <- win %>%
  start()
end <- win %>%
  end()
seq <- getSeq(dna, win)
gc <- seq %>%
  letterFrequency(letters = "GC", OR = "") %>%
  .[,1]
tbl <- tibble(
  start = start,
  end = end,
  gc = gc
) %>% 
  mutate(
    seqnames = chr,
    strand = str,
    cumsum = cumsum(gc),
    cummean = cummean(gc),
    lag = dplyr::lag(cumsum, n = 50, default = NA),
    lead = dplyr::lead(cumsum, n = 50, default = NA),
    gcsum = lead - lag,
    gcprop = gcsum / 101
  ) 
gcTrack <- tbl %>% 
  dplyr::select(seqnames, start, end, strand, gcprop) %>%
  makeGRangesFromDataFrame(
    keep.extra.columns = TRUE
  ) %>%
  DataTrack(
    type = c("b"),
    ylim = c(0, 1), 
    cex = 0.2, 
    name = "GC %",
    col.title="black",
    fontface.title=2,
    fontsize=10,
    grid = TRUE,
    lwd.grid = 0.5,
    lty.grid = "dashed",
    v = 0
  )
```

```{r}
files <- list.files(
  "/data/biohub/20170327_Psen2S4Ter_RNASeq/data/2_alignedData/bam", 
  pattern = ".bam$", 
  full.names = TRUE
)
param <- ScanBamParam(
  what = c("pos", "qwidth"),
  which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
  flag=scanBamFlag(isUnmappedQuery=FALSE)
)
covFunc <- function(x){
  prop <- order_3$proportion[order_3$filenames == x] %>%
    percent(accuracy = 0.01)
  reads <- counts[order_3$sample[order_3$filenames == x]]
  pileup(x, scanBamParam = param) %>%
    dplyr::filter(strand == str) %>%
    mutate(start = pos, end = pos, count = count/reads) %>% 
    dplyr::select(seqnames, start, end, strand, count) %>%
    makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>%
    DataTrack(
      type = c("heatmap"),
      cex = 0.2, 
      name = prop,
      col.title="black",
      fontface.title=2,
      fontsize=10,
    )
}
covList <- lapply(files, covFunc) 
covOrd <- covList[as.integer(order_3$order)]
```

```{r, fig.height=9}
plotTracks(
  c(gTrack, rrTrack, rTrack, covOrd, gcTrack),
  from = st - 100,
  to = en + 100,
  chromosome = chr,
  sizes = c(1,0.5,1,rep(0.5, length(covOrd)),1), 
  main = sym,
  cex.main = 1.5,
  fontface.main = 1
)
```

***

### *sdhc*

```{r}
sym <- "sdhc"
goi <- genesFilt_Dr[mcols(genesFilt_Dr)$gene_name == sym,]$gene_id
counts <- dgeList_3$counts[goi,]
gm <- subset(exons_Dr, gene == goi) %>%
  subset(transcript %in% rownames(dgeFiltTx_3))
gmRed <- gm %>% 
  GenomicRanges::reduce()
st <- GenomicRanges::start(gmRed) %>% 
  min()
en <- GenomicRanges::end(gmRed) %>% 
  max()
chr <- seqnames(gmRed)@values
str <- gmRed %>% 
  strand() %>%
  .@values
gTrack <- GenomeAxisTrack()
rrTrack <- GeneRegionTrack(
  gmRed,
  name = NULL,
  fill = "darkorange"
)
rTrack <- GeneRegionTrack(
  gm,
  name = NULL
)
```

```{r}
win <- slidingWindows(gmRed, width = 1) %>% 
  unlist()
start <- win %>%
  start()
end <- win %>%
  end()
seq <- getSeq(dna, win)
gc <- seq %>%
  letterFrequency(letters = "GC", OR = "") %>%
  .[,1]
tbl <- tibble(
  start = start,
  end = end,
  gc = gc
) %>% 
  mutate(
    seqnames = chr,
    strand = str,
    cumsum = cumsum(gc),
    cummean = cummean(gc),
    lag = dplyr::lag(cumsum, n = 50, default = NA),
    lead = dplyr::lead(cumsum, n = 50, default = NA),
    gcsum = lead - lag,
    gcprop = gcsum / 101
  ) 
gcTrack <- tbl %>% 
  dplyr::select(seqnames, start, end, strand, gcprop) %>%
  makeGRangesFromDataFrame(
    keep.extra.columns = TRUE
  ) %>%
  DataTrack(
    type = c("b"),
    ylim = c(0, 1), 
    cex = 0.2, 
    name = "GC %",
    col.title="black",
    fontface.title=2,
    fontsize=10,
    grid = TRUE,
    lwd.grid = 0.5,
    lty.grid = "dashed",
    v = 0
  )
```

```{r}
files <- list.files(
  "/data/biohub/20170327_Psen2S4Ter_RNASeq/data/2_alignedData/bam", 
  pattern = ".bam$", 
  full.names = TRUE
)
param <- ScanBamParam(
  what = c("pos", "qwidth"),
  which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
  flag=scanBamFlag(isUnmappedQuery=FALSE)
)
covFunc <- function(x){
  prop <- order_3$proportion[order_3$filenames == x] %>%
    percent(accuracy = 0.01)
  reads <- counts[order_3$sample[order_3$filenames == x]]
  pileup(x, scanBamParam = param) %>%
    dplyr::filter(strand == str) %>%
    mutate(start = pos, end = pos, count = count/reads) %>% 
    dplyr::select(seqnames, start, end, strand, count) %>%
    makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>%
    DataTrack(
      type = c("heatmap"),
      cex = 0.2, 
      name = prop,
      col.title="black",
      fontface.title=2,
      fontsize=10,
    )
}
covList <- lapply(files, covFunc) 
covOrd <- covList[as.integer(order_3$order)]
```

```{r, fig.height=9}
plotTracks(
  c(gTrack, rrTrack, rTrack, covOrd, gcTrack),
  from = st - 100,
  to = en + 100,
  chromosome = chr,
  sizes = c(1,0.5,1,rep(0.5, length(covOrd)),1), 
  main = sym,
  cex.main = 1.5,
  fontface.main = 1
)
```

***

### *rpl29*

```{r}
sym <- "rpl29"
goi <- genesFilt_Dr[mcols(genesFilt_Dr)$gene_name == sym,]$gene_id
counts <- dgeList_3$counts[goi,]
gm <- subset(exons_Dr, gene == goi) %>%
  subset(transcript %in% rownames(dgeFiltTx_3))
gmRed <- gm %>% 
  GenomicRanges::reduce()
st <- GenomicRanges::start(gmRed) %>% 
  min()
en <- GenomicRanges::end(gmRed) %>% 
  max()
chr <- seqnames(gmRed)@values
str <- gmRed %>% 
  strand() %>%
  .@values
gTrack <- GenomeAxisTrack()
rrTrack <- GeneRegionTrack(
  gmRed,
  name = NULL,
  fill = "darkorange"
)
rTrack <- GeneRegionTrack(
  gm,
  name = NULL
)
```

```{r}
## GC content track
win <- slidingWindows(gmRed, width = 1) %>% 
  unlist()
start <- win %>%
  start()
end <- win %>%
  end()
seq <- getSeq(dna, win)
gc <- seq %>%
  letterFrequency(letters = "GC", OR = "") %>%
  .[,1]
tbl <- tibble(
  start = start,
  end = end,
  gc = gc
) %>% 
  mutate(
    seqnames = chr,
    strand = str,
    cumsum = cumsum(gc),
    cummean = cummean(gc),
    lag = dplyr::lag(cumsum, n = 50, default = NA),
    lead = dplyr::lead(cumsum, n = 50, default = NA),
    gcsum = lead - lag,
    gcprop = gcsum / 101
  ) 
gcTrack <- tbl %>% 
  dplyr::select(seqnames, start, end, strand, gcprop) %>%
  makeGRangesFromDataFrame(
    keep.extra.columns = TRUE
  ) %>%
  DataTrack(
    type = c("b"),
    ylim = c(0, 1), 
    cex = 0.2, 
    name = "GC %",
    col.title="black",
    fontface.title=2,
    fontsize=10,
    grid = TRUE,
    lwd.grid = 0.5,
    lty.grid = "dashed",
    v = 0
  )
```

```{r}
## Coverage tracks
files <- list.files(
  "/data/biohub/20170327_Psen2S4Ter_RNASeq/data/2_alignedData/bam", 
  pattern = ".bam$", 
  full.names = TRUE
)
param <- ScanBamParam(
  what = c("pos", "qwidth"),
  which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
  flag=scanBamFlag(isUnmappedQuery=FALSE)
)
covFunc <- function(x){
  prop <- order_3$proportion[order_3$filenames == x] %>%
    percent(accuracy = 0.01)
  reads <- counts[order_3$sample[order_3$filenames == x]]
  pileup(x, scanBamParam = param) %>%
    dplyr::filter(strand == str) %>%
    mutate(start = pos, end = pos, count = count/reads) %>% 
    dplyr::select(seqnames, start, end, strand, count) %>%
    makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>%
    DataTrack(
      type = c("heatmap"),
      cex = 0.2, 
      name = prop,
      col.title="black",
      fontface.title=2,
      fontsize=10,
    )
}
covList <- lapply(files, covFunc) 
covOrd <- covList[as.integer(order_3$order)]
```

```{r}
## k-mer track
geneMod <- subset(genesGR_Dr, gene_id == goi)
geneSeq <- getSeq(dna, geneMod)
merTop <- merP500k8$mer[1:50]
# merLoc <- vmatchPattern(merTop[25], geneSeq)
merLoc <- lapply(merTop, function(x){
  vmatchPattern(x, geneSeq) %>%
    unlist()
})
merStart <- rapply(merLoc, function(x){
  start(x)
})
merWidth <- rapply(merLoc, function(x){
  width(x)
})
hTrack <- HighlightTrack(
  trackList = gcTrack, 
  start = start(geneMod) + merStart,
  width = merWidth,
  chromosome = chr,
  inBackground = TRUE,
  col = "white",
  fill = "red"
)
```

```{r, fig.height=9}
## Plot the tracks
plotTracks(
  c(gTrack, rrTrack, rTrack, covOrd, hTrack),
  from = st - 100,
  to = en + 100,
  chromosome = chr,
  sizes = c(1,0.5,1,rep(0.5, length(covOrd)),1), 
  main = sym,
  cex.main = 1.5,
  fontface.main = 1
)
```

***

### *rpl34*

```{r}
sym <- "rpl34"
goi <- genesFilt_Dr[mcols(genesFilt_Dr)$gene_name == sym,]$gene_id
counts <- dgeList_3$counts[goi,]
gm <- subset(exons_Dr, gene == goi) %>%
  subset(transcript %in% rownames(dgeFiltTx_3))
gmRed <- gm %>% 
  GenomicRanges::reduce()
st <- GenomicRanges::start(gmRed) %>% 
  min()
en <- GenomicRanges::end(gmRed) %>% 
  max()
chr <- seqnames(gmRed)@values
str <- gmRed %>% 
  strand() %>%
  .@values
gTrack <- GenomeAxisTrack()
rrTrack <- GeneRegionTrack(
  gmRed,
  name = NULL,
  fill = "darkorange"
)
rTrack <- GeneRegionTrack(
  gm,
  name = NULL
)
```

```{r}
win <- slidingWindows(gmRed, width = 1) %>% 
  unlist()
start <- win %>%
  start()
end <- win %>%
  end()
seq <- getSeq(dna, win)
gc <- seq %>%
  letterFrequency(letters = "GC", OR = "") %>%
  .[,1]
tbl <- tibble(
  start = start,
  end = end,
  gc = gc
) %>% 
  mutate(
    seqnames = chr,
    strand = str,
    cumsum = cumsum(gc),
    cummean = cummean(gc),
    lag = dplyr::lag(cumsum, n = 50, default = NA),
    lead = dplyr::lead(cumsum, n = 50, default = NA),
    gcsum = lead - lag,
    gcprop = gcsum / 101
  ) 
gcTrack <- tbl %>% 
  dplyr::select(seqnames, start, end, strand, gcprop) %>%
  makeGRangesFromDataFrame(
    keep.extra.columns = TRUE
  ) %>%
  DataTrack(
    type = c("b"),
    ylim = c(0, 1), 
    cex = 0.2, 
    name = "GC %",
    col.title="black",
    fontface.title=2,
    fontsize=10,
    grid = TRUE,
    lwd.grid = 0.5,
    lty.grid = "dashed",
    v = 0
  )
```

```{r}
files <- list.files(
  "/data/biohub/20170327_Psen2S4Ter_RNASeq/data/2_alignedData/bam", 
  pattern = ".bam$", 
  full.names = TRUE
)
param <- ScanBamParam(
  what = c("pos", "qwidth"),
  which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
  flag=scanBamFlag(isUnmappedQuery=FALSE)
)
covFunc <- function(x){
  prop <- order_3$proportion[order_3$filenames == x] %>%
    percent(accuracy = 0.01)
  reads <- counts[order_3$sample[order_3$filenames == x]]
  pileup(x, scanBamParam = param) %>%
    dplyr::filter(strand == str) %>%
    mutate(start = pos, end = pos, count = count/reads) %>% 
    dplyr::select(seqnames, start, end, strand, count) %>%
    makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>%
    DataTrack(
      type = c("heatmap"),
      cex = 0.2, 
      name = prop,
      col.title="black",
      fontface.title=2,
      fontsize=10,
    )
}
covList <- lapply(files, covFunc) 
covOrd <- covList[as.integer(order_3$order)]
```

```{r, fig.height=9}
plotTracks(
  c(gTrack, rrTrack, rTrack, covOrd, gcTrack),
  from = st - 100,
  to = en + 100,
  chromosome = chr,
  sizes = c(1,0.5,1,rep(0.5, length(covOrd)),1), 
  main = sym,
  cex.main = 1.5,
  fontface.main = 1
)
```

***

### *COLGALT1*

```{r}
sym <- "COLGALT1"
goi <- genesFilt_Dr[mcols(genesFilt_Dr)$gene_name == sym,]$gene_id
counts <- dgeList_3$counts[goi,]
gm <- subset(exons_Dr, gene == goi) %>%
  subset(transcript %in% rownames(dgeFiltTx_3))
gmRed <- gm %>% 
  GenomicRanges::reduce()
st <- GenomicRanges::start(gmRed) %>% 
  min()
en <- GenomicRanges::end(gmRed) %>% 
  max()
chr <- seqnames(gmRed)@values
str <- gmRed %>% 
  strand() %>%
  .@values
gTrack <- GenomeAxisTrack()
rrTrack <- GeneRegionTrack(
  gmRed,
  name = NULL,
  fill = "darkorange"
)
rTrack <- GeneRegionTrack(
  gm,
  name = NULL
)
```

```{r}
win <- slidingWindows(gmRed, width = 1) %>% 
  unlist()
start <- win %>%
  start()
end <- win %>%
  end()
seq <- getSeq(dna, win)
gc <- seq %>%
  letterFrequency(letters = "GC", OR = "") %>%
  .[,1]
tbl <- tibble(
  start = start,
  end = end,
  gc = gc
) %>% 
  mutate(
    seqnames = chr,
    strand = str,
    cumsum = cumsum(gc),
    cummean = cummean(gc),
    lag = dplyr::lag(cumsum, n = 50, default = NA),
    lead = dplyr::lead(cumsum, n = 50, default = NA),
    gcsum = lead - lag,
    gcprop = gcsum / 101
  ) 
gcTrack <- tbl %>% 
  dplyr::select(seqnames, start, end, strand, gcprop) %>%
  makeGRangesFromDataFrame(
    keep.extra.columns = TRUE
  ) %>%
  DataTrack(
    type = c("b"),
    ylim = c(0, 1), 
    cex = 0.2, 
    name = "GC %",
    col.title="black",
    fontface.title=2,
    fontsize=10,
    grid = TRUE,
    lwd.grid = 0.5,
    lty.grid = "dashed",
    v = 0
  )
```

```{r}
files <- list.files(
  "/data/biohub/20170327_Psen2S4Ter_RNASeq/data/2_alignedData/bam", 
  pattern = ".bam$", 
  full.names = TRUE
)
param <- ScanBamParam(
  what = c("pos", "qwidth"),
  which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
  flag=scanBamFlag(isUnmappedQuery=FALSE)
)
covFunc <- function(x){
  prop <- order_3$proportion[order_3$filenames == x] %>%
    percent(accuracy = 0.01)
  reads <- counts[order_3$sample[order_3$filenames == x]]
  pileup(x, scanBamParam = param) %>%
    dplyr::filter(strand == str) %>%
    mutate(start = pos, end = pos, count = count/reads) %>% 
    dplyr::select(seqnames, start, end, strand, count) %>%
    makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>%
    DataTrack(
      type = c("heatmap"),
      cex = 0.2, 
      name = prop,
      col.title="black",
      fontface.title=2,
      fontsize=10,
    )
}
covList <- lapply(files, covFunc) 
covOrd <- covList[as.integer(order_3$order)]
```

```{r, fig.height=9}
plotTracks(
  c(gTrack, rrTrack, rTrack, covOrd, gcTrack),
  from = st - 100,
  to = en + 100,
  chromosome = chr,
  sizes = c(1,0.5,1,rep(0.5, length(covOrd)),1), 
  main = sym,
  cex.main = 1.5,
  fontface.main = 1
)
```

***

### *sub1b*

```{r}
sym <- "sub1b"
goi <- genesFilt_Dr[mcols(genesFilt_Dr)$gene_name == sym,]$gene_id
counts <- dgeList_3$counts[goi,]
gm <- subset(exons_Dr, gene == goi) %>%
  subset(transcript %in% rownames(dgeFiltTx_3))
gmRed <- gm %>% 
  GenomicRanges::reduce()
st <- GenomicRanges::start(gmRed) %>% 
  min()
en <- GenomicRanges::end(gmRed) %>% 
  max()
chr <- seqnames(gmRed)@values
str <- gmRed %>% 
  strand() %>%
  .@values
gTrack <- GenomeAxisTrack()
rrTrack <- GeneRegionTrack(
  gmRed,
  name = NULL,
  fill = "darkorange"
)
rTrack <- GeneRegionTrack(
  gm,
  name = NULL
)
```

```{r}
win <- slidingWindows(gmRed, width = 1) %>% 
  unlist()
start <- win %>%
  start()
end <- win %>%
  end()
seq <- getSeq(dna, win)
gc <- seq %>%
  letterFrequency(letters = "GC", OR = "") %>%
  .[,1]
tbl <- tibble(
  start = start,
  end = end,
  gc = gc
) %>% 
  mutate(
    seqnames = chr,
    strand = str,
    cumsum = cumsum(gc),
    cummean = cummean(gc),
    lag = dplyr::lag(cumsum, n = 50, default = NA),
    lead = dplyr::lead(cumsum, n = 50, default = NA),
    gcsum = lead - lag,
    gcprop = gcsum / 101
  ) 
gcTrack <- tbl %>% 
  dplyr::select(seqnames, start, end, strand, gcprop) %>%
  makeGRangesFromDataFrame(
    keep.extra.columns = TRUE
  ) %>%
  DataTrack(
    type = c("b"),
    ylim = c(0, 1), 
    cex = 0.2, 
    name = "GC %",
    col.title="black",
    fontface.title=2,
    fontsize=10,
    grid = TRUE,
    lwd.grid = 0.5,
    lty.grid = "dashed",
    v = 0
  )
```

```{r}
files <- list.files(
  "/data/biohub/20170327_Psen2S4Ter_RNASeq/data/2_alignedData/bam", 
  pattern = ".bam$", 
  full.names = TRUE
)
param <- ScanBamParam(
  what = c("pos", "qwidth"),
  which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
  flag=scanBamFlag(isUnmappedQuery=FALSE)
)
covFunc <- function(x){
  prop <- order_3$proportion[order_3$filenames == x] %>%
    percent(accuracy = 0.01)
  reads <- counts[order_3$sample[order_3$filenames == x]]
  pileup(x, scanBamParam = param) %>%
    dplyr::filter(strand == str) %>%
    mutate(start = pos, end = pos, count = count/reads) %>% 
    dplyr::select(seqnames, start, end, strand, count) %>%
    makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>%
    DataTrack(
      type = c("heatmap"),
      cex = 0.2, 
      name = prop,
      col.title="black",
      fontface.title=2,
      fontsize=10,
    )
}
covList <- lapply(files, covFunc) 
covOrd <- covList[as.integer(order_3$order)]
```

```{r, fig.height=9}
plotTracks(
  c(gTrack, rrTrack, rTrack, covOrd, gcTrack),
  from = st - 100,
  to = en + 100,
  chromosome = chr,
  sizes = c(1,0.5,1,rep(0.5, length(covOrd)),1), 
  main = sym,
  cex.main = 1.5,
  fontface.main = 1
)
```

***

### *serinc1*

```{r}
sym <- "serinc1"
goi <- genesFilt_Dr[mcols(genesFilt_Dr)$gene_name == sym,]$gene_id
counts <- dgeList_3$counts[goi,]
gm <- subset(exons_Dr, gene == goi) %>%
  subset(transcript %in% rownames(dgeFiltTx_3))
gmRed <- gm %>% 
  GenomicRanges::reduce()
st <- GenomicRanges::start(gmRed) %>% 
  min()
en <- GenomicRanges::end(gmRed) %>% 
  max()
chr <- seqnames(gmRed)@values
str <- gmRed %>% 
  strand() %>%
  .@values
gTrack <- GenomeAxisTrack()
rrTrack <- GeneRegionTrack(
  gmRed,
  name = NULL,
  fill = "darkorange"
)
rTrack <- GeneRegionTrack(
  gm,
  name = NULL
)
```

```{r}
win <- slidingWindows(gmRed, width = 1) %>% 
  unlist()
start <- win %>%
  start()
end <- win %>%
  end()
seq <- getSeq(dna, win)
gc <- seq %>%
  letterFrequency(letters = "GC", OR = "") %>%
  .[,1]
tbl <- tibble(
  start = start,
  end = end,
  gc = gc
) %>% 
  mutate(
    seqnames = chr,
    strand = str,
    cumsum = cumsum(gc),
    cummean = cummean(gc),
    lag = dplyr::lag(cumsum, n = 50, default = NA),
    lead = dplyr::lead(cumsum, n = 50, default = NA),
    gcsum = lead - lag,
    gcprop = gcsum / 101
  ) 
gcTrack <- tbl %>% 
  dplyr::select(seqnames, start, end, strand, gcprop) %>%
  makeGRangesFromDataFrame(
    keep.extra.columns = TRUE
  ) %>%
  DataTrack(
    type = c("b"),
    ylim = c(0, 1), 
    cex = 0.2, 
    name = "GC %",
    col.title="black",
    fontface.title=2,
    fontsize=10,
    grid = TRUE,
    lwd.grid = 0.5,
    lty.grid = "dashed",
    v = 0
  )
```

```{r}
files <- list.files(
  "/data/biohub/20170327_Psen2S4Ter_RNASeq/data/2_alignedData/bam", 
  pattern = ".bam$", 
  full.names = TRUE
)
param <- ScanBamParam(
  what = c("pos", "qwidth"),
  which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
  flag=scanBamFlag(isUnmappedQuery=FALSE)
)
covFunc <- function(x){
  prop <- order_3$proportion[order_3$filenames == x] %>%
    percent(accuracy = 0.01)
  reads <- counts[order_3$sample[order_3$filenames == x]]
  pileup(x, scanBamParam = param) %>%
    dplyr::filter(strand == str) %>%
    mutate(start = pos, end = pos, count = count/reads) %>% 
    dplyr::select(seqnames, start, end, strand, count) %>%
    makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>%
    DataTrack(
      type = c("heatmap"),
      cex = 0.2, 
      name = prop,
      col.title="black",
      fontface.title=2,
      fontsize=10,
    )
}
covList <- lapply(files, covFunc) 
covOrd <- covList[as.integer(order_3$order)]
```

```{r, fig.height=9}
plotTracks(
  c(gTrack, rrTrack, rTrack, covOrd, gcTrack),
  from = st - 100,
  to = en + 100,
  chromosome = chr,
  sizes = c(1,0.5,1,rep(0.5, length(covOrd)),1), 
  main = sym,
  cex.main = 1.5,
  fontface.main = 1
)
```

***

### *acin1b*

```{r}
sym <- "acin1b"
goi <- genesFilt_Dr[mcols(genesFilt_Dr)$gene_name == sym,]$gene_id
counts <- dgeList_3$counts[goi,]
gm <- subset(exons_Dr, gene == goi) %>%
  subset(transcript %in% rownames(dgeFiltTx_3))
gmRed <- gm %>% 
  GenomicRanges::reduce()
st <- GenomicRanges::start(gmRed) %>% 
  min()
en <- GenomicRanges::end(gmRed) %>% 
  max()
chr <- seqnames(gmRed)@values
str <- gmRed %>% 
  strand() %>%
  .@values
gTrack <- GenomeAxisTrack()
rrTrack <- GeneRegionTrack(
  gmRed,
  name = NULL,
  fill = "darkorange"
)
rTrack <- GeneRegionTrack(
  gm,
  name = NULL
)
```

```{r}
win <- slidingWindows(gmRed, width = 1) %>% 
  unlist()
start <- win %>%
  start()
end <- win %>%
  end()
seq <- getSeq(dna, win)
gc <- seq %>%
  letterFrequency(letters = "GC", OR = "") %>%
  .[,1]
tbl <- tibble(
  start = start,
  end = end,
  gc = gc
) %>% 
  mutate(
    seqnames = chr,
    strand = str,
    cumsum = cumsum(gc),
    cummean = cummean(gc),
    lag = dplyr::lag(cumsum, n = 50, default = NA),
    lead = dplyr::lead(cumsum, n = 50, default = NA),
    gcsum = lead - lag,
    gcprop = gcsum / 101
  ) 
gcTrack <- tbl %>% 
  dplyr::select(seqnames, start, end, strand, gcprop) %>%
  makeGRangesFromDataFrame(
    keep.extra.columns = TRUE
  ) %>%
  DataTrack(
    type = c("b"),
    ylim = c(0, 1), 
    cex = 0.2, 
    name = "GC %",
    col.title="black",
    fontface.title=2,
    fontsize=10,
    grid = TRUE,
    lwd.grid = 0.5,
    lty.grid = "dashed",
    v = 0
  )
```

```{r}
files <- list.files(
  "/data/biohub/20170327_Psen2S4Ter_RNASeq/data/2_alignedData/bam", 
  pattern = ".bam$", 
  full.names = TRUE
)
param <- ScanBamParam(
  what = c("pos", "qwidth"),
  which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
  flag=scanBamFlag(isUnmappedQuery=FALSE)
)
covFunc <- function(x){
  prop <- order_3$proportion[order_3$filenames == x] %>%
    percent(accuracy = 0.01)
  reads <- counts[order_3$sample[order_3$filenames == x]]
  pileup(x, scanBamParam = param) %>%
    dplyr::filter(strand == str) %>%
    mutate(start = pos, end = pos, count = count/reads) %>% 
    dplyr::select(seqnames, start, end, strand, count) %>%
    makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>%
    DataTrack(
      type = c("heatmap"),
      cex = 0.2, 
      name = prop,
      col.title="black",
      fontface.title=2,
      fontsize=10,
    )
}
covList <- lapply(files, covFunc) 
covOrd <- covList[as.integer(order_3$order)]
```

```{r, fig.height=9}
plotTracks(
  c(gTrack, rrTrack, rTrack, covOrd, gcTrack),
  from = st - 100,
  to = en + 100,
  chromosome = chr,
  sizes = c(1,0.5,1,rep(0.5, length(covOrd)),1), 
  main = sym,
  cex.main = 1.5,
  fontface.main = 1
)
```

***

### *elof1*

```{r}
sym <- "elof1"
goi <- genesFilt_Dr[mcols(genesFilt_Dr)$gene_name == sym,]$gene_id
counts <- dgeList_3$counts[goi,]
gm <- subset(exons_Dr, gene == goi) %>%
  subset(transcript %in% rownames(dgeFiltTx_3))
gmRed <- gm %>% 
  GenomicRanges::reduce()
st <- GenomicRanges::start(gmRed) %>% 
  min()
en <- GenomicRanges::end(gmRed) %>% 
  max()
chr <- seqnames(gmRed)@values
str <- gmRed %>% 
  strand() %>%
  .@values
gTrack <- GenomeAxisTrack()
rrTrack <- GeneRegionTrack(
  gmRed,
  name = NULL,
  fill = "darkorange"
)
rTrack <- GeneRegionTrack(
  gm,
  name = NULL
)
```

```{r}
win <- slidingWindows(gmRed, width = 1) %>% 
  unlist()
start <- win %>%
  start()
end <- win %>%
  end()
seq <- getSeq(dna, win)
gc <- seq %>%
  letterFrequency(letters = "GC", OR = "") %>%
  .[,1]
tbl <- tibble(
  start = start,
  end = end,
  gc = gc
) %>% 
  mutate(
    seqnames = chr,
    strand = str,
    cumsum = cumsum(gc),
    cummean = cummean(gc),
    lag = dplyr::lag(cumsum, n = 50, default = NA),
    lead = dplyr::lead(cumsum, n = 50, default = NA),
    gcsum = lead - lag,
    gcprop = gcsum / 101
  ) 
gcTrack <- tbl %>% 
  dplyr::select(seqnames, start, end, strand, gcprop) %>%
  makeGRangesFromDataFrame(
    keep.extra.columns = TRUE
  ) %>%
  DataTrack(
    type = c("b"),
    ylim = c(0, 1), 
    cex = 0.2, 
    name = "GC %",
    col.title="black",
    fontface.title=2,
    fontsize=10,
    grid = TRUE,
    lwd.grid = 0.5,
    lty.grid = "dashed",
    v = 0
  )
```

```{r}
files <- list.files(
  "/data/biohub/20170327_Psen2S4Ter_RNASeq/data/2_alignedData/bam", 
  pattern = ".bam$", 
  full.names = TRUE
)
param <- ScanBamParam(
  what = c("pos", "qwidth"),
  which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
  flag=scanBamFlag(isUnmappedQuery=FALSE)
)
covFunc <- function(x){
  prop <- order_3$proportion[order_3$filenames == x] %>%
    percent(accuracy = 0.01)
  reads <- counts[order_3$sample[order_3$filenames == x]]
  pileup(x, scanBamParam = param) %>%
    dplyr::filter(strand == str) %>%
    mutate(start = pos, end = pos, count = count/reads) %>% 
    dplyr::select(seqnames, start, end, strand, count) %>%
    makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>%
    DataTrack(
      type = c("heatmap"),
      cex = 0.2, 
      name = prop,
      col.title="black",
      fontface.title=2,
      fontsize=10,
    )
}
covList <- lapply(files, covFunc) 
covOrd <- covList[as.integer(order_3$order)]
```

```{r, fig.height=9}
plotTracks(
  c(gTrack, rrTrack, rTrack, covOrd, gcTrack),
  from = st - 100,
  to = en + 100,
  chromosome = chr,
  sizes = c(1,0.5,1,rep(0.5, length(covOrd)),1), 
  main = sym,
  cex.main = 1.5,
  fontface.main = 1
)
```

***

### *cox7c*

```{r}
sym <- "cox7c"
goi <- genesFilt_Dr[mcols(genesFilt_Dr)$gene_name == sym,]$gene_id
counts <- dgeList_3$counts[goi,]
gm <- subset(exons_Dr, gene == goi) %>%
  subset(transcript %in% rownames(dgeFiltTx_3))
gmRed <- gm %>% 
  GenomicRanges::reduce()
st <- GenomicRanges::start(gmRed) %>% 
  min()
en <- GenomicRanges::end(gmRed) %>% 
  max()
chr <- seqnames(gmRed)@values
str <- gmRed %>% 
  strand() %>%
  .@values
gTrack <- GenomeAxisTrack()
rrTrack <- GeneRegionTrack(
  gmRed,
  name = NULL,
  fill = "darkorange"
)
rTrack <- GeneRegionTrack(
  gm,
  name = NULL
)
```

```{r}
win <- slidingWindows(gmRed, width = 1) %>% 
  unlist()
start <- win %>%
  start()
end <- win %>%
  end()
seq <- getSeq(dna, win)
gc <- seq %>%
  letterFrequency(letters = "GC", OR = "") %>%
  .[,1]
tbl <- tibble(
  start = start,
  end = end,
  gc = gc
) %>% 
  mutate(
    seqnames = chr,
    strand = str,
    cumsum = cumsum(gc),
    cummean = cummean(gc),
    lag = dplyr::lag(cumsum, n = 50, default = NA),
    lead = dplyr::lead(cumsum, n = 50, default = NA),
    gcsum = lead - lag,
    gcprop = gcsum / 101
  ) 
gcTrack <- tbl %>% 
  dplyr::select(seqnames, start, end, strand, gcprop) %>%
  makeGRangesFromDataFrame(
    keep.extra.columns = TRUE
  ) %>%
  DataTrack(
    type = c("b"),
    ylim = c(0, 1), 
    cex = 0.2, 
    name = "GC %",
    col.title="black",
    fontface.title=2,
    fontsize=10,
    grid = TRUE,
    lwd.grid = 0.5,
    lty.grid = "dashed",
    v = 0
  )
```

```{r}
files <- list.files(
  "/data/biohub/20170327_Psen2S4Ter_RNASeq/data/2_alignedData/bam", 
  pattern = ".bam$", 
  full.names = TRUE
)
param <- ScanBamParam(
  what = c("pos", "qwidth"),
  which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
  flag=scanBamFlag(isUnmappedQuery=FALSE)
)
covFunc <- function(x){
  prop <- order_3$proportion[order_3$filenames == x] %>%
    percent(accuracy = 0.01)
  reads <- counts[order_3$sample[order_3$filenames == x]]
  pileup(x, scanBamParam = param) %>%
    dplyr::filter(strand == str) %>%
    mutate(start = pos, end = pos, count = count/reads) %>% 
    dplyr::select(seqnames, start, end, strand, count) %>%
    makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>%
    DataTrack(
      type = c("heatmap"),
      cex = 0.2, 
      name = prop,
      col.title="black",
      fontface.title=2,
      fontsize=10,
    )
}
covList <- lapply(files, covFunc) 
covOrd <- covList[as.integer(order_3$order)]
```

```{r, fig.height=9}
plotTracks(
  c(gTrack, rrTrack, rTrack, covOrd, gcTrack),
  from = st - 100,
  to = en + 100,
  chromosome = chr,
  sizes = c(1,0.5,1,rep(0.5, length(covOrd)),1), 
  main = sym,
  cex.main = 1.5,
  fontface.main = 1
)
```

***

### *atpv0e2*

```{r}
sym <- "atpv0e2"
goi <- genesFilt_Dr[mcols(genesFilt_Dr)$gene_name == sym,]$gene_id
counts <- dgeList_3$counts[goi,]
gm <- subset(exons_Dr, gene == goi) %>%
  subset(transcript %in% rownames(dgeFiltTx_3))
gmRed <- gm %>% 
  GenomicRanges::reduce()
st <- GenomicRanges::start(gmRed) %>% 
  min()
en <- GenomicRanges::end(gmRed) %>% 
  max()
chr <- seqnames(gmRed)@values
str <- gmRed %>% 
  strand() %>%
  .@values
gTrack <- GenomeAxisTrack()
rrTrack <- GeneRegionTrack(
  gmRed,
  name = NULL,
  fill = "darkorange"
)
rTrack <- GeneRegionTrack(
  gm,
  name = NULL
)
```

```{r}
win <- slidingWindows(gmRed, width = 1) %>% 
  unlist()
start <- win %>%
  start()
end <- win %>%
  end()
seq <- getSeq(dna, win)
gc <- seq %>%
  letterFrequency(letters = "GC", OR = "") %>%
  .[,1]
tbl <- tibble(
  start = start,
  end = end,
  gc = gc
) %>% 
  mutate(
    seqnames = chr,
    strand = str,
    cumsum = cumsum(gc),
    cummean = cummean(gc),
    lag = dplyr::lag(cumsum, n = 50, default = NA),
    lead = dplyr::lead(cumsum, n = 50, default = NA),
    gcsum = lead - lag,
    gcprop = gcsum / 101
  ) 
gcTrack <- tbl %>% 
  dplyr::select(seqnames, start, end, strand, gcprop) %>%
  makeGRangesFromDataFrame(
    keep.extra.columns = TRUE
  ) %>%
  DataTrack(
    type = c("b"),
    ylim = c(0, 1), 
    cex = 0.2, 
    name = "GC %",
    col.title="black",
    fontface.title=2,
    fontsize=10,
    grid = TRUE,
    lwd.grid = 0.5,
    lty.grid = "dashed",
    v = 0
  )
```

```{r}
files <- list.files(
  "/data/biohub/20170327_Psen2S4Ter_RNASeq/data/2_alignedData/bam", 
  pattern = ".bam$", 
  full.names = TRUE
)
param <- ScanBamParam(
  what = c("pos", "qwidth"),
  which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
  flag=scanBamFlag(isUnmappedQuery=FALSE)
)
covFunc <- function(x){
  prop <- order_3$proportion[order_3$filenames == x] %>%
    percent(accuracy = 0.01)
  reads <- counts[order_3$sample[order_3$filenames == x]]
  pileup(x, scanBamParam = param) %>%
    dplyr::filter(strand == str) %>%
    mutate(start = pos, end = pos, count = count/reads) %>% 
    dplyr::select(seqnames, start, end, strand, count) %>%
    makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>%
    DataTrack(
      type = c("heatmap"),
      cex = 0.2, 
      name = prop,
      col.title="black",
      fontface.title=2,
      fontsize=10,
    )
}
covList <- lapply(files, covFunc) 
covOrd <- covList[as.integer(order_3$order)]
```

```{r, fig.height=9}
plotTracks(
  c(gTrack, rrTrack, rTrack, covOrd, gcTrack),
  from = st - 100,
  to = en + 100,
  chromosome = chr,
  sizes = c(1,0.5,1,rep(0.5, length(covOrd)),1), 
  main = sym,
  cex.main = 1.5,
  fontface.main = 1
)
```

***

### *dido1*

```{r}
sym <- "dido1"
goi <- genesFilt_Dr[mcols(genesFilt_Dr)$gene_name == sym,]$gene_id
counts <- dgeList_3$counts[goi,]
gm <- subset(exons_Dr, gene == goi) %>%
  subset(transcript %in% rownames(dgeFiltTx_3))
gmRed <- gm %>% 
  GenomicRanges::reduce()
st <- GenomicRanges::start(gmRed) %>% 
  min()
en <- GenomicRanges::end(gmRed) %>% 
  max()
chr <- seqnames(gmRed)@values
str <- gmRed %>% 
  strand() %>%
  .@values
gTrack <- GenomeAxisTrack()
rrTrack <- GeneRegionTrack(
  gmRed,
  name = NULL,
  fill = "darkorange"
)
rTrack <- GeneRegionTrack(
  gm,
  name = NULL
)
```

```{r}
win <- slidingWindows(gmRed, width = 1) %>% 
  unlist()
start <- win %>%
  start()
end <- win %>%
  end()
seq <- getSeq(dna, win)
gc <- seq %>%
  letterFrequency(letters = "GC", OR = "") %>%
  .[,1]
tbl <- tibble(
  start = start,
  end = end,
  gc = gc
) %>% 
  mutate(
    seqnames = chr,
    strand = str,
    cumsum = cumsum(gc),
    cummean = cummean(gc),
    lag = dplyr::lag(cumsum, n = 50, default = NA),
    lead = dplyr::lead(cumsum, n = 50, default = NA),
    gcsum = lead - lag,
    gcprop = gcsum / 101
  ) 
gcTrack <- tbl %>% 
  dplyr::select(seqnames, start, end, strand, gcprop) %>%
  makeGRangesFromDataFrame(
    keep.extra.columns = TRUE
  ) %>%
  DataTrack(
    type = c("b"),
    ylim = c(0, 1), 
    cex = 0.2, 
    name = "GC %",
    col.title="black",
    fontface.title=2,
    fontsize=10,
    grid = TRUE,
    lwd.grid = 0.5,
    lty.grid = "dashed",
    v = 0
  )
```

```{r}
files <- list.files(
  "/data/biohub/20170327_Psen2S4Ter_RNASeq/data/2_alignedData/bam", 
  pattern = ".bam$", 
  full.names = TRUE
)
param <- ScanBamParam(
  what = c("pos", "qwidth"),
  which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
  flag=scanBamFlag(isUnmappedQuery=FALSE)
)
covFunc <- function(x){
  prop <- order_3$proportion[order_3$filenames == x] %>%
    percent(accuracy = 0.01)
  reads <- counts[order_3$sample[order_3$filenames == x]]
  pileup(x, scanBamParam = param) %>%
    dplyr::filter(strand == str) %>%
    mutate(start = pos, end = pos, count = count/reads) %>% 
    dplyr::select(seqnames, start, end, strand, count) %>%
    makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>%
    DataTrack(
      type = c("heatmap"),
      cex = 0.2, 
      name = prop,
      col.title="black",
      fontface.title=2,
      fontsize=10,
    )
}
covList <- lapply(files, covFunc) 
covOrd <- covList[as.integer(order_3$order)]
```

```{r, fig.height=9}
plotTracks(
  c(gTrack, rrTrack, rTrack, covOrd, gcTrack),
  from = st - 100,
  to = en + 100,
  chromosome = chr,
  sizes = c(1,0.5,1,rep(0.5, length(covOrd)),1), 
  main = sym,
  cex.main = 1.5,
  fontface.main = 1
)
```

***

### *brd2b*

```{r}
sym <- "brd2b"
goi <- genesFilt_Dr[mcols(genesFilt_Dr)$gene_name == sym,]$gene_id
counts <- dgeList_3$counts[goi,]
gm <- subset(exons_Dr, gene == goi) %>%
  subset(transcript %in% rownames(dgeFiltTx_3))
gmRed <- gm %>% 
  GenomicRanges::reduce()
st <- GenomicRanges::start(gmRed) %>% 
  min()
en <- GenomicRanges::end(gmRed) %>% 
  max()
chr <- seqnames(gmRed)@values
str <- gmRed %>% 
  strand() %>%
  .@values
gTrack <- GenomeAxisTrack()
rrTrack <- GeneRegionTrack(
  gmRed,
  name = NULL,
  fill = "darkorange"
)
rTrack <- GeneRegionTrack(
  gm,
  name = NULL
)
```

```{r}
win <- slidingWindows(gmRed, width = 1) %>% 
  unlist()
start <- win %>%
  start()
end <- win %>%
  end()
seq <- getSeq(dna, win)
gc <- seq %>%
  letterFrequency(letters = "GC", OR = "") %>%
  .[,1]
tbl <- tibble(
  start = start,
  end = end,
  gc = gc
) %>% 
  mutate(
    seqnames = chr,
    strand = str,
    cumsum = cumsum(gc),
    cummean = cummean(gc),
    lag = dplyr::lag(cumsum, n = 50, default = NA),
    lead = dplyr::lead(cumsum, n = 50, default = NA),
    gcsum = lead - lag,
    gcprop = gcsum / 101
  ) 
gcTrack <- tbl %>% 
  dplyr::select(seqnames, start, end, strand, gcprop) %>%
  makeGRangesFromDataFrame(
    keep.extra.columns = TRUE
  ) %>%
  DataTrack(
    type = c("b"),
    ylim = c(0, 1), 
    cex = 0.2, 
    name = "GC %",
    col.title="black",
    fontface.title=2,
    fontsize=10,
    grid = TRUE,
    lwd.grid = 0.5,
    lty.grid = "dashed",
    v = 0
  )
```

```{r}
files <- list.files(
  "/data/biohub/20170327_Psen2S4Ter_RNASeq/data/2_alignedData/bam", 
  pattern = ".bam$", 
  full.names = TRUE
)
param <- ScanBamParam(
  what = c("pos", "qwidth"),
  which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
  flag=scanBamFlag(isUnmappedQuery=FALSE)
)
covFunc <- function(x){
  prop <- order_3$proportion[order_3$filenames == x] %>%
    percent(accuracy = 0.01)
  reads <- counts[order_3$sample[order_3$filenames == x]]
  pileup(x, scanBamParam = param) %>%
    dplyr::filter(strand == str) %>%
    mutate(start = pos, end = pos, count = count/reads) %>% 
    dplyr::select(seqnames, start, end, strand, count) %>%
    makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>%
    DataTrack(
      type = c("heatmap"),
      cex = 0.2, 
      name = prop,
      col.title="black",
      fontface.title=2,
      fontsize=10,
    )
}
covList <- lapply(files, covFunc) 
covOrd <- covList[as.integer(order_3$order)]
```

```{r, fig.height=9}
plotTracks(
  c(gTrack, rrTrack, rTrack, covOrd, gcTrack),
  from = st - 100,
  to = en + 100,
  chromosome = chr,
  sizes = c(1,0.5,1,rep(0.5, length(covOrd)),1), 
  main = sym,
  cex.main = 1.5,
  fontface.main = 1
)
```

***

### *sp3a*

```{r}
sym <- "sp3a"
goi <- genesFilt_Dr[mcols(genesFilt_Dr)$gene_name == sym,]$gene_id
counts <- dgeList_3$counts[goi,]
gm <- subset(exons_Dr, gene == goi) %>%
  subset(transcript %in% rownames(dgeFiltTx_3))
gmRed <- gm %>% 
  GenomicRanges::reduce()
st <- GenomicRanges::start(gmRed) %>% 
  min()
en <- GenomicRanges::end(gmRed) %>% 
  max()
chr <- seqnames(gmRed)@values
str <- gmRed %>% 
  strand() %>%
  .@values
gTrack <- GenomeAxisTrack()
rrTrack <- GeneRegionTrack(
  gmRed,
  name = NULL,
  fill = "darkorange"
)
rTrack <- GeneRegionTrack(
  gm,
  name = NULL
)
```

```{r}
win <- slidingWindows(gmRed, width = 1) %>% 
  unlist()
start <- win %>%
  start()
end <- win %>%
  end()
seq <- getSeq(dna, win)
gc <- seq %>%
  letterFrequency(letters = "GC", OR = "") %>%
  .[,1]
tbl <- tibble(
  start = start,
  end = end,
  gc = gc
) %>% 
  mutate(
    seqnames = chr,
    strand = str,
    cumsum = cumsum(gc),
    cummean = cummean(gc),
    lag = dplyr::lag(cumsum, n = 50, default = NA),
    lead = dplyr::lead(cumsum, n = 50, default = NA),
    gcsum = lead - lag,
    gcprop = gcsum / 101
  ) 
gcTrack <- tbl %>% 
  dplyr::select(seqnames, start, end, strand, gcprop) %>%
  makeGRangesFromDataFrame(
    keep.extra.columns = TRUE
  ) %>%
  DataTrack(
    type = c("b"),
    ylim = c(0, 1), 
    cex = 0.2, 
    name = "GC %",
    col.title="black",
    fontface.title=2,
    fontsize=10,
    grid = TRUE,
    lwd.grid = 0.5,
    lty.grid = "dashed",
    v = 0
  )
```

```{r}
files <- list.files(
  "/data/biohub/20170327_Psen2S4Ter_RNASeq/data/2_alignedData/bam", 
  pattern = ".bam$", 
  full.names = TRUE
)
param <- ScanBamParam(
  what = c("pos", "qwidth"),
  which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
  flag=scanBamFlag(isUnmappedQuery=FALSE)
)
covFunc <- function(x){
  prop <- order_3$proportion[order_3$filenames == x] %>%
    percent(accuracy = 0.01)
  reads <- counts[order_3$sample[order_3$filenames == x]]
  pileup(x, scanBamParam = param) %>%
    dplyr::filter(strand == str) %>%
    mutate(start = pos, end = pos, count = count/reads) %>% 
    dplyr::select(seqnames, start, end, strand, count) %>%
    makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>%
    DataTrack(
      type = c("heatmap"),
      cex = 0.2, 
      name = prop,
      col.title="black",
      fontface.title=2,
      fontsize=10,
    )
}
covList <- lapply(files, covFunc) 
covOrd <- covList[as.integer(order_3$order)]
```

```{r, fig.height=9}
plotTracks(
  c(gTrack, rrTrack, rTrack, covOrd, gcTrack),
  from = st - 100,
  to = en + 100,
  chromosome = chr,
  sizes = c(1,0.5,1,rep(0.5, length(covOrd)),1), 
  main = sym,
  cex.main = 1.5,
  fontface.main = 1
)
```

***

### *sugp1*

```{r}
sym <- "sugp1"
goi <- genesFilt_Dr[mcols(genesFilt_Dr)$gene_name == sym,]$gene_id
counts <- dgeList_3$counts[goi,]
gm <- subset(exons_Dr, gene == goi) %>%
  subset(transcript %in% rownames(dgeFiltTx_3))
gmRed <- gm %>% 
  GenomicRanges::reduce()
st <- GenomicRanges::start(gmRed) %>% 
  min()
en <- GenomicRanges::end(gmRed) %>% 
  max()
chr <- seqnames(gmRed)@values
str <- gmRed %>% 
  strand() %>%
  .@values
gTrack <- GenomeAxisTrack()
rrTrack <- GeneRegionTrack(
  gmRed,
  name = NULL,
  fill = "darkorange"
)
rTrack <- GeneRegionTrack(
  gm,
  name = NULL
)
```

```{r}
win <- slidingWindows(gmRed, width = 1) %>% 
  unlist()
start <- win %>%
  start()
end <- win %>%
  end()
seq <- getSeq(dna, win)
gc <- seq %>%
  letterFrequency(letters = "GC", OR = "") %>%
  .[,1]
tbl <- tibble(
  start = start,
  end = end,
  gc = gc
) %>% 
  mutate(
    seqnames = chr,
    strand = str,
    cumsum = cumsum(gc),
    cummean = cummean(gc),
    lag = dplyr::lag(cumsum, n = 50, default = NA),
    lead = dplyr::lead(cumsum, n = 50, default = NA),
    gcsum = lead - lag,
    gcprop = gcsum / 101
  ) 
gcTrack <- tbl %>% 
  dplyr::select(seqnames, start, end, strand, gcprop) %>%
  makeGRangesFromDataFrame(
    keep.extra.columns = TRUE
  ) %>%
  DataTrack(
    type = c("b"),
    ylim = c(0, 1), 
    cex = 0.2, 
    name = "GC %",
    col.title="black",
    fontface.title=2,
    fontsize=10,
    grid = TRUE,
    lwd.grid = 0.5,
    lty.grid = "dashed",
    v = 0
  )
```

```{r}
files <- list.files(
  "/data/biohub/20170327_Psen2S4Ter_RNASeq/data/2_alignedData/bam", 
  pattern = ".bam$", 
  full.names = TRUE
)
param <- ScanBamParam(
  what = c("pos", "qwidth"),
  which = GRanges(seqnames = chr, ranges = IRanges(st, en), strand = str),
  flag=scanBamFlag(isUnmappedQuery=FALSE)
)
covFunc <- function(x){
  prop <- order_3$proportion[order_3$filenames == x] %>%
    percent(accuracy = 0.01)
  reads <- counts[order_3$sample[order_3$filenames == x]]
  pileup(x, scanBamParam = param) %>%
    dplyr::filter(strand == str) %>%
    mutate(start = pos, end = pos, count = count/reads) %>% 
    dplyr::select(seqnames, start, end, strand, count) %>%
    makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>%
    DataTrack(
      type = c("heatmap"),
      cex = 0.2, 
      name = prop,
      col.title="black",
      fontface.title=2,
      fontsize=10,
    )
}
covList <- lapply(files, covFunc) 
covOrd <- covList[as.integer(order_3$order)]
```

```{r, fig.height=9}
plotTracks(
  c(gTrack, rrTrack, rTrack, covOrd, gcTrack),
  from = st - 100,
  to = en + 100,
  chromosome = chr,
  sizes = c(1,0.5,1,rep(0.5, length(covOrd)),1), 
  main = sym,
  cex.main = 1.5,
  fontface.main = 1
)
```

